{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GIFT数据处理\n",
    "\n",
    "这一步是为了获得所有GIFT需要的数据\n",
    "\n",
    "\n",
    "输入：\n",
    "- GTEx等eQTL\n",
    "- GTEx等对应的基因表达量矩阵\n",
    "- GWAS结果\n",
    "- pgen对应的LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "from pathlib import Path\n",
    "import numpy as np \n",
    "from io import StringIO\n",
    "import pyranges as pr\n",
    "\n",
    "# snpid tools\n",
    "from finemap_tools.utils import add_ID, sort_ID\n",
    "# load eQTLs function\n",
    "from finemap_tools.reader.eQTL import  GTEx_tabix_reader\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# topLoci = (1, 15583355, 15583355) # (chr, start, end)\n",
    "# locus_range = 500 # 100kb \n",
    "# locus_range = locus_range * 1000 \n",
    "# locus_range_tuple = (topLoci[0], topLoci[1] - locus_range, topLoci[2] + locus_range)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Tuple, Dict \n",
    "# 取顺式区域\n",
    "\n",
    "def is_col_map_in_df(df:pd.DataFrame, col_map:Dict):\n",
    "\n",
    "    not_in = []\n",
    "    assert isinstance(col_map, dict), f\"col_map should be a dict {col_map}\"\n",
    "    for _, v in col_map.items():\n",
    "        if v not in df.columns.tolist():\n",
    "            not_in.append(v)\n",
    "    if len(not_in) == 0:\n",
    "        return True\n",
    "    else:\n",
    "        print(f\"{not_in } not in {df.columns.tolist()}\")\n",
    "        return False \n",
    "\n",
    "\n",
    "def detect_in_region(value, region_list):\n",
    "    \"\"\"\n",
    "    Check if a value falls within any of the regions in the given region list.\n",
    "\n",
    "    Parameters:\n",
    "    value (float): The value to be checked.\n",
    "    region_list (list): A list of tuples representing the regions. Each tuple should contain two elements: the lower bound and the upper bound of the region.\n",
    "\n",
    "    Returns:\n",
    "    bool: True if the value falls within any of the regions, False otherwise.\n",
    "    \"\"\"\n",
    "    if isinstance(region_list[0], int):\n",
    "        region_list = [region_list]\n",
    "\n",
    "    for region in region_list:\n",
    "        if all((value >= region[0], value < region[1])):\n",
    "\n",
    "            return True \n",
    "\n",
    "    return False\n",
    "\n",
    "\n",
    "def extract_cis_snps(df, pos_col,one_based_start_end):\n",
    "    selector = df[pos_col].apply(lambda x: detect_in_region(x, one_based_start_end))\n",
    "    return df[selector]\n",
    "\n",
    "def extract_cis_snp_by_anno(eQTL_df:pd.DataFrame, genes_in_locus_df:pd.DataFrame, cis_range:int, eQTL_col_map:Dict, anno_file_col_map:Dict) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Input:\n",
    "    - eQTL_df: with chr pos \n",
    "\n",
    "    Output:\n",
    "    - eQTL_df: with only cis-snp of any gene and have two new columns: gene_name, gene_id, which matched to the value in anno_file_col_map\n",
    "    \n",
    "    \"\"\"\n",
    "    eQTL_chr_col = eQTL_col_map['chr']\n",
    "    eQTL_postion_col = eQTL_col_map['pos']\n",
    "    eQTL_ref_col = eQTL_col_map['ref']\n",
    "    eQTL_alt_col = eQTL_col_map['alt']\n",
    "\n",
    "    eQTL_gene_id = eQTL_col_map['gene_id']\n",
    "\n",
    "    anno_gene_id = anno_file_col_map['gene_id']\n",
    "    anno_gene_name = anno_file_col_map['gene_name']\n",
    "    anno_Start = anno_file_col_map['Start']\n",
    "    anno_End = anno_file_col_map['End']\n",
    "\n",
    "    eQTL_df['ID_sorted'] = add_ID(eQTL_df, [eQTL_chr_col, eQTL_postion_col, eQTL_ref_col, eQTL_alt_col], sort=True, inplace=False)\n",
    "\n",
    "    # 1. eQTLs的基因必须在transcirpt list \n",
    "    # 2. eQTLs必须是cis-SNP\n",
    "\n",
    "    ## 1. eQTLs的基因必须在transcirpt list\n",
    "\n",
    "    print(f\"the genes (from gene annotation file) in the locus: {genes_in_locus_df.shape[0]}\")\n",
    "\n",
    "    gene_in_locus_df_id_Name = genes_in_locus_df[[anno_gene_id, anno_gene_name]]\n",
    "\n",
    "    matched = eQTL_df[eQTL_df[eQTL_gene_id].isin(gene_in_locus_df_id_Name[anno_gene_id])].merge(gene_in_locus_df_id_Name, left_on=eQTL_gene_id, right_on=anno_gene_id, how='inner')\n",
    "\n",
    "    print(f\"matched: {matched[eQTL_gene_id].unique().shape}\")\n",
    "\n",
    "    not_in_eQTL_gene_name = set(gene_in_locus_df_id_Name[anno_gene_name].tolist()) - set(matched[anno_gene_name].unique().tolist())\n",
    "\n",
    "    print(f\"not_in_eQTL_gene: {not_in_eQTL_gene_name}\")\n",
    "    gene_in_analysis = matched[anno_gene_name].unique().tolist()\n",
    "    print(f\"gene_in_analysis: {gene_in_analysis}\")\n",
    "    eQTL_df = matched\n",
    "\n",
    "    genes_in_locus_df = genes_in_locus_df[genes_in_locus_df['Name'].isin(gene_in_analysis)] # update genes_in_locus_df\n",
    "\n",
    "\n",
    "    ## 2. check cis-snp \n",
    "    cis_snp_list = []\n",
    "    for geneName , gene_eQTLs in eQTL_df.groupby(anno_gene_name): \n",
    "        gene_region = genes_in_locus_df[genes_in_locus_df[anno_gene_name] == geneName].iloc[0][[anno_Start, anno_End]].tolist()\n",
    "        upstream = (gene_region[0] - cis_range, gene_region[0])\n",
    "        downstream = (gene_region[1], gene_region[1] + cis_range)\n",
    "        gene_cis_range = [upstream, downstream]\n",
    "\n",
    "        gene_cis_eQTLs_df = extract_cis_snps(gene_eQTLs, eQTL_postion_col, gene_cis_range)\n",
    "\n",
    "        cis_snp_list += gene_cis_eQTLs_df['ID_sorted'].tolist()\n",
    "        print(f\"{geneName} have cis-snp num: {gene_cis_eQTLs_df.shape[0]} while total snp num: {gene_eQTLs.shape[0]} and gene_cis_range: {gene_cis_range}\")\n",
    "\n",
    "\n",
    "    cis_snp_list = set(cis_snp_list)\n",
    "    print(f\"total cis-snp num: {len(cis_snp_list)}\")\n",
    "\n",
    "    print(\"keep only it is cis-snp of any gene\")\n",
    "    print(f\"before filter, eQTL_df shape: {eQTL_df.shape}\")\n",
    "    eQTL_df = eQTL_df[eQTL_df['ID_sorted'].isin(cis_snp_list)]\n",
    "    print(f\"after filter, eQTL_df shape: {eQTL_df.shape}\")\n",
    "    print(f\"after filter, eQTL_df gene num: {eQTL_df['Name'].unique().shape}\")\n",
    "    print(f\"after filter, eQTL_df snp num: {eQTL_df['ID_sorted'].unique().shape}\")\n",
    "    return eQTL_df\n",
    "\n",
    "def GIFT_eQTL_pipline(eQTL_df, genes_in_locus_df, cis_range, eQTL_necessary_col_map, anno_file_necessary_col_map):\n",
    "    ### cal TSTAT\n",
    "    eQTL_df['TSTAT'] = eQTL_df[eQTL_necessary_col_map['beta']].astype(float) / eQTL_df[eQTL_necessary_col_map['se']].astype(float)\n",
    "    print(\"have TSTAT na num is :\",eQTL_df['TSTAT'].isna().sum())\n",
    "    eQTL_df = eQTL_df.dropna(subset=['TSTAT'])\n",
    "    print(f\"after dropna: {eQTL_df.shape[0]}\")\n",
    "\n",
    "    ### extract cis-snp and two new columns: anno_file_necessary_col_map[\"gene_name\"], anno_file_necessary_col_map[\"gene_id\"] and \"ID_sorted\" in eQTL_df\n",
    "    eQTL_df = extract_cis_snp_by_anno(eQTL_df = eQTL_df, genes_in_locus_df=genes_in_locus_df, cis_range=cis_range, eQTL_col_map=eQTL_necessary_col_map, anno_file_col_map=anno_file_necessary_col_map)\n",
    "\n",
    "    ### gene in analysis \n",
    "    gene_in_analysis = eQTL_df[anno_file_necessary_col_map[\"gene_name\"]].unique().tolist()\n",
    "\n",
    "    ### remove any cis-snp not in all gene \n",
    "    gene_nums = len(gene_in_analysis)\n",
    "    missing_cis_snp_rate = eQTL_df.groupby(\"ID_sorted\").apply(lambda x: x.shape[0]/gene_nums).sort_values(ascending=False)\n",
    "    missing_cis_snp = missing_cis_snp_rate[missing_cis_snp_rate == 0].index.tolist()\n",
    "    print(f\"missing rate >0: {len(missing_cis_snp)}\")\n",
    "    eQTL_df = eQTL_df[~eQTL_df['ID_sorted'].isin(missing_cis_snp)]\n",
    "    print(f\"eQTL_df: {eQTL_df.shape}\")\n",
    "\n",
    "    return eQTL_df, gene_in_analysis\n",
    "\n",
    "default_eQTL_necessary_col = {\n",
    "    'chr': 'chromosome',\n",
    "    'pos': 'position',\n",
    "    'ref': 'ref',\n",
    "    'alt': 'alt',\n",
    "    'beta': 'beta',\n",
    "    'se': 'se',\n",
    "    'gene_id': 'gene_id'\n",
    "}\n",
    "default_anno_file_necessary_col = {\n",
    "    \"gene_id\": \"gene_id\",\n",
    "    \"gene_name\": \"Name\",\n",
    "    \"Start\": \"Start\",\n",
    "    \"End\": \"End\"\n",
    "}\n",
    "default_gwas_necessary_col = {\n",
    "    'chr': 'chrom',\n",
    "    'pos': 'pos',\n",
    "    'ref': 'ref',\n",
    "    'alt': 'alt',\n",
    "    'beta': 'beta',\n",
    "    'se': 'sebeta'\n",
    "}\n",
    "\n",
    "def reorder_snplist_by_cis_snp_gene(\n",
    "        eQTL_df:pd.DataFrame,\n",
    "        genes_in_locus_df:pd.DataFrame,\n",
    "        cis_range:int,\n",
    "        eQTL_col_map:Dict,\n",
    "        anno_file_col_map:Dict,\n",
    "):\n",
    "    snplist = [] \n",
    "    pindex = []\n",
    "    gene = [] \n",
    "    eQTL_postion_col = eQTL_col_map['pos']\n",
    "    eQTL_ID_col = \"ID\" # from pvar \n",
    "    anno_gene_name = anno_file_col_map['gene_name']\n",
    "    anno_Start = anno_file_col_map['Start']\n",
    "    anno_End = anno_file_col_map['End']\n",
    "\n",
    "    for geneName, gene_eQTLs in eQTL_df.groupby(anno_gene_name):\n",
    "        gene_region = genes_in_locus_df[genes_in_locus_df[anno_gene_name] == geneName].iloc[0][[anno_Start, anno_End]].tolist()\n",
    "        upstream = (gene_region[0] - cis_range, gene_region[0])\n",
    "        downstream = (gene_region[1], gene_region[1] + cis_range)\n",
    "        gene_cis_range = [upstream, downstream]\n",
    "\n",
    "        # filter eQTLs only keep the cis snps at cis range \n",
    "        gene_cis_eQTLs_df = extract_cis_snps(gene_eQTLs, eQTL_postion_col, gene_cis_range).set_index(eQTL_ID_col)[[\"TSTAT\"]]\n",
    "        if (na_num:=sum(gene_cis_eQTLs_df[['TSTAT']].isna().sum())) > 0:\n",
    "            print(f\"gene: {geneName} has {na_num} NA value in TSTAT\")\n",
    "            # gene_cis_eQTLs_df = gene_cis_eQTLs_df.dropna()\n",
    "            raise Exception(f\"gene: {geneName} has {na_num} NA value in TSTAT, plz check the data to ensure all cis-snp have TSTAT in eQTL data.\")\n",
    "            \n",
    "        if gene_cis_eQTLs_df.shape[0] > 0 : # if there is any snp in the cis range\n",
    "            # gene_eQTLs_cis[geneName] = gene_cis_eQTLs_df # TODO: append is not necessary \n",
    "\n",
    "            gene.append(geneName)\n",
    "            snplist += gene_cis_eQTLs_df.index.tolist() # add cis snps to snplist\n",
    "            pindex.append(gene_cis_eQTLs_df.shape[0]) # add the number of cis snps in the gene to pindex\n",
    "        else:\n",
    "            print(f\"gene: {geneName} has no snp in cis range\")\n",
    "\n",
    "\n",
    "\n",
    "    print(f\"total cis snps: {len(snplist)}\")\n",
    "    for snpnum, g in zip(pindex, gene):\n",
    "        print(f\"gene: {g} has {snpnum} cis snps\")\n",
    "\n",
    "    return snplist, pindex, gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ToGiFTFormatH5(eQTL_path, \n",
    "                   gwasPath,\n",
    "                   locus_range_tuple,\n",
    "                    ensemble_anno_hs_path,\n",
    "                    pfile_path, \n",
    "                    gene_expression_path,\n",
    "\n",
    "                   cis_range = 100, # kb \n",
    "                   gene_expression_format=\"GTEx\",\n",
    "                   eQTL_col_map=None,\n",
    "                   gwas_col_map=None,\n",
    "                   anno_file_col_map=None\n",
    "                     ):\n",
    "    \"\"\"\n",
    "    Input Format:\n",
    "    1. eQTLpath: 应当是包含所有探针的eQTL文件，并且必须有对应的tabix索引；默认是对GTEx格式进行处理。如来源其他的地方，至少包含以下列：chr, pos, ref, alt ,beta, se, gene_id 7列，\n",
    "    2. gwasPath: 应当是一个表型的GWAS文件，并且必须有tabix索引；默认是对GWASFormatted的格式进行处理，如果来源于其他的地方，至少包含以下列：chr, pos, ref, alt ,beta, se\n",
    "    3. pvar_path \n",
    "    \n",
    "    \"\"\"\n",
    "    # locus_range_update\n",
    "    \n",
    "    ## extract gene annotation from gff file \n",
    "    print(\"-------------------- loading anno file --------------------\")\n",
    "    try:\n",
    "        ### load gff3\n",
    "        ensemble_anno_hs = pr.read_gff3(ensemble_anno_hs_path) # load gff3\n",
    "        ensemble_anno_hs_gene = ensemble_anno_hs[ensemble_anno_hs.Feature == 'gene'] \n",
    "        genes_in_locus_df = ensemble_anno_hs_gene[str(locus_range_tuple[0]), locus_range_tuple[1]:locus_range_tuple[2]].as_df()\n",
    "\n",
    "        ### check col_map\n",
    "        anno_file_necessary_col_map = default_anno_file_necessary_col\n",
    "        if anno_file_col_map is not None:\n",
    "            anno_file_necessary_col_map.update(anno_file_col_map)\n",
    "        if not is_col_map_in_df(genes_in_locus_df, anno_file_necessary_col_map):\n",
    "            raise ValueError(f\"Error: anno_file_col_map is not valid: {anno_file_col_map}\")\n",
    "\n",
    "        ### step2 update new locus range with considering the boundary genes with cis-range \n",
    "        print(f\"Considering the boundary of locus with cis range {cis_range//1000}kb with the old locus range: {locus_range_tuple}\")\n",
    "        locus_range_tuple_update = (locus_range_tuple[0], genes_in_locus_df['Start'].min() + cis_range, genes_in_locus_df['End'].max() + cis_range)\n",
    "        print(f\"new locus range: {locus_range_tuple_update}\")\n",
    "\n",
    "        ### locus_range_tuple_update\n",
    "        locus_range_tuple = locus_range_tuple_update\n",
    "    \n",
    "    except:\n",
    "        raise ValueError(f\"Error: ensemble_anno_hs_path is not valid: {ensemble_anno_hs_path}\")\n",
    "    \n",
    "    ## locus range\n",
    "    chr = locus_range_tuple[0]\n",
    "    start = locus_range_tuple[1]\n",
    "    end = locus_range_tuple[2]\n",
    "    region = f\"{chr}:{start}-{end}\"\n",
    "\n",
    "    ## eQTL file parse \n",
    "    ### load eQTL file\n",
    "    eQTL_df = GTEx_tabix_reader(eQTL_path, region)\n",
    "\n",
    "    ### check col_map\n",
    "    eQTL_necessary_col_map = default_eQTL_necessary_col\n",
    "    if eQTL_col_map is not None:\n",
    "        eQTL_necessary_col_map.update(eQTL_col_map)\n",
    "\n",
    "    if not is_col_map_in_df(eQTL_df, eQTL_necessary_col_map):\n",
    "        raise ValueError(f\"Error: eQTL_col_map is not valid: {eQTL_col_map}\")\n",
    "\n",
    "    ### run GIFT eQTL pipline\n",
    "    eQTL_df, gene_in_analysis = GIFT_eQTL_pipline(\n",
    "        eQTL_df= eQTL_df,\n",
    "        genes_in_locus_df=genes_in_locus_df,\n",
    "        cis_range=cis_range,\n",
    "        eQTL_necessary_col_map=eQTL_necessary_col_map,\n",
    "        anno_file_necessary_col_map=anno_file_necessary_col_map\n",
    "    )\n",
    "    # ### cal TSTAT\n",
    "    # eQTL_df['TSTAT'] = eQTL_df[eQTL_necessary_col_map['beta']].astype(float) / eQTL_df[eQTL_necessary_col_map['se']].astype(float)\n",
    "    # print(\"have TSTAT na num is :\",eQTL_df['TSTAT'].isna().sum())\n",
    "    # eQTL_df = eQTL_df.dropna(subset=['TSTAT'])\n",
    "    # print(f\"after dropna: {eQTL_df.shape[0]}\")\n",
    "\n",
    "    # ### extract cis-snp and two new columns: anno_file_necessary_col_map[\"gene_name\"], anno_file_necessary_col_map[\"gene_id\"] and \"ID_sorted\" in eQTL_df\n",
    "    # eQTL_df = extract_cis_snp_by_anno(eQTL_df = eQTL_df, genes_in_locus_df=genes_in_locus_df, cis_range=cis_range, eQTL_col_map=eQTL_necessary_col_map, anno_file_col_map=anno_file_necessary_col_map)\n",
    "\n",
    "    # ### gene in analysis \n",
    "    # gene_in_analysis = eQTL_df[anno_file_necessary_col_map[\"gene_name\"]].unique().tolist()\n",
    "\n",
    "    # ### remove any cis-snp not in all gene \n",
    "    # gene_nums = len(gene_in_analysis)\n",
    "    # missing_cis_snp_rate = eQTL_df.groupby(\"ID_sorted\").apply(lambda x: x.shape[0]/gene_nums).sort_values(ascending=False)\n",
    "    # missing_cis_snp = missing_cis_snp_rate[missing_cis_snp_rate == 0].index.tolist()\n",
    "    # print(f\"missing rate >0: {len(missing_cis_snp)}\")\n",
    "    # eQTL_df = eQTL_df[~eQTL_df['ID_sorted'].isin(missing_cis_snp)]\n",
    "    # print(f\"eQTL_df: {eQTL_df.shape}\")\n",
    "\n",
    "\n",
    "    ## gwas file parse \n",
    "    ### load gwas file\n",
    "    from finemap_tools.reader.gwas import load_GWASFormated_file\n",
    "\n",
    "    GWAS_df = load_GWASFormated_file(file_path=gwasPath, region=region)\n",
    "    ### check col_map\n",
    "    gwas_necessary_col_map = default_gwas_necessary_col\n",
    "    if gwas_col_map is not None:\n",
    "        gwas_necessary_col_map.update(gwas_col_map)\n",
    "\n",
    "    if not is_col_map_in_df(GWAS_df, gwas_necessary_col_map):\n",
    "        raise ValueError(f\"Error: gwas_col_map is not valid: {gwas_col_map}\")\n",
    "    \n",
    "    ### run GIFT GWAS pipline\n",
    "    gwas_chr_col, gwas_pos_col, gwas_ref_col, gwas_alt_col, gwas_beta_col, gwas_se_col = gwas_necessary_col_map['chr'], gwas_necessary_col_map['pos'], gwas_necessary_col_map['ref'], gwas_necessary_col_map['alt'], gwas_necessary_col_map['beta'], gwas_necessary_col_map['se']\n",
    "\n",
    "    GWAS_df['ID_sorted'] = add_ID(GWAS_df, [gwas_chr_col, gwas_pos_col, gwas_ref_col, gwas_alt_col], inplace=False, sort = True)\n",
    "    GWAS_df[\"TSTAT\"] = GWAS_df[gwas_beta_col].astype(float) / GWAS_df[gwas_se_col].astype(float)\n",
    "    GWAS_df = GWAS_df.dropna(subset=[\"TSTAT\"])\n",
    "\n",
    "    ## check pfile for LD \n",
    "\n",
    "    from finemap_tools.reader.plink import read_pvar\n",
    "    ### load pvar\n",
    "    pvar_path = pfile_path + \".pvar\"\n",
    "    pvar_df = read_pvar(pvar_path)\n",
    "    pvar_df['ID_sorted'] = sort_ID(pvar_df, \"ID\", inplace=False)\n",
    "\n",
    "    ## intersection of eQTL and GWAS and pvar\n",
    "\n",
    "    pvar_SNP = pvar_df['ID_sorted'].tolist()\n",
    "    GWAS_SNP = GWAS_df['ID_sorted'].tolist()\n",
    "    eQTL_SNP = eQTL_df['ID_sorted'].tolist()\n",
    "\n",
    "    print(f\"pvar_SNP: {len(pvar_SNP)}\")\n",
    "    print(f\"GWAS_SNP: {len(GWAS_SNP)}\")\n",
    "    print(f\"eQTL_SNP: {len(eQTL_SNP)}\")\n",
    "\n",
    "    pvar_SNP_set = set(pvar_SNP)\n",
    "    GWAS_SNP_set = set(GWAS_SNP)\n",
    "    eQTL_SNP_set = set(eQTL_SNP)\n",
    "\n",
    "    intersection = pvar_SNP_set & GWAS_SNP_set & eQTL_SNP_set\n",
    "    print(f\"intersection: {len(intersection)}\")\n",
    "\n",
    "    pvar_df_intersction = pvar_df[pvar_df['ID_sorted'].isin(intersection)]\n",
    "    # add real ID \n",
    "    GWAS_df_intersction = GWAS_df[GWAS_df['ID_sorted'].isin(intersection)].merge(pvar_df_intersction[['ID_sorted', \"ID\"]], on=\"ID_sorted\")\n",
    "    # add real ID \n",
    "    eQTL_df_intersction = eQTL_df[eQTL_df['ID_sorted'].isin(intersection)].merge(pvar_df_intersction[['ID_sorted', \"ID\"]], on=\"ID_sorted\")\n",
    "\n",
    "    SNP_ID_real = pvar_df_intersction[\"ID\"].tolist()\n",
    "\n",
    "    ## check afreq by plink2 and filter freq < 1e-2\n",
    "    import tempfile\n",
    "    from finemap_tools.plink import plink2_cal_freq\n",
    "\n",
    "    with tempfile.TemporaryDirectory() as tmpdirname:\n",
    "        print('created temporary directory', tmpdirname)\n",
    "        freq = plink2_cal_freq(pgen=pfile_path, snplist=SNP_ID_real, outSuffix=tmpdirname + '/plink2_cal_freq')\n",
    "        # print(freq)\n",
    "    \n",
    "    final_SNP_list = freq[freq['ALT_FREQS'] >1e-2]['ID'].tolist() \n",
    "\n",
    "    eQTL_df = eQTL_df_intersction[eQTL_df_intersction['ID'].isin(final_SNP_list)]\n",
    "    GWAS_df = GWAS_df_intersction[GWAS_df_intersction['ID'].isin(final_SNP_list)]\n",
    "\n",
    "    print(f\"eQTL_df_passed_qc: {eQTL_df.shape}\")\n",
    "    print(f\"GWAS_df_passed_qc: {GWAS_df.shape}\")\n",
    "\n",
    "    ### check any gene in eQTL_df has no cis-snp after QC\n",
    "    gene_cis_snp_count = eQTL_df.groupby(anno_file_necessary_col_map[\"gene_name\"]).apply(lambda x: x.shape[0])\n",
    "    gene_cis_snp_count = gene_cis_snp_count[gene_cis_snp_count == 0]\n",
    "    if gene_cis_snp_count.shape[0] > 0:\n",
    "        # print(f\"gene_cis_snp_count: {gene_cis_snp_count}\")\n",
    "        print(f\"After QC theses genes have no cis-snp: {gene_cis_snp_count.index.tolist()}\")\n",
    "        eQTL_df = eQTL_df[~eQTL_df[anno_file_necessary_col_map[\"gene_name\"]].isin(gene_cis_snp_count.index.tolist())]\n",
    "    \n",
    "    gene_in_analysis = eQTL_df[anno_file_necessary_col_map[\"gene_name\"]].unique().tolist()\n",
    "\n",
    "    ## cal gene expression df \n",
    "\n",
    "    ### load gene expression file\n",
    "\n",
    "    from functools import reduce\n",
    "    # snplist is all genes snp concat together, so may have duplicate snp\n",
    "    def load_GTEx_gene_expression(geneExpression):\n",
    "        geneExpression_df = pd.read_csv(geneExpression, sep = '\\t', skiprows=2 ).iloc[:, 2:].set_index('Description').T\n",
    "        return geneExpression_df\n",
    "    geneExpression_df = load_GTEx_gene_expression(gene_expression_path)[gene_in_analysis]\n",
    "    geneExpression_df_corr = geneExpression_df.corr()\n",
    "\n",
    "    ## cal LD \n",
    "    from finemap_tools.plink import plink2_cal_LD\n",
    "    with tempfile.TemporaryDirectory() as tmpdirname:\n",
    "        print('created temporary directory', tmpdirname)\n",
    "        LD_df = plink2_cal_LD(pgen=pfile_path, snplist=final_SNP_list, outSuffix=tmpdirname + '/plink2_cal_ld')\n",
    "\n",
    "    ##  raarange cis-snp order of eQTL_df, GWAS_df, LD \n",
    "    ### generate snplist order \n",
    "    snplist, pindex, gene = reorder_snplist_by_cis_snp_gene(\n",
    "        eQTL_df=eQTL_df,\n",
    "        genes_in_locus_df=genes_in_locus_df, # only provide Start End info\n",
    "        cis_range=cis_range,\n",
    "        eQTL_col_map=eQTL_necessary_col_map,\n",
    "        anno_file_col_map=anno_file_necessary_col_map\n",
    "    )\n",
    "\n",
    "    ### reorder eQTL_df, GWAS_df, LD_df\n",
    "    snplist_df = pd.DataFrame(snplist, columns=[\"ID\"])\n",
    "\n",
    "    #Zscore_GWAS 计算\n",
    "    Zscore_GWAS_df = snplist_df.merge(GWAS_df, how=\"left\", on = \"ID\")[[ \"ID\", \"TSTAT\"]].set_index(\"ID\")\n",
    "    print(f\"Zscore_GWAS_df: {Zscore_GWAS_df.shape}\")\n",
    "    # Zscore_eQTLs 计算\n",
    "    Zscore_eQTLs_list = []\n",
    "\n",
    "    for geneName, gene_eQTLs in eQTL_df.groupby(anno_file_necessary_col_map[\"gene_name\"]):\n",
    "\n",
    "        Zscore_eQTLs = snplist_df.merge(gene_eQTLs[[ \"ID\", \"TSTAT\"]].rename(columns = {\"TSTAT\": geneName}), how=\"left\", on = \"ID\").set_index(\"ID\")\n",
    "        Zscore_eQTLs_list.append(Zscore_eQTLs)\n",
    "\n",
    "\n",
    "    assert True == all((x.index == Zscore_eQTLs_list[0].index ).all() for x in Zscore_eQTLs_list)  # check if all index are the same\n",
    "    Zscore_eQTLs_df = pd.concat(Zscore_eQTLs_list, axis = 1)\n",
    "    print(f\"Zscore_eQTLs_df: {Zscore_eQTLs_df.shape}\")\n",
    "    assert Zscore_eQTLs_df.isna().sum().sum() == 0, \"eQTLs Zscore has NA value, plz check the data to ensure all cis-snp have TSTAT in eQTL data.\"\n",
    "\n",
    "    # LD 重排顺序\n",
    "    LD_eQTL = LD_df.loc[snplist, snplist]\n",
    "    LD_GWAS = LD_df.loc[snplist, snplist]\n",
    "\n",
    "    # expression corr \n",
    "    expression_corr = geneExpression_df_corr.loc[gene, gene]\n",
    "\n",
    "    # check all \n",
    "    print(\"check all at final part~\")\n",
    "    print(Zscore_GWAS_df.shape[0], sum(pindex))\n",
    "    assert sum(pindex) == Zscore_GWAS_df.shape[0], print(f\"eQTLs snps are not equal to GWAS snps, eQTLs snps are {sum(pindex)} and GWAS snps are {Zscore_GWAS_df.shape[0]}\")\n",
    "    assert [col for col in Zscore_eQTLs_df.columns] == gene, print(f\"gene is {gene}, while Zscore_eQTLs columns is {Zscore_eQTLs_df.columns.tolist()}\")\n",
    "    assert Zscore_GWAS_df.index.tolist() == Zscore_eQTLs_df.index.tolist()\n",
    "    assert Zscore_GWAS_df.index.tolist() == snplist\n",
    "    assert Zscore_eQTLs_df.index.tolist() == LD_eQTL.index.tolist()\n",
    "    assert Zscore_GWAS_df.index.tolist() == LD_GWAS.index.tolist()\n",
    "    assert expression_corr.index.tolist() == gene\n",
    "\n",
    "    return gene, snplist, pindex, Zscore_eQTLs_df, Zscore_GWAS_df, LD_eQTL, LD_GWAS, expression_corr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis_range: 10kb\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# cis-snp range \n",
    "cis_range = 10 # 100kb \n",
    "print(f\"cis_range: {cis_range}kb\")\n",
    "cis_range = cis_range * 1000\n",
    "\n",
    "# locus range \n",
    "locus_range_tuple = (1, 15339446, 15817866)\n",
    "base_dir = \"/home/xutingfeng/GIFT/\"\n",
    "# necessary file path \n",
    "## ensemble annotation\n",
    "ensemble_anno_hs_path = base_dir + \"hs_anno/refseq/Homo_sapiens.GRCh38.111.gff3.gz\" # refseq annotation\n",
    "\n",
    "## GTEx tissue eQTL file\n",
    "eQTLfile_path = base_dir+ \"GTEx_V8/ge/Kidney_Cortex.tsv.gz\"\n",
    "\n",
    "## gwas file path \n",
    "gwasPath = base_dir +  \"data/gwas/T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz\"\n",
    "\n",
    "## pvar path \n",
    "pfile_path = base_dir + \"data/pgen/DNAJC16_GRCh38_union\"\n",
    "\n",
    "## gene expression\n",
    "\n",
    "gene_expression_path = base_dir + \"data/expression/gene_reads_2017-06-05_v8_kidney_cortex.gct.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------- loading anno file --------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/xutingfeng/miniforge3/lib/python3.10/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/home/xutingfeng/miniforge3/lib/python3.10/site-packages/pyranges/methods/init.py:160: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  empty_removed = df.groupby([\"Chromosome\", \"Strand\"])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Considering the boundary of locus with cis range 10kb with the old locus range: (1, 15339446, 15817866)\n",
      "new locus range: (1, 15246520, 15819348)\n",
      "have TSTAT na num is : 3090\n",
      "after dropna: 103680\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3637/2822344048.py:65: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  eQTL_df['ID_sorted'] = add_ID(eQTL_df, [eQTL_chr_col, eQTL_postion_col, eQTL_ref_col, eQTL_alt_col], sort=True, inplace=False)\n",
      "/tmp/ipykernel_3637/2822344048.py:130: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  missing_cis_snp_rate = eQTL_df.groupby(\"ID_sorted\").apply(lambda x: x.shape[0]/gene_nums).sort_values(ascending=False)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the genes (from gene annotation file) in the locus: 15\n",
      "matched: (14,)\n",
      "not_in_eQTL_gene: {'RSC1A1'}\n",
      "gene_in_analysis: ['CELA2B', 'AGMAT', 'EFHD2', 'DNAJC16', 'SLC25A34', 'FHAD1', 'UQCRHL', 'DDI2', 'FBLIM1', 'CTRC', 'CELA2A', 'TMEM82', 'PLEKHM2', 'CASP9']\n",
      "AGMAT have cis-snp num: 120 while total snp num: 2210 and gene_cis_range: [(15561698, 15571698), (15585051, 15595051)]\n",
      "CASP9 have cis-snp num: 67 while total snp num: 2210 and gene_cis_range: [(15480831, 15490831), (15526534, 15536534)]\n",
      "CELA2A have cis-snp num: 58 while total snp num: 2210 and gene_cis_range: [(15446727, 15456727), (15472091, 15482091)]\n",
      "CELA2B have cis-snp num: 93 while total snp num: 2210 and gene_cis_range: [(15455908, 15465908), (15491395, 15501395)]\n",
      "CTRC have cis-snp num: 65 while total snp num: 2210 and gene_cis_range: [(15428441, 15438441), (15449242, 15459242)]\n",
      "DDI2 have cis-snp num: 59 while total snp num: 2210 and gene_cis_range: [(15607457, 15617457), (15669044, 15679044)]\n",
      "DNAJC16 have cis-snp num: 134 while total snp num: 2210 and gene_cis_range: [(15516812, 15526812), (15592379, 15602379)]\n",
      "EFHD2 have cis-snp num: 70 while total snp num: 2210 and gene_cis_range: [(15399887, 15409887), (15430339, 15440339)]\n",
      "FBLIM1 have cis-snp num: 66 while total snp num: 2210 and gene_cis_range: [(15746606, 15756606), (15786594, 15796594)]\n",
      "FHAD1 have cis-snp num: 43 while total snp num: 2210 and gene_cis_range: [(15226520, 15236520), (15400283, 15410283)]\n",
      "PLEKHM2 have cis-snp num: 66 while total snp num: 2210 and gene_cis_range: [(15674319, 15684319), (15734769, 15744769)]\n",
      "SLC25A34 have cis-snp num: 63 while total snp num: 2210 and gene_cis_range: [(15726257, 15736257), (15741392, 15751392)]\n",
      "TMEM82 have cis-snp num: 66 while total snp num: 2210 and gene_cis_range: [(15732498, 15742498), (15747982, 15757982)]\n",
      "UQCRHL have cis-snp num: 117 while total snp num: 2210 and gene_cis_range: [(15797168, 15807168), (15809348, 15819348)]\n",
      "total cis-snp num: 824\n",
      "keep only it is cis-snp of any gene\n",
      "before filter, eQTL_df shape: (30940, 21)\n",
      "after filter, eQTL_df shape: (11536, 21)\n",
      "after filter, eQTL_df gene num: (14,)\n",
      "after filter, eQTL_df snp num: (824,)\n",
      "missing rate >0: 0\n",
      "eQTL_df: (11536, 21)\n",
      "tabix have a header, so will take the first line as header and remove it.\n",
      "pvar_SNP: 138075\n",
      "GWAS_SNP: 4496\n",
      "eQTL_SNP: 11536\n",
      "intersection: 812\n",
      "created temporary directory /tmp/tmpwk8xiwp9\n",
      "eQTL_df_passed_qc: (10010, 22)\n",
      "GWAS_df_passed_qc: (715, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3637/1729080893.py:167: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  gene_cis_snp_count = eQTL_df.groupby(anno_file_necessary_col_map[\"gene_name\"]).apply(lambda x: x.shape[0])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "created temporary directory /tmp/tmpfn1_w_f7\n",
      "total cis snps: 924\n",
      "gene: AGMAT has 101 cis snps\n",
      "gene: CASP9 has 65 cis snps\n",
      "gene: CELA2A has 52 cis snps\n",
      "gene: CELA2B has 86 cis snps\n",
      "gene: CTRC has 57 cis snps\n",
      "gene: DDI2 has 52 cis snps\n",
      "gene: DNAJC16 has 115 cis snps\n",
      "gene: EFHD2 has 58 cis snps\n",
      "gene: FBLIM1 has 60 cis snps\n",
      "gene: FHAD1 has 34 cis snps\n",
      "gene: PLEKHM2 has 48 cis snps\n",
      "gene: SLC25A34 has 48 cis snps\n",
      "gene: TMEM82 has 45 cis snps\n",
      "gene: UQCRHL has 103 cis snps\n",
      "Zscore_GWAS_df: (924, 1)\n",
      "Zscore_eQTLs_df: (924, 14)\n",
      "check all at final part~\n",
      "924 924\n"
     ]
    }
   ],
   "source": [
    "gene, snplist, pindex, Zscore_eQTLs_df, Zscore_GWAS_df, LD_eQTL, LD_GWAS, expression_corr = ToGiFTFormatH5(\n",
    "    eQTL_path= eQTLfile_path,\n",
    "    gwasPath= gwasPath,\n",
    "    ensemble_anno_hs_path=ensemble_anno_hs_path,\n",
    "    pfile_path= pfile_path,\n",
    "    gene_expression_path=gene_expression_path,\n",
    "    cis_range = cis_range,\n",
    "    locus_range_tuple = locus_range_tuple\n",
    "\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to h5 \n",
    "import h5py\n",
    "\n",
    "with h5py.File(base_dir+\"GIFT.h5\", \"w\") as f:\n",
    "    f.create_group(\"data\")\n",
    "    f['data'].create_dataset('gene', data=gene)\n",
    "    f['data'].create_dataset('snplist', data=snplist)\n",
    "    f['data'].create_dataset('pindex', data=pindex)\n",
    "    f['data'].create_dataset('Zscore_eQTLs_df', data=Zscore_eQTLs_df)\n",
    "    f['data'].create_dataset('Zscore_GWAS_df', data=Zscore_GWAS_df)\n",
    "    f['data'].create_dataset('geneExpression_corr_df', data=expression_corr)\n",
    "    f['data'].create_dataset('LD_eQTL', data=LD_eQTL)\n",
    "    f['data'].create_dataset('LD_GWAS', data=LD_GWAS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14587143</td>\n",
       "      <td>1:14587143:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14587157</td>\n",
       "      <td>1:14587157:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14587159</td>\n",
       "      <td>1:14587159:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14587190</td>\n",
       "      <td>1:14587190:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14587192</td>\n",
       "      <td>1:14587192:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109352</th>\n",
       "      <td>1</td>\n",
       "      <td>16587077</td>\n",
       "      <td>1:16587077:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109353</th>\n",
       "      <td>1</td>\n",
       "      <td>16587078</td>\n",
       "      <td>1:16587078:G:T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109354</th>\n",
       "      <td>1</td>\n",
       "      <td>16587079</td>\n",
       "      <td>1:16587079:G:T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109355</th>\n",
       "      <td>1</td>\n",
       "      <td>16587082</td>\n",
       "      <td>1:16587082:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109356</th>\n",
       "      <td>1</td>\n",
       "      <td>16587084</td>\n",
       "      <td>1:16587084:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>109357 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM       POS              ID REF ALT\n",
       "0            1  14587143  1:14587143:C:T   T   C\n",
       "1            1  14587157  1:14587157:C:T   C   T\n",
       "2            1  14587159  1:14587159:A:G   G   A\n",
       "3            1  14587190  1:14587190:A:C   C   A\n",
       "4            1  14587192  1:14587192:C:T   T   C\n",
       "...        ...       ...             ...  ..  ..\n",
       "109352       1  16587077  1:16587077:A:G   A   G\n",
       "109353       1  16587078  1:16587078:G:T   T   G\n",
       "109354       1  16587079  1:16587079:G:T   T   G\n",
       "109355       1  16587082  1:16587082:C:T   T   C\n",
       "109356       1  16587084  1:16587084:A:G   A   G\n",
       "\n",
       "[109357 rows x 5 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from finemap_tools.reader.plink import read_pvar\n",
    "\n",
    "pvar = read_pvar(\"data/pgen/DNAJC16_ukb.pvar\")\n",
    "pvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import overload\n",
    "@overload\n",
    "def is_ambiguous_alleles(id:str, ref:str=None, alt:str=None):...\n",
    "@overload\n",
    "def is_ambiguous_alleles( ref:str, alt:str,id:str=None):...\n",
    "\n",
    "def is_ambiguous_alleles(id:str=None, ref:str=None, alt:str=None):\n",
    "    \"\"\"\n",
    "    check if ref and alt is ambiguous alleles\n",
    "    \"\"\"\n",
    "    if id is not None:\n",
    "        chr, pos ,a1, a2 = id.split(\":\")\n",
    "        # alt = id.split(\":\")[3]\n",
    "    elif ref is not None and alt is not None:\n",
    "        a1 = ref\n",
    "        a2 = alt\n",
    "    else:\n",
    "        raise ValueError(\"plz provide id or ref and alt\")\n",
    "    \n",
    "    if ref == \"A\" and alt == \"T\":\n",
    "        return True\n",
    "    elif ref == \"T\" and alt == \"A\":\n",
    "        return True\n",
    "    elif ref == \"C\" and alt == \"G\":\n",
    "        return True\n",
    "    elif ref == \"G\" and alt == \"C\":\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "from typing import List\n",
    "@overload \n",
    "def find_Biallelic_snp(snp_id_list:List[str], id_sep:str):...\n",
    "@overload\n",
    "def find_Biallelic_snp(snplist_df:pd.DataFrame, pos_col:str, id_col:str, chr_col:str):...\n",
    "def find_Biallelic_snp(snp_id_list:List[str]=None,id_sep:str=None,snplist_df:pd.DataFrame=None, pos_col:str=None, id_col:str=None, chr_col:str=None):\n",
    "    if snp_id_list is not None and id_sep is not None:\n",
    "        snplist_df = pd.DataFrame([[i] + i.split(id_sep) for i in snp_id_list], columns=[\"ID\",\"chr\", \"pos\", \"A1\", \"A2\"])\n",
    "        pos_col = \"pos\"\n",
    "        id_col = \"ID\"\n",
    "        chr_col = \"chr\"\n",
    "    elif snplist_df is not None and pos_col is not None and id_col is not None:\n",
    "        snplist_df = snplist_df \n",
    "\n",
    "    pos_is_Biallelic = snplist_df.groupby([chr_col, pos_col]).apply(lambda x:  reset_df if (reset_df := x.drop_duplicates(subset=[id_col])).shape[0] > 1 else None).reset_index(drop=True)\n",
    "\n",
    "\n",
    "    if len(pos_is_Biallelic.shape) == 0:\n",
    "        return None \n",
    "    else:\n",
    "        return pos_is_Biallelic\n",
    "    # return pos_is_Biallelic\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_27093/2044317120.py:44: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  pos_is_Biallelic = snplist_df.groupby([chr_col, pos_col]).apply(lambda x:  reset_df if (reset_df := x.drop_duplicates(subset=[id_col])).shape[0] > 1 else None).reset_index(drop=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:1234:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>1234</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:1234:C:G</td>\n",
       "      <td>1</td>\n",
       "      <td>1234</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID chr   pos A1 A2\n",
       "0  1:1234:A:T   1  1234  A  T\n",
       "1  1:1234:C:G   1  1234  C  G"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_Biallelic_snp([\"1:1234:A:T\", \"1:1234:C:G\", \"2:12345:C:G\"], \":\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_27093/3581233435.py:45: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  pos_is_Biallelic = snplist_df.groupby([chr_col, pos_col]).apply(lambda x:  reset_df if (reset_df := x.drop_duplicates(subset=[id_col])).shape[0] > 1 else None).reset_index(drop=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14598639</td>\n",
       "      <td>1:14598639:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14598639</td>\n",
       "      <td>1:14598639:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14598640</td>\n",
       "      <td>1:14598640:C:G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14598640</td>\n",
       "      <td>1:14598640:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14598663</td>\n",
       "      <td>1:14598663:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10852</th>\n",
       "      <td>1</td>\n",
       "      <td>16586951</td>\n",
       "      <td>1:16586951:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10853</th>\n",
       "      <td>1</td>\n",
       "      <td>16586959</td>\n",
       "      <td>1:16586959:C:G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10854</th>\n",
       "      <td>1</td>\n",
       "      <td>16586959</td>\n",
       "      <td>1:16586959:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10855</th>\n",
       "      <td>1</td>\n",
       "      <td>16587021</td>\n",
       "      <td>1:16587021:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10856</th>\n",
       "      <td>1</td>\n",
       "      <td>16587021</td>\n",
       "      <td>1:16587021:G:T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10857 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #CHROM       POS              ID REF ALT\n",
       "0           1  14598639  1:14598639:A:C   C   A\n",
       "1           1  14598639  1:14598639:C:T   C   T\n",
       "2           1  14598640  1:14598640:C:G   G   C\n",
       "3           1  14598640  1:14598640:G:T   G   T\n",
       "4           1  14598663  1:14598663:A:G   G   A\n",
       "...       ...       ...             ...  ..  ..\n",
       "10852       1  16586951  1:16586951:G:T   G   T\n",
       "10853       1  16586959  1:16586959:C:G   G   C\n",
       "10854       1  16586959  1:16586959:G:T   G   T\n",
       "10855       1  16587021  1:16587021:A:G   G   A\n",
       "10856       1  16587021  1:16587021:G:T   G   T\n",
       "\n",
       "[10857 rows x 5 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_Biallelic_snp(snplist_df=pvar,pos_col=\"POS\", id_col=\"ID\", chr_col=\"#CHROM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos\n",
       "15471702    1\n",
       "15471822    1\n",
       "15471938    1\n",
       "15472580    1\n",
       "15473003    1\n",
       "           ..\n",
       "15905314    1\n",
       "15906139    1\n",
       "15895123    1\n",
       "15893979    1\n",
       "15908595    1\n",
       "Name: count, Length: 1726, dtype: int64"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame( [i.split(\":\") for i in snplist], columns= [\"chr\", \"pos\", \"ref\", \"alt\"]).drop_duplicates()['pos'].value_counts().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 提取该locus全部的基因"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/xutingfeng/miniforge3/lib/python3.10/site-packages/pyranges/methods/init.py:45: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return {k: v for k, v in df.groupby(grpby_key)}\n",
      "/home/xutingfeng/miniforge3/lib/python3.10/site-packages/pyranges/methods/init.py:160: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  empty_removed = df.groupby([\"Chromosome\", \"Strand\"])\n"
     ]
    }
   ],
   "source": [
    "# step1 extract locus gene list \n",
    "# extract gene annotation from gff file \n",
    "ensemble_anno_hs = pr.read_gff3(ensemble_anno_hs_path) # load gff3\n",
    "ensemble_anno_hs_gene = ensemble_anno_hs[ensemble_anno_hs.Feature == 'gene'] # focus on gene \n",
    "genes_in_locus_df = ensemble_anno_hs_gene[str(locus_range_tuple[0]), locus_range_tuple[1]:locus_range_tuple[2]].as_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 根据转录本的情况更新locus的范围\n",
    "\n",
    "这步是为了locus两端的基因可以考虑到cis的全部"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Considering the boundary of locus with cis range 100kb with the old locus range: (1, 15339446, 15817866)\n",
      "new locus range: (1, 15336520, 15909348)\n"
     ]
    }
   ],
   "source": [
    "# step2 update new locus range with considering the boundary genes with cis-range \n",
    "print(f\"Considering the boundary of locus with cis range {cis_range//1000}kb with the old locus range: {locus_range_tuple}\")\n",
    "locus_range_tuple_update = (locus_range_tuple[0], genes_in_locus_df['Start'].min() + cis_range, genes_in_locus_df['End'].max() + cis_range)\n",
    "print(f\"new locus range: {locus_range_tuple_update}\")\n",
    "# locus_range_tuple_update\n",
    "locus_range_tuple = locus_range_tuple_update"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 读取 eQTL数据(GTEx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "have TSTAT na num is : 5217\n",
      "after dropna: 101141\n",
      "the genes (from gene annotation file) in the locus: 15\n",
      "matched: (14,)\n",
      "not_in_eQTL_gene: {'RSC1A1'}\n",
      "gene_in_analysis: ['CELA2B', 'TMEM82', 'CELA2A', 'CTRC', 'UQCRHL', 'DNAJC16', 'CASP9', 'FBLIM1', 'DDI2', 'SLC25A34', 'AGMAT', 'PLEKHM2', 'EFHD2', 'FHAD1']\n",
      "AGMAT have cis-snp num: 719 while total snp num: 2058\n",
      "CASP9 have cis-snp num: 808 while total snp num: 2058\n",
      "CELA2A have cis-snp num: 731 while total snp num: 2058\n",
      "CELA2B have cis-snp num: 798 while total snp num: 2058\n",
      "CTRC have cis-snp num: 764 while total snp num: 2058\n",
      "DDI2 have cis-snp num: 770 while total snp num: 2058\n",
      "DNAJC16 have cis-snp num: 731 while total snp num: 2058\n",
      "EFHD2 have cis-snp num: 653 while total snp num: 2058\n",
      "FBLIM1 have cis-snp num: 704 while total snp num: 2058\n",
      "FHAD1 have cis-snp num: 398 while total snp num: 2058\n",
      "PLEKHM2 have cis-snp num: 769 while total snp num: 2058\n",
      "SLC25A34 have cis-snp num: 711 while total snp num: 2058\n",
      "TMEM82 have cis-snp num: 705 while total snp num: 2058\n",
      "UQCRHL have cis-snp num: 669 while total snp num: 2058\n",
      "total cis-snp num: 2058\n",
      "keep only it is cis-snp of any gene\n",
      "before filter, eQTL_df shape: (28812, 21)\n",
      "after filter, eQTL_df shape: (28812, 21)\n",
      "after filter, eQTL_df gene num: (14,)\n",
      "after filter, eQTL_df snp num: (2058,)\n",
      "missing rate >0: 0\n",
      "eQTL_df: (28812, 21)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_968/3075796394.py:73: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  missing_SNP_rate = eQTL_df.groupby(\"ID_sorted\").apply(lambda x: x.shape[0]/gene_nums).sort_values(ascending=False)\n"
     ]
    }
   ],
   "source": [
    "# set the position col var \n",
    "\n",
    "default_eQTL_necessary_col = {\n",
    "    'chr': 'chromosome',\n",
    "    'pos': 'position',\n",
    "    'ref': 'ref',\n",
    "    'alt': 'alt',\n",
    "    'beta': 'beta',\n",
    "    'se': 'se',\n",
    "    'gene_id': 'gene_id'\n",
    "}\n",
    "eQTL_chr_col = default_eQTL_necessary_col['chr']\n",
    "eQTL_postion_col = default_eQTL_necessary_col['pos']\n",
    "eQTL_ref_col = default_eQTL_necessary_col['ref']\n",
    "eQTL_alt_col = default_eQTL_necessary_col['alt']\n",
    "\n",
    "eQTL_beta_col = default_eQTL_necessary_col['beta']\n",
    "eQTL_se_col = default_eQTL_necessary_col['se']\n",
    "\n",
    "eQTL_gene_id = default_eQTL_necessary_col['gene_id']\n",
    "\n",
    "\n",
    "chr = locus_range_tuple[0]\n",
    "start = locus_range_tuple[1]\n",
    "end = locus_range_tuple[2]\n",
    "region = f\"{chr}:{start}-{end}\"\n",
    "\n",
    "eQTL_df = GTEx_tabix_reader(eQTLfile_path, region)\n",
    "\n",
    "# set ID_sorted for intersection at later \n",
    "eQTL_df['ID_sorted'] = add_ID(eQTL_df, [eQTL_chr_col, eQTL_postion_col, eQTL_ref_col, eQTL_alt_col], sort=True, inplace=False)\n",
    "\n",
    "# cal TSTAT for zscore matrix and drop na with TSTAT\n",
    "eQTL_df['TSTAT'] = eQTL_df[eQTL_beta_col].astype(float) / eQTL_df[eQTL_se_col].astype(float)\n",
    "print(\"have TSTAT na num is :\",eQTL_df['TSTAT'].isna().sum())\n",
    "eQTL_df = eQTL_df.dropna(subset=['TSTAT'])\n",
    "print(f\"after dropna: {eQTL_df.shape[0]}\")\n",
    "\n",
    "\n",
    "### eQTL和 transcript 取交集\n",
    "# 1. eQTLs的基因必须在transcirpt list \n",
    "# 2. eQTLs必须是cis-SNP\n",
    "# 3. 这个cis-SNP 必须所有gene都有对应的TSTAT\n",
    "\n",
    "# 1. check gene in transcript gene list \n",
    "print(f\"the genes (from gene annotation file) in the locus: {genes_in_locus_df.shape[0]}\")\n",
    "\n",
    "gene_in_locus_df_id_Name = genes_in_locus_df[['gene_id', 'Name']]\n",
    "\n",
    "matched = eQTL_df[eQTL_df['gene_id'].isin(gene_in_locus_df_id_Name['gene_id'])].merge(gene_in_locus_df_id_Name, on=\"gene_id\")\n",
    "\n",
    "print(f\"matched: {matched['gene_id'].unique().shape}\")\n",
    "\n",
    "not_in_eQTL_gene_name = set(gene_in_locus_df_id_Name['Name'].tolist()) - set(matched['Name'].unique().tolist())\n",
    "\n",
    "print(f\"not_in_eQTL_gene: {not_in_eQTL_gene_name}\")\n",
    "gene_in_analysis = matched['Name'].unique().tolist()\n",
    "print(f\"gene_in_analysis: {gene_in_analysis}\")\n",
    "eQTL_df = matched\n",
    "\n",
    "genes_in_locus_df = genes_in_locus_df[genes_in_locus_df['Name'].isin(gene_in_analysis)] # update genes_in_locus_df\n",
    "\n",
    "# 2. check cis-snp \n",
    "\n",
    "cis_snp_list = []\n",
    "\n",
    "for geneName , gene_eQTLs in eQTL_df.groupby('Name'): \n",
    "    gene_region = genes_in_locus_df[genes_in_locus_df['Name'] == geneName].iloc[0][['Start', \"End\"]].tolist()\n",
    "    upstream = (gene_region[0] - cis_range, gene_region[0])\n",
    "    downstream = (gene_region[1], gene_region[1] + cis_range)\n",
    "    gene_cis_range = [upstream, downstream]\n",
    "\n",
    "    gene_cis_eQTLs_df = extract_cis_snps(gene_eQTLs, eQTL_postion_col, gene_cis_range)\n",
    "\n",
    "    cis_snp_list = cis_snp_list + gene_cis_eQTLs_df['ID_sorted'].tolist()\n",
    "    print(f\"{geneName} have cis-snp num: {gene_cis_eQTLs_df.shape[0]} while total snp num: {gene_eQTLs.shape[0]}\")\n",
    "\n",
    "\n",
    "cis_snp_list = set(cis_snp_list)\n",
    "print(f\"total cis-snp num: {len(cis_snp_list)}\")\n",
    "\n",
    "print(\"keep only it is cis-snp of any gene\")\n",
    "print(f\"before filter, eQTL_df shape: {eQTL_df.shape}\")\n",
    "eQTL_df = eQTL_df[eQTL_df['ID_sorted'].isin(cis_snp_list)]\n",
    "print(f\"after filter, eQTL_df shape: {eQTL_df.shape}\")\n",
    "print(f\"after filter, eQTL_df gene num: {eQTL_df['Name'].unique().shape}\")\n",
    "print(f\"after filter, eQTL_df snp num: {eQTL_df['ID_sorted'].unique().shape}\")\n",
    "# 3. 去除缺失率过高的SNP\n",
    "gene_nums = len(gene_in_analysis)\n",
    "missing_SNP_rate = eQTL_df.groupby(\"ID_sorted\").apply(lambda x: x.shape[0]/gene_nums).sort_values(ascending=False)\n",
    "# missing_SNP_rate \n",
    "missing_SNP = missing_SNP_rate[missing_SNP_rate == 0].index.tolist()\n",
    "print(f\"missing rate >0: {len(missing_SNP)}\")\n",
    "\n",
    "eQTL_df = eQTL_df[~eQTL_df['ID_sorted'].isin(missing_SNP)]\n",
    "\n",
    "print(f\"eQTL_df: {eQTL_df.shape}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 读取GWAS数据 (GWASFormatted file)\n",
    "<!-- regenie格式 -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading region: 1:15336520-15909348\n",
      "tabix have a header, so will take the first line as header and remove it.\n"
     ]
    }
   ],
   "source": [
    "from finemap_tools.reader.gwas import load_GWASFormated_file\n",
    "\n",
    "default_gwas_necessary_col = {\n",
    "    'chr': 'chrom',\n",
    "    'pos': 'pos',\n",
    "    'ref': 'ref',\n",
    "    'alt': 'alt',\n",
    "    'beta': 'beta',\n",
    "    'se': 'sebeta'\n",
    "}\n",
    "gwas_chr_col = default_gwas_necessary_col['chr']\n",
    "gwas_pos_col = default_gwas_necessary_col['pos']\n",
    "gwas_ref_col = default_gwas_necessary_col['ref']\n",
    "gwas_alt_col = default_gwas_necessary_col['alt']\n",
    "\n",
    "gwas_beta_col = default_gwas_necessary_col['beta']\n",
    "gwas_se_col = default_gwas_necessary_col['se']\n",
    "\n",
    "\n",
    "print(f\"loading region: {region}\")\n",
    "GWAS_df = load_GWASFormated_file(file_path=gwasPath, region=region)\n",
    "\n",
    "GWAS_df['ID_sorted'] = add_ID(GWAS_df, [gwas_chr_col, gwas_pos_col, gwas_ref_col, gwas_alt_col], inplace=False, sort = True)\n",
    "GWAS_df[\"TSTAT\"] = GWAS_df[gwas_beta_col].astype(float) / GWAS_df[gwas_se_col].astype(float)\n",
    "GWAS_df = GWAS_df.dropna(subset=[\"TSTAT\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check pfile for LD "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14587143</td>\n",
       "      <td>1:14587143:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14587157</td>\n",
       "      <td>1:14587157:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14587159</td>\n",
       "      <td>1:14587159:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14587190</td>\n",
       "      <td>1:14587190:A:C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14587192</td>\n",
       "      <td>1:14587192:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109352</th>\n",
       "      <td>1</td>\n",
       "      <td>16587077</td>\n",
       "      <td>1:16587077:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109353</th>\n",
       "      <td>1</td>\n",
       "      <td>16587078</td>\n",
       "      <td>1:16587078:G:T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109354</th>\n",
       "      <td>1</td>\n",
       "      <td>16587079</td>\n",
       "      <td>1:16587079:G:T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109355</th>\n",
       "      <td>1</td>\n",
       "      <td>16587082</td>\n",
       "      <td>1:16587082:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109356</th>\n",
       "      <td>1</td>\n",
       "      <td>16587084</td>\n",
       "      <td>1:16587084:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>109357 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM       POS              ID REF ALT\n",
       "0            1  14587143  1:14587143:C:T   T   C\n",
       "1            1  14587157  1:14587157:C:T   C   T\n",
       "2            1  14587159  1:14587159:A:G   G   A\n",
       "3            1  14587190  1:14587190:A:C   C   A\n",
       "4            1  14587192  1:14587192:C:T   T   C\n",
       "...        ...       ...             ...  ..  ..\n",
       "109352       1  16587077  1:16587077:A:G   A   G\n",
       "109353       1  16587078  1:16587078:G:T   T   G\n",
       "109354       1  16587079  1:16587079:G:T   T   G\n",
       "109355       1  16587082  1:16587082:C:T   T   C\n",
       "109356       1  16587084  1:16587084:A:G   A   G\n",
       "\n",
       "[109357 rows x 5 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from finemap_tools.reader.plink import read_pvar\n",
    "\n",
    "pvar_df = read_pvar(pvar_path)\n",
    "\n",
    "pvar_df['ID_sorted'] = sort_ID(pvar_df, \"ID\", inplace=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 所有纳入的SNP进行QC\n",
    "\n",
    "\n",
    "1. 交集\n",
    "2. MAF > 1e-2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- 这里PGEN是UKB的数据，因此MAF以UKB的数据为准进行计算\n",
    "\n",
    "- 也可以gwas的数据为准进行MAF过滤，不可以eQTL（因为summary数据大多样本量较小）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pvar_SNP: 109357\n",
      "GWAS_SNP: 4458\n",
      "eQTL_SNP: 28812\n",
      "intersection: 2027\n"
     ]
    }
   ],
   "source": [
    "pvar_SNP = pvar_df['ID_sorted'].tolist()\n",
    "GWAS_SNP = GWAS_df['ID_sorted'].tolist()\n",
    "eQTL_SNP = eQTL_df['ID_sorted'].tolist()\n",
    "\n",
    "\n",
    "print(f\"pvar_SNP: {len(pvar_SNP)}\")\n",
    "print(f\"GWAS_SNP: {len(GWAS_SNP)}\")\n",
    "print(f\"eQTL_SNP: {len(eQTL_SNP)}\")\n",
    "\n",
    "\n",
    "pvar_SNP_set = set(pvar_SNP)\n",
    "GWAS_SNP_set = set(GWAS_SNP)\n",
    "eQTL_SNP_set = set(eQTL_SNP)\n",
    "\n",
    "\n",
    "intersection = pvar_SNP_set & GWAS_SNP_set & eQTL_SNP_set\n",
    "print(f\"intersection: {len(intersection)}\")\n",
    "\n",
    "pvar_df_intersction = pvar_df[pvar_df['ID_sorted'].isin(intersection)]\n",
    "# add real ID \n",
    "GWAS_df_intersction = GWAS_df[GWAS_df['ID_sorted'].isin(intersection)].merge(pvar_df_intersction[['ID_sorted', \"ID\"]], on=\"ID_sorted\")\n",
    "# add real ID \n",
    "eQTL_df_intersction = eQTL_df[eQTL_df['ID_sorted'].isin(intersection)].merge(pvar_df_intersction[['ID_sorted', \"ID\"]], on=\"ID_sorted\")\n",
    "\n",
    "\n",
    "SNP_ID_real = pvar_df_intersction[\"ID\"].tolist()\n",
    "# SNP_ID_real\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "from finemap_tools.plink import plink2_cal_freq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "created temporary directory /tmp/tmp80w217vi\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      #CHROM                 ID REF   ALT PROVISIONAL_REF?  ALT_FREQS  OBS_CT\n",
      "0          1     1:15336612:A:C   C     A                Y   0.013270  911472\n",
      "1          1     1:15336888:A:G   A     G                Y   0.072879  932296\n",
      "2          1     1:15336935:C:T   T     C                Y   0.306207  921016\n",
      "3          1     1:15337191:A:G   G     A                Y   0.013222  931276\n",
      "4          1     1:15337448:C:T   C     T                Y   0.065020  930034\n",
      "...      ...                ...  ..   ...              ...        ...     ...\n",
      "2022       1     1:15905314:C:G   C     G                Y   0.085668  932098\n",
      "2023       1     1:15906139:C:T   C     T                Y   0.071549  932296\n",
      "2024       1     1:15906557:C:T   C     T                Y   0.002449  932132\n",
      "2025       1  1:15908170:G:GCTA   G  GCTA                Y   0.002481  932166\n",
      "2026       1     1:15908595:A:G   A     G                Y   0.017554  929088\n",
      "\n",
      "[2027 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "# 创建临时文件夹来存储数据\n",
    "with tempfile.TemporaryDirectory() as tmpdirname:\n",
    "    print('created temporary directory', tmpdirname)\n",
    "    freq = plink2_cal_freq(pgen=\"data/pgen/DNAJC16_ukb\", snplist=SNP_ID_real, outSuffix=tmpdirname + '/plink2_cal_freq')\n",
    "    print(freq)\n",
    "    \n",
    "# freq = plink2_cal_freq(pgen=\"data/pgen/DNAJC16_ukb\", snplist=SNP_ID_real, outSuffix='test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eQTL_df_passed_qc: (24164, 22)\n",
      "GWAS_df_passed_qc: (1726, 11)\n"
     ]
    }
   ],
   "source": [
    "final_SNP_list = freq[freq['ALT_FREQS'] >1e-2]['ID'].tolist() \n",
    "\n",
    "eQTL_df = eQTL_df_intersction[eQTL_df_intersction['ID'].isin(final_SNP_list)]\n",
    "GWAS_df_passed_qc = GWAS_df_intersction[GWAS_df_intersction['ID'].isin(final_SNP_list)]\n",
    "print(f\"eQTL_df_passed_qc: {eQTL_df.shape}\")\n",
    "print(f\"GWAS_df_passed_qc: {GWAS_df_passed_qc.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 基因表达量相关性计算\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    \n",
    "expression_filepath = \"data/expression/gene_reads_2017-06-05_v8_kidney_cortex.gct.gz\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Description</th>\n",
       "      <th>CELA2B</th>\n",
       "      <th>TMEM82</th>\n",
       "      <th>CELA2A</th>\n",
       "      <th>CTRC</th>\n",
       "      <th>UQCRHL</th>\n",
       "      <th>DNAJC16</th>\n",
       "      <th>CASP9</th>\n",
       "      <th>FBLIM1</th>\n",
       "      <th>DDI2</th>\n",
       "      <th>SLC25A34</th>\n",
       "      <th>AGMAT</th>\n",
       "      <th>PLEKHM2</th>\n",
       "      <th>EFHD2</th>\n",
       "      <th>FHAD1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Description</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CELA2B</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.184775</td>\n",
       "      <td>0.942457</td>\n",
       "      <td>0.979194</td>\n",
       "      <td>-0.083674</td>\n",
       "      <td>-0.078751</td>\n",
       "      <td>0.165051</td>\n",
       "      <td>0.208723</td>\n",
       "      <td>0.038826</td>\n",
       "      <td>-0.048049</td>\n",
       "      <td>-0.111831</td>\n",
       "      <td>0.071931</td>\n",
       "      <td>0.062023</td>\n",
       "      <td>0.249947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM82</th>\n",
       "      <td>-0.184775</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.202855</td>\n",
       "      <td>-0.180160</td>\n",
       "      <td>0.363203</td>\n",
       "      <td>0.553856</td>\n",
       "      <td>0.136347</td>\n",
       "      <td>-0.095800</td>\n",
       "      <td>0.418527</td>\n",
       "      <td>0.657048</td>\n",
       "      <td>0.377137</td>\n",
       "      <td>0.087738</td>\n",
       "      <td>0.050699</td>\n",
       "      <td>0.229548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CELA2A</th>\n",
       "      <td>0.942457</td>\n",
       "      <td>-0.202855</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.947389</td>\n",
       "      <td>-0.069936</td>\n",
       "      <td>-0.090878</td>\n",
       "      <td>0.265227</td>\n",
       "      <td>0.135480</td>\n",
       "      <td>0.010767</td>\n",
       "      <td>-0.033191</td>\n",
       "      <td>-0.120507</td>\n",
       "      <td>0.031208</td>\n",
       "      <td>0.012756</td>\n",
       "      <td>0.200312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTRC</th>\n",
       "      <td>0.979194</td>\n",
       "      <td>-0.180160</td>\n",
       "      <td>0.947389</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.058232</td>\n",
       "      <td>-0.053476</td>\n",
       "      <td>0.200043</td>\n",
       "      <td>0.199330</td>\n",
       "      <td>0.062109</td>\n",
       "      <td>-0.024640</td>\n",
       "      <td>-0.088197</td>\n",
       "      <td>0.071908</td>\n",
       "      <td>0.067095</td>\n",
       "      <td>0.243770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UQCRHL</th>\n",
       "      <td>-0.083674</td>\n",
       "      <td>0.363203</td>\n",
       "      <td>-0.069936</td>\n",
       "      <td>-0.058232</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.583717</td>\n",
       "      <td>0.579547</td>\n",
       "      <td>0.042745</td>\n",
       "      <td>0.460671</td>\n",
       "      <td>0.389708</td>\n",
       "      <td>0.438251</td>\n",
       "      <td>0.359547</td>\n",
       "      <td>0.239764</td>\n",
       "      <td>0.227805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNAJC16</th>\n",
       "      <td>-0.078751</td>\n",
       "      <td>0.553856</td>\n",
       "      <td>-0.090878</td>\n",
       "      <td>-0.053476</td>\n",
       "      <td>0.583717</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.391077</td>\n",
       "      <td>-0.060071</td>\n",
       "      <td>0.475638</td>\n",
       "      <td>0.478915</td>\n",
       "      <td>0.891022</td>\n",
       "      <td>0.094745</td>\n",
       "      <td>0.094650</td>\n",
       "      <td>0.293590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CASP9</th>\n",
       "      <td>0.165051</td>\n",
       "      <td>0.136347</td>\n",
       "      <td>0.265227</td>\n",
       "      <td>0.200043</td>\n",
       "      <td>0.579547</td>\n",
       "      <td>0.391077</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.176401</td>\n",
       "      <td>0.582608</td>\n",
       "      <td>0.338586</td>\n",
       "      <td>0.128561</td>\n",
       "      <td>0.491348</td>\n",
       "      <td>0.284614</td>\n",
       "      <td>0.460086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FBLIM1</th>\n",
       "      <td>0.208723</td>\n",
       "      <td>-0.095800</td>\n",
       "      <td>0.135480</td>\n",
       "      <td>0.199330</td>\n",
       "      <td>0.042745</td>\n",
       "      <td>-0.060071</td>\n",
       "      <td>0.176401</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.517010</td>\n",
       "      <td>0.066621</td>\n",
       "      <td>-0.294383</td>\n",
       "      <td>0.766109</td>\n",
       "      <td>0.740864</td>\n",
       "      <td>0.356500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DDI2</th>\n",
       "      <td>0.038826</td>\n",
       "      <td>0.418527</td>\n",
       "      <td>0.010767</td>\n",
       "      <td>0.062109</td>\n",
       "      <td>0.460671</td>\n",
       "      <td>0.475638</td>\n",
       "      <td>0.582608</td>\n",
       "      <td>0.517010</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.139019</td>\n",
       "      <td>0.730252</td>\n",
       "      <td>0.586365</td>\n",
       "      <td>0.541152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC25A34</th>\n",
       "      <td>-0.048049</td>\n",
       "      <td>0.657048</td>\n",
       "      <td>-0.033191</td>\n",
       "      <td>-0.024640</td>\n",
       "      <td>0.389708</td>\n",
       "      <td>0.478915</td>\n",
       "      <td>0.338586</td>\n",
       "      <td>0.066621</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.344579</td>\n",
       "      <td>0.211022</td>\n",
       "      <td>0.063415</td>\n",
       "      <td>0.248887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGMAT</th>\n",
       "      <td>-0.111831</td>\n",
       "      <td>0.377137</td>\n",
       "      <td>-0.120507</td>\n",
       "      <td>-0.088197</td>\n",
       "      <td>0.438251</td>\n",
       "      <td>0.891022</td>\n",
       "      <td>0.128561</td>\n",
       "      <td>-0.294383</td>\n",
       "      <td>0.139019</td>\n",
       "      <td>0.344579</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.209865</td>\n",
       "      <td>-0.148470</td>\n",
       "      <td>0.061704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLEKHM2</th>\n",
       "      <td>0.071931</td>\n",
       "      <td>0.087738</td>\n",
       "      <td>0.031208</td>\n",
       "      <td>0.071908</td>\n",
       "      <td>0.359547</td>\n",
       "      <td>0.094745</td>\n",
       "      <td>0.491348</td>\n",
       "      <td>0.766109</td>\n",
       "      <td>0.730252</td>\n",
       "      <td>0.211022</td>\n",
       "      <td>-0.209865</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.803527</td>\n",
       "      <td>0.455571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EFHD2</th>\n",
       "      <td>0.062023</td>\n",
       "      <td>0.050699</td>\n",
       "      <td>0.012756</td>\n",
       "      <td>0.067095</td>\n",
       "      <td>0.239764</td>\n",
       "      <td>0.094650</td>\n",
       "      <td>0.284614</td>\n",
       "      <td>0.740864</td>\n",
       "      <td>0.586365</td>\n",
       "      <td>0.063415</td>\n",
       "      <td>-0.148470</td>\n",
       "      <td>0.803527</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.392527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FHAD1</th>\n",
       "      <td>0.249947</td>\n",
       "      <td>0.229548</td>\n",
       "      <td>0.200312</td>\n",
       "      <td>0.243770</td>\n",
       "      <td>0.227805</td>\n",
       "      <td>0.293590</td>\n",
       "      <td>0.460086</td>\n",
       "      <td>0.356500</td>\n",
       "      <td>0.541152</td>\n",
       "      <td>0.248887</td>\n",
       "      <td>0.061704</td>\n",
       "      <td>0.455571</td>\n",
       "      <td>0.392527</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Description    CELA2B    TMEM82    CELA2A      CTRC    UQCRHL   DNAJC16  \\\n",
       "Description                                                               \n",
       "CELA2B       1.000000 -0.184775  0.942457  0.979194 -0.083674 -0.078751   \n",
       "TMEM82      -0.184775  1.000000 -0.202855 -0.180160  0.363203  0.553856   \n",
       "CELA2A       0.942457 -0.202855  1.000000  0.947389 -0.069936 -0.090878   \n",
       "CTRC         0.979194 -0.180160  0.947389  1.000000 -0.058232 -0.053476   \n",
       "UQCRHL      -0.083674  0.363203 -0.069936 -0.058232  1.000000  0.583717   \n",
       "DNAJC16     -0.078751  0.553856 -0.090878 -0.053476  0.583717  1.000000   \n",
       "CASP9        0.165051  0.136347  0.265227  0.200043  0.579547  0.391077   \n",
       "FBLIM1       0.208723 -0.095800  0.135480  0.199330  0.042745 -0.060071   \n",
       "DDI2         0.038826  0.418527  0.010767  0.062109  0.460671  0.475638   \n",
       "SLC25A34    -0.048049  0.657048 -0.033191 -0.024640  0.389708  0.478915   \n",
       "AGMAT       -0.111831  0.377137 -0.120507 -0.088197  0.438251  0.891022   \n",
       "PLEKHM2      0.071931  0.087738  0.031208  0.071908  0.359547  0.094745   \n",
       "EFHD2        0.062023  0.050699  0.012756  0.067095  0.239764  0.094650   \n",
       "FHAD1        0.249947  0.229548  0.200312  0.243770  0.227805  0.293590   \n",
       "\n",
       "Description     CASP9    FBLIM1      DDI2  SLC25A34     AGMAT   PLEKHM2  \\\n",
       "Description                                                               \n",
       "CELA2B       0.165051  0.208723  0.038826 -0.048049 -0.111831  0.071931   \n",
       "TMEM82       0.136347 -0.095800  0.418527  0.657048  0.377137  0.087738   \n",
       "CELA2A       0.265227  0.135480  0.010767 -0.033191 -0.120507  0.031208   \n",
       "CTRC         0.200043  0.199330  0.062109 -0.024640 -0.088197  0.071908   \n",
       "UQCRHL       0.579547  0.042745  0.460671  0.389708  0.438251  0.359547   \n",
       "DNAJC16      0.391077 -0.060071  0.475638  0.478915  0.891022  0.094745   \n",
       "CASP9        1.000000  0.176401  0.582608  0.338586  0.128561  0.491348   \n",
       "FBLIM1       0.176401  1.000000  0.517010  0.066621 -0.294383  0.766109   \n",
       "DDI2         0.582608  0.517010  1.000000  0.454545  0.139019  0.730252   \n",
       "SLC25A34     0.338586  0.066621  0.454545  1.000000  0.344579  0.211022   \n",
       "AGMAT        0.128561 -0.294383  0.139019  0.344579  1.000000 -0.209865   \n",
       "PLEKHM2      0.491348  0.766109  0.730252  0.211022 -0.209865  1.000000   \n",
       "EFHD2        0.284614  0.740864  0.586365  0.063415 -0.148470  0.803527   \n",
       "FHAD1        0.460086  0.356500  0.541152  0.248887  0.061704  0.455571   \n",
       "\n",
       "Description     EFHD2     FHAD1  \n",
       "Description                      \n",
       "CELA2B       0.062023  0.249947  \n",
       "TMEM82       0.050699  0.229548  \n",
       "CELA2A       0.012756  0.200312  \n",
       "CTRC         0.067095  0.243770  \n",
       "UQCRHL       0.239764  0.227805  \n",
       "DNAJC16      0.094650  0.293590  \n",
       "CASP9        0.284614  0.460086  \n",
       "FBLIM1       0.740864  0.356500  \n",
       "DDI2         0.586365  0.541152  \n",
       "SLC25A34     0.063415  0.248887  \n",
       "AGMAT       -0.148470  0.061704  \n",
       "PLEKHM2      0.803527  0.455571  \n",
       "EFHD2        1.000000  0.392527  \n",
       "FHAD1        0.392527  1.000000  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneExpression_df = load_GTEx_gene_expression(expression_filepath)[gene_in_analysis]\n",
    "geneExpression_df_corr = geneExpression_df.corr()\n",
    "geneExpression_df_corr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LD 数据计算"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1:15336612:A:C</th>\n",
       "      <th>1:15336888:A:G</th>\n",
       "      <th>1:15336935:C:T</th>\n",
       "      <th>1:15337191:A:G</th>\n",
       "      <th>1:15337448:C:T</th>\n",
       "      <th>1:15337531:C:T</th>\n",
       "      <th>1:15337851:A:C</th>\n",
       "      <th>1:15337914:C:T</th>\n",
       "      <th>1:15339446:G:T</th>\n",
       "      <th>1:15340368:A:G</th>\n",
       "      <th>...</th>\n",
       "      <th>1:15903102:A:T</th>\n",
       "      <th>1:15903567:A:G</th>\n",
       "      <th>1:15903632:T:TC</th>\n",
       "      <th>1:15904162:C:T</th>\n",
       "      <th>1:15904535:C:T</th>\n",
       "      <th>1:15904647:A:C</th>\n",
       "      <th>1:15905251:C:T</th>\n",
       "      <th>1:15905314:C:G</th>\n",
       "      <th>1:15906139:C:T</th>\n",
       "      <th>1:15908595:A:G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1:15336612:A:C</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000999</td>\n",
       "      <td>0.006069</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000780</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005622</td>\n",
       "      <td>0.005153</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.004777</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.008599</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15336888:A:G</th>\n",
       "      <td>0.000999</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.036063</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.885309</td>\n",
       "      <td>0.002126</td>\n",
       "      <td>0.882178</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.742131</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.001731</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.001791</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000623</td>\n",
       "      <td>0.001050</td>\n",
       "      <td>0.001502</td>\n",
       "      <td>0.000186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15336935:C:T</th>\n",
       "      <td>0.006069</td>\n",
       "      <td>0.036063</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.006285</td>\n",
       "      <td>0.031516</td>\n",
       "      <td>0.026295</td>\n",
       "      <td>0.031567</td>\n",
       "      <td>0.007372</td>\n",
       "      <td>0.005547</td>\n",
       "      <td>0.026510</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.001505</td>\n",
       "      <td>0.000388</td>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.000615</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.003779</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.003086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15337191:A:G</th>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.006285</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.000368</td>\n",
       "      <td>0.000746</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.956575</td>\n",
       "      <td>0.000776</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15337448:C:T</th>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.885309</td>\n",
       "      <td>0.031516</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001775</td>\n",
       "      <td>0.997336</td>\n",
       "      <td>0.001231</td>\n",
       "      <td>0.000670</td>\n",
       "      <td>0.857628</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001038</td>\n",
       "      <td>0.001305</td>\n",
       "      <td>0.000171</td>\n",
       "      <td>0.001419</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.002889</td>\n",
       "      <td>0.000066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15904647:A:C</th>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>0.002198</td>\n",
       "      <td>0.466997</td>\n",
       "      <td>0.002287</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.013380</td>\n",
       "      <td>0.001743</td>\n",
       "      <td>0.001306</td>\n",
       "      <td>0.000305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15905251:C:T</th>\n",
       "      <td>0.000242</td>\n",
       "      <td>0.000623</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.000418</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.001184</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000526</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078008</td>\n",
       "      <td>0.084164</td>\n",
       "      <td>0.034112</td>\n",
       "      <td>0.089055</td>\n",
       "      <td>0.009350</td>\n",
       "      <td>0.013380</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.071025</td>\n",
       "      <td>0.052120</td>\n",
       "      <td>0.012263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15905314:C:G</th>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.001050</td>\n",
       "      <td>0.003779</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000372</td>\n",
       "      <td>0.002673</td>\n",
       "      <td>0.000354</td>\n",
       "      <td>0.000612</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010825</td>\n",
       "      <td>0.010410</td>\n",
       "      <td>0.004692</td>\n",
       "      <td>0.011098</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>0.001743</td>\n",
       "      <td>0.071025</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.006990</td>\n",
       "      <td>0.001791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15906139:C:T</th>\n",
       "      <td>0.008599</td>\n",
       "      <td>0.001502</td>\n",
       "      <td>0.000306</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.002889</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.002890</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.003745</td>\n",
       "      <td>...</td>\n",
       "      <td>0.664067</td>\n",
       "      <td>0.639381</td>\n",
       "      <td>0.003635</td>\n",
       "      <td>0.605338</td>\n",
       "      <td>0.000914</td>\n",
       "      <td>0.001306</td>\n",
       "      <td>0.052120</td>\n",
       "      <td>0.006990</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.001348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1:15908595:A:G</th>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000186</td>\n",
       "      <td>0.003086</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001934</td>\n",
       "      <td>0.002103</td>\n",
       "      <td>0.000802</td>\n",
       "      <td>0.002214</td>\n",
       "      <td>0.000247</td>\n",
       "      <td>0.000305</td>\n",
       "      <td>0.012263</td>\n",
       "      <td>0.001791</td>\n",
       "      <td>0.001348</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1726 rows × 1726 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                1:15336612:A:C  1:15336888:A:G  1:15336935:C:T  \\\n",
       "1:15336612:A:C        1.000000        0.000999        0.006069   \n",
       "1:15336888:A:G        0.000999        1.000000        0.036063   \n",
       "1:15336935:C:T        0.006069        0.036063        1.000000   \n",
       "1:15337191:A:G        0.000215        0.000217        0.006285   \n",
       "1:15337448:C:T        0.000782        0.885309        0.031516   \n",
       "...                        ...             ...             ...   \n",
       "1:15904647:A:C        0.000065        0.000009        0.000028   \n",
       "1:15905251:C:T        0.000242        0.000623        0.000916   \n",
       "1:15905314:C:G        0.000001        0.001050        0.003779   \n",
       "1:15906139:C:T        0.008599        0.001502        0.000306   \n",
       "1:15908595:A:G        0.000021        0.000186        0.003086   \n",
       "\n",
       "                1:15337191:A:G  1:15337448:C:T  1:15337531:C:T  \\\n",
       "1:15336612:A:C        0.000215        0.000782        0.000269   \n",
       "1:15336888:A:G        0.000217        0.885309        0.002126   \n",
       "1:15336935:C:T        0.006285        0.031516        0.026295   \n",
       "1:15337191:A:G        1.000000        0.000729        0.000368   \n",
       "1:15337448:C:T        0.000729        1.000000        0.001775   \n",
       "...                        ...             ...             ...   \n",
       "1:15904647:A:C        0.000117        0.000009        0.000006   \n",
       "1:15905251:C:T        0.000418        0.000008        0.001184   \n",
       "1:15905314:C:G        0.000042        0.000372        0.002673   \n",
       "1:15906139:C:T        0.000010        0.002889        0.000036   \n",
       "1:15908595:A:G        0.000036        0.000066        0.000300   \n",
       "\n",
       "                1:15337851:A:C  1:15337914:C:T  1:15339446:G:T  \\\n",
       "1:15336612:A:C        0.000780        0.000256        0.000180   \n",
       "1:15336888:A:G        0.882178        0.001493        0.000233   \n",
       "1:15336935:C:T        0.031567        0.007372        0.005547   \n",
       "1:15337191:A:G        0.000746        0.000225        0.956575   \n",
       "1:15337448:C:T        0.997336        0.001231        0.000670   \n",
       "...                        ...             ...             ...   \n",
       "1:15904647:A:C        0.000010        0.000217        0.000107   \n",
       "1:15905251:C:T        0.000019        0.000044        0.000526   \n",
       "1:15905314:C:G        0.000354        0.000612        0.000087   \n",
       "1:15906139:C:T        0.002890        0.000530        0.000003   \n",
       "1:15908595:A:G        0.000068        0.000059        0.000044   \n",
       "\n",
       "                1:15340368:A:G  ...  1:15903102:A:T  1:15903567:A:G  \\\n",
       "1:15336612:A:C        0.000625  ...        0.005622        0.005153   \n",
       "1:15336888:A:G        0.742131  ...        0.000334        0.001731   \n",
       "1:15336935:C:T        0.026510  ...        0.002184        0.001505   \n",
       "1:15337191:A:G        0.000776  ...        0.000008        0.000008   \n",
       "1:15337448:C:T        0.857628  ...        0.001038        0.001305   \n",
       "...                        ...  ...             ...             ...   \n",
       "1:15904647:A:C        0.000020  ...        0.002028        0.002198   \n",
       "1:15905251:C:T        0.000074  ...        0.078008        0.084164   \n",
       "1:15905314:C:G        0.000334  ...        0.010825        0.010410   \n",
       "1:15906139:C:T        0.003745  ...        0.664067        0.639381   \n",
       "1:15908595:A:G        0.000021  ...        0.001934        0.002103   \n",
       "\n",
       "                1:15903632:T:TC  1:15904162:C:T  1:15904535:C:T  \\\n",
       "1:15336612:A:C         0.000126        0.004777        0.000066   \n",
       "1:15336888:A:G         0.000460        0.001791        0.000046   \n",
       "1:15336935:C:T         0.000388        0.001194        0.000615   \n",
       "1:15337191:A:G         0.000007        0.000024        0.000069   \n",
       "1:15337448:C:T         0.000171        0.001419        0.000005   \n",
       "...                         ...             ...             ...   \n",
       "1:15904647:A:C         0.466997        0.002287        0.000225   \n",
       "1:15905251:C:T         0.034112        0.089055        0.009350   \n",
       "1:15905314:C:G         0.004692        0.011098        0.001258   \n",
       "1:15906139:C:T         0.003635        0.605338        0.000914   \n",
       "1:15908595:A:G         0.000802        0.002214        0.000247   \n",
       "\n",
       "                1:15904647:A:C  1:15905251:C:T  1:15905314:C:G  \\\n",
       "1:15336612:A:C        0.000065        0.000242        0.000001   \n",
       "1:15336888:A:G        0.000009        0.000623        0.001050   \n",
       "1:15336935:C:T        0.000028        0.000916        0.003779   \n",
       "1:15337191:A:G        0.000117        0.000418        0.000042   \n",
       "1:15337448:C:T        0.000009        0.000008        0.000372   \n",
       "...                        ...             ...             ...   \n",
       "1:15904647:A:C        1.000000        0.013380        0.001743   \n",
       "1:15905251:C:T        0.013380        1.000000        0.071025   \n",
       "1:15905314:C:G        0.001743        0.071025        1.000000   \n",
       "1:15906139:C:T        0.001306        0.052120        0.006990   \n",
       "1:15908595:A:G        0.000305        0.012263        0.001791   \n",
       "\n",
       "                1:15906139:C:T  1:15908595:A:G  \n",
       "1:15336612:A:C        0.008599        0.000021  \n",
       "1:15336888:A:G        0.001502        0.000186  \n",
       "1:15336935:C:T        0.000306        0.003086  \n",
       "1:15337191:A:G        0.000010        0.000036  \n",
       "1:15337448:C:T        0.002889        0.000066  \n",
       "...                        ...             ...  \n",
       "1:15904647:A:C        0.001306        0.000305  \n",
       "1:15905251:C:T        0.052120        0.012263  \n",
       "1:15905314:C:G        0.006990        0.001791  \n",
       "1:15906139:C:T        1.000000        0.001348  \n",
       "1:15908595:A:G        0.001348        1.000000  \n",
       "\n",
       "[1726 rows x 1726 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LD_df = plink2_cal_LD(pgen=\"data/pgen/DNAJC16_ukb\", snplist=final_SNP_list, outSuffix='test')\n",
    "LD_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- ## eQTLs 提取顺式区域\n",
    "\n",
    "这一步会生成：\n",
    "- snplist: (gene1_cis_snp, ...) \n",
    "- pindex: (gene1_cis_snp_num, gene2_cis_snp_num ...)\n",
    "- Zscore_eQTLs (snplist_nums, gene_nums) -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis range: 100kb\n",
      "total cis snps: 8430\n",
      "gene: AGMAT has 647 cis snps\n",
      "gene: CASP9 has 674 cis snps\n",
      "gene: CELA2A has 618 cis snps\n",
      "gene: CELA2B has 678 cis snps\n",
      "gene: CTRC has 639 cis snps\n",
      "gene: DDI2 has 656 cis snps\n",
      "gene: DNAJC16 has 639 cis snps\n",
      "gene: EFHD2 has 552 cis snps\n",
      "gene: FBLIM1 has 583 cis snps\n",
      "gene: FHAD1 has 324 cis snps\n",
      "gene: PLEKHM2 has 650 cis snps\n",
      "gene: SLC25A34 has 614 cis snps\n",
      "gene: TMEM82 has 600 cis snps\n",
      "gene: UQCRHL has 556 cis snps\n"
     ]
    }
   ],
   "source": [
    "eQTL_postion_col = \"position\"\n",
    "eQTL_ID_col = \"ID\"\n",
    "snplist = [] \n",
    "pindex = []\n",
    "gene = [] \n",
    "# gene_eQTLs_cis = {} \n",
    "\n",
    "print(f\"cis range: {cis_range//1000}kb\")\n",
    "\n",
    "\n",
    "\n",
    "for geneName, gene_eQTLs in eQTL_df.groupby(\"Name\"):\n",
    "    gene_region = genes_in_locus_df[genes_in_locus_df[\"Name\"] == geneName].iloc[0][['Start', \"End\"]].tolist()\n",
    "    upstream = (gene_region[0] - cis_range, gene_region[0])\n",
    "    downstream = (gene_region[1], gene_region[1] + cis_range)\n",
    "    gene_cis_range = [upstream, downstream]\n",
    "\n",
    "    # filter eQTLs only keep the cis snps at cis range \n",
    "    gene_cis_eQTLs_df = extract_cis_snps(gene_eQTLs, eQTL_postion_col, gene_cis_range).set_index(eQTL_ID_col)[[\"TSTAT\"]]\n",
    "    if (na_num:=sum(gene_cis_eQTLs_df[['TSTAT']].isna().sum())) > 0:\n",
    "        print(f\"gene: {geneName} has {na_num} NA value in TSTAT\")\n",
    "        # gene_cis_eQTLs_df = gene_cis_eQTLs_df.dropna()\n",
    "        raise Exception(f\"gene: {geneName} has {na_num} NA value in TSTAT, plz check the data to ensure all cis-snp have TSTAT in eQTL data.\")\n",
    "        \n",
    "    if gene_cis_eQTLs_df.shape[0] > 0 : # if there is any snp in the cis range\n",
    "        # gene_eQTLs_cis[geneName] = gene_cis_eQTLs_df # TODO: append is not necessary \n",
    "\n",
    "        gene.append(geneName)\n",
    "        snplist += gene_cis_eQTLs_df.index.tolist() # add cis snps to snplist\n",
    "        pindex.append(gene_cis_eQTLs_df.shape[0]) # add the number of cis snps in the gene to pindex\n",
    "    else:\n",
    "        print(f\"gene: {geneName} has no snp in cis range\")\n",
    "\n",
    "\n",
    "\n",
    "print(f\"total cis snps: {len(snplist)}\")\n",
    "for snpnum, g in zip(pindex, gene):\n",
    "    print(f\"gene: {g} has {snpnum} cis snps\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 整合数据=> GIFT格式\n",
    "需要\n",
    "- eQTL\n",
    "- GWAS \n",
    "- LD \n",
    "- R  N by N  matrix, where N is gene nums\n",
    "- gene list (gene1, gene2 ...)\n",
    "\n",
    "\n",
    "\n",
    "重排LD、eQTL，GWAS，R以及genelist:\n",
    "\n",
    "- eQTL: row is snplist , columns is genes \n",
    "- GWAS: row is snp list \n",
    "- LD: row and column is snplist \n",
    "- R: N by N matrix, N is gene nums \n",
    "- gene list (gene1, gene2 ...)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8430 8430\n",
      "eQTLs matrix shape is :(8430, 14)\n"
     ]
    }
   ],
   "source": [
    "snplist_df = pd.DataFrame(snplist, columns=[\"ID\"])\n",
    "\n",
    "#Zscore_GWAS 计算\n",
    "Zscore_GWAS_df = snplist_df.merge(GWAS_df_passed_qc, how=\"left\", on = \"ID\")[[ \"ID\", \"TSTAT\"]].set_index(\"ID\")\n",
    "\n",
    "# Zscore_eQTLs 计算\n",
    "Zscore_eQTLs_list = []\n",
    "for geneName, gene_eQTLs in eQTL_df.groupby(\"Name\"):\n",
    "\n",
    "    Zscore_eQTLs = snplist_df.merge(gene_eQTLs[[ \"ID\", \"TSTAT\"]].rename(columns = {\"TSTAT\": geneName}), how=\"left\", on = \"ID\").set_index(\"ID\")\n",
    "    Zscore_eQTLs_list.append(Zscore_eQTLs)\n",
    "\n",
    "assert True == all((x.index == Zscore_eQTLs_list[0].index ).all() for x in Zscore_eQTLs_list)  # check if all index are the same\n",
    "Zscore_eQTLs_df = pd.concat(Zscore_eQTLs_list, axis = 1)\n",
    "assert Zscore_eQTLs_df.isna().sum().sum() == 0, \"eQTLs Zscore has NA value, plz check the data to ensure all cis-snp have TSTAT in eQTL data.\"\n",
    "\n",
    "# LD 重排顺序\n",
    "LD_eQTL = LD_df.loc[snplist, snplist]\n",
    "LD_GWAS = LD_df.loc[snplist, snplist]\n",
    "\n",
    "# expression corr \n",
    "expression_corr = geneExpression_df_corr.loc[gene, gene]\n",
    "\n",
    "# check all \n",
    "print(Zscore_GWAS_df.shape[0], sum(pindex))\n",
    "assert sum(pindex) == Zscore_GWAS_df.shape[0], print(f\"eQTLs snps are not equal to GWAS snps, eQTLs snps are {sum(pindex)} and GWAS snps are {Zscore_GWAS_df.shape[0]}\")\n",
    "assert [col for col in Zscore_eQTLs_df.columns] == gene\n",
    "assert Zscore_GWAS_df.index.tolist() == Zscore_eQTLs_df.index.tolist()\n",
    "assert Zscore_GWAS_df.index.tolist() == snplist\n",
    "assert Zscore_eQTLs_df.index.tolist() == LD_eQTL.index.tolist()\n",
    "assert Zscore_GWAS_df.index.tolist() == LD_GWAS.index.tolist()\n",
    "assert expression_corr.index.tolist() == gene\n",
    "\n",
    "print(f\"eQTLs matrix shape is :{Zscore_eQTLs_df.shape}\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to h5 \n",
    "import h5py\n",
    "\n",
    "with h5py.File(\"GIFT.h5\", \"w\") as f:\n",
    "    f.create_group(\"data\")\n",
    "    f['data'].create_dataset('gene', data=gene)\n",
    "    f['data'].create_dataset('snplist', data=snplist)\n",
    "    f['data'].create_dataset('pindex', data=pindex)\n",
    "    f['data'].create_dataset('Zscore_eQTLs_df', data=Zscore_eQTLs_df)\n",
    "    f['data'].create_dataset('Zscore_GWAS_df', data=Zscore_GWAS_df)\n",
    "    f['data'].create_dataset('geneExpression_corr_df', data=expression_corr)\n",
    "    f['data'].create_dataset('LD_eQTL', data=LD_eQTL)\n",
    "    f['data'].create_dataset('LD_GWAS', data=LD_GWAS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aaa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwasPath = \"/home/xutingfeng/GIFT/data/GWAS/DNAJC16_gift.tsv\" \n",
    "eQTLsLocation = \"/home/xutingfeng/GIFT/data/eQTL/Kidney_Cortex/genes\"\n",
    "plink_vcor2_LDPath = \"/home/xutingfeng/GIFT/data/pgen/DNAJC16_LD.unphased.vcor2\"\n",
    "\n",
    "geneExpressionPath = \"/home/xutingfeng/GIFT/data/expression/gene_tpm_2017-06-05_v8_kidney_cortex.gct.gz\"\n",
    "Ensemble_gff3 =  \"Homo_sapiens.GRCh38.111.chromosome.1.gff3.gz\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_in_region(value, region_list):\n",
    "    \"\"\"\n",
    "    Check if a value falls within any of the regions in the given region list.\n",
    "\n",
    "    Parameters:\n",
    "    value (float): The value to be checked.\n",
    "    region_list (list): A list of tuples representing the regions. Each tuple should contain two elements: the lower bound and the upper bound of the region.\n",
    "\n",
    "    Returns:\n",
    "    bool: True if the value falls within any of the regions, False otherwise.\n",
    "    \"\"\"\n",
    "    if isinstance(region_list[0], int):\n",
    "        region_list = [region_list]\n",
    "    for region in region_list:\n",
    "        if all((value >= region[0], value < region[1])):\n",
    "\n",
    "            return True \n",
    "\n",
    "    return False\n",
    "\n",
    "\n",
    "def extract_cis_snps(df, pos_col,one_based_start_end):\n",
    "    selector = df[pos_col].apply(lambda x: detect_in_region(x, one_based_start_end))\n",
    "    return df[selector]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "# snplist is all genes snp concat together, so may have duplicate snp\n",
    "def load_GTEx_gene_expression(geneExpression):\n",
    "    geneExpression_df = pd.read_csv(geneExpression, sep = '\\t', skiprows=2 ).iloc[:, 2:].set_index('Description').T\n",
    "    return geneExpression_df\n",
    "\n",
    "def preprocess(GWASPath, eQTLLocation, LDPath, geneExpressionPath, used_genes = None, gff=None, cis_range = 100,\n",
    "    eQTLs_col_dict=None, gwas_col_dict=None,\n",
    "    force_same = False\n",
    "    ):\n",
    "    \"\"\"\n",
    "    cis_range: int, default 100, means 100kb, the range to define cis eQTLs\n",
    "    \"\"\"\n",
    "\n",
    "    if cis_range: \n",
    "        cis_range = cis_range*1000\n",
    "\n",
    "    # Step1 load data \n",
    "    # read GWAS file\n",
    "    if gwas_col_dict is None:\n",
    "        gwas_col_dict = {}\n",
    "    gwasPOS_col = gwas_col_dict.get(\"POS\", \"pos\")\n",
    "    gwasSNP_col = gwas_col_dict.get(\"SNP\", \"ID\")\n",
    "    gwasBETA_col = gwas_col_dict.get(\"BETA\", \"BETA\")\n",
    "    gwasSE_col = gwas_col_dict.get(\"SE\", \"SE\")\n",
    "    gwas = pd.read_csv(GWASPath, sep = \"\\t\")[[gwasPOS_col, gwasSNP_col, gwasBETA_col, gwasSE_col]]\n",
    "\n",
    "    # read eQTLs\n",
    "\n",
    "    if eQTLs_col_dict is None:\n",
    "        eQTLs_col_dict = {} \n",
    "    \n",
    "    eQTLPOS_col = eQTLs_col_dict.get(\"POS\", \"BP_hg38\")\n",
    "    eQTLSNP_col = eQTLs_col_dict.get(\"SNP\", \"SNP\")\n",
    "    eQTLBETA_col = eQTLs_col_dict.get(\"BETA\", \"BETA\")\n",
    "    eQTLSE_col = eQTLs_col_dict.get(\"SE\", \"SE\")\n",
    "    eQTLs_list  = {i.stem:pd.read_csv( str(i), sep='\\t')[[eQTLPOS_col, eQTLSNP_col, eQTLBETA_col, eQTLSE_col]] for i in Path(eQTLsLocation).glob(\"*\") }\n",
    "\n",
    "    # read LD \n",
    "    # TODO: support for different LD of eQTL and GWAS\n",
    "    LD_martix = pd.read_csv(LDPath, sep = \"\\t\", header = None)\n",
    "    LD_variants = pd.read_csv(LDPath + \".vars\", sep = \"\\t\", header = None)\n",
    "    LD_martix.columns = LD_variants[0].tolist()\n",
    "    LD_martix.index = LD_variants[0].tolist()\n",
    "    # read gene expression\n",
    "    geneExpression_df = load_GTEx_gene_expression(geneExpressionPath)\n",
    "\n",
    "    genes_in_analysis = list(eQTLs_list.keys())\n",
    "            \n",
    "    # check with geneExpression_df \n",
    "    genes_in_analysis = list(set(genes_in_analysis) & set(geneExpression_df.columns.tolist()))\n",
    "    \n",
    "    if used_genes:\n",
    "        intersection = list(set(used_genes) & set(genes_in_analysis))\n",
    "        if len(intersection) != len(used_genes):\n",
    "            print(\"Some genes are not found in the eQTLs or gene expression data\")\n",
    "            print(f\"final used {len(intersection)} genes while raw used {len(genes_in_analysis)} genes\")\n",
    "        genes_in_analysis = intersection\n",
    "\n",
    "    if gff:\n",
    "        gff = pyranges.read_gff3(gff)\n",
    "        gene_gff = gff[gff.Feature == 'gene'].as_df()\n",
    "        gene_in_gff = list(set(gene_gff.Name.tolist()) & set(genes_in_analysis))\n",
    "        if len(gene_in_gff) != len(genes_in_analysis):\n",
    "            print(\"Some genes are not found in the gff file\", list(set(genes_in_analysis) - set(gene_in_gff)))\n",
    "            print(f\"final used {len(gene_in_gff)} genes while raw used {len(genes_in_analysis)} genes\")\n",
    "        genes_in_analysis = gene_in_gff\n",
    "\n",
    "        # filter eQTL_list with cis snp \n",
    "        gff_filter_cis_eQTLs_list = {}\n",
    "        gff_filter_cis_eQTLs_gene_list = [] \n",
    "        for gene in sorted(genes_in_analysis):\n",
    "            gene_region = gene_gff[gene_gff.Name == gene].iloc[0][['Start', \"End\"]].tolist()\n",
    "            upsteam = (gene_region[0] - cis_range, gene_region[0])\n",
    "            downstream = (gene_region[1], gene_region[1] + cis_range)\n",
    "            gene_cis_range = [upsteam, downstream]\n",
    "\n",
    "            filted_cis_eQTLs_snp_df = extract_cis_snps(eQTLs_list[gene], eQTLPOS_col, gene_region, gene_cis_range)\n",
    "            if filted_cis_eQTLs_snp_df.shape[0] > 0:\n",
    "                gff_filter_cis_eQTLs_list[gene] = filted_cis_eQTLs_snp_df\n",
    "                gff_filter_cis_eQTLs_gene_list.append(gene)\n",
    "            print(f\"gene {gene} has {filted_cis_eQTLs_snp_df.shape[0]} cis snps while raw has {eQTLs_list[gene].shape[0]} snps with cis range {cis_range//1000}kb and gene region is {gene_region} so the cis range of gene is {gene_cis_range}\")\n",
    "\n",
    "        genes_in_analysis = gff_filter_cis_eQTLs_gene_list\n",
    "        eQTLs_list = gff_filter_cis_eQTLs_list\n",
    "    \n",
    "\n",
    "    geneExpression_df = geneExpression_df[genes_in_analysis]\n",
    "    eQTLs_list = {gene: eQTLs_list[gene] for gene in genes_in_analysis if gene in eQTLs_list.keys()}\n",
    "\n",
    "    # step2, generate the snplist\n",
    "    if force_same:\n",
    "        snplist = []\n",
    "\n",
    "        for gene, gene_df in eQTLs_list.items():\n",
    "            snplist.append(gene_df[eQTLSNP_col].tolist())\n",
    "        from functools import reduce\n",
    "        force_same_snplist = reduce(lambda x, y: list(set(x) & set(y)), snplist)\n",
    "\n",
    "        if len(force_same_snplist) == 0:\n",
    "            raise ValueError(\"No snps are shared by all genes\")\n",
    "        else:\n",
    "            snplist =reduce(lambda x, y: x+ y,  [force_same_snplist] * len(snplist))\n",
    "        print(len(force_same_snplist))\n",
    "        force_same_eQTLs_list = {} \n",
    "        for gene, gene_df in eQTLs_list.items():\n",
    "            force_same_eQTLs_list[gene] = gene_df[gene_df[eQTLSNP_col].isin(snplist)]\n",
    "\n",
    "        eQTLs_list = force_same_eQTLs_list \n",
    "\n",
    "    else:\n",
    "        snplist = []\n",
    "\n",
    "        for gene, gene_df in eQTLs_list.items():\n",
    "            snplist += gene_df[eQTLSNP_col].tolist()\n",
    "\n",
    "\n",
    "    # step3 generate Zscore eQTLs\n",
    "    snplist_df = pd.DataFrame({eQTLSNP_col: snplist})\n",
    "    pindex = []\n",
    "    Zscore_eQTLs_list = []\n",
    "    for gene, gene_df in eQTLs_list.items():\n",
    "\n",
    "        gene_df[f'TSTAT_{gene}'] = gene_df[eQTLBETA_col]/gene_df[eQTLSE_col]# TODO: accept pre calculated tstat\n",
    "        if any(gene_df[f'TSTAT_{gene}'].isna()):\n",
    "            print(gene_df[gene_df[f'TSTAT_{gene}'].isna()])\n",
    "            raise ValueError(f\"eQTLs of {gene} has NA value in your eQTLs data at TSTAT col, plz check\")\n",
    "        # print([eQTLSNP_col, 'TSTAT'])\n",
    "\n",
    "        Zscore_eQTLs_df = snplist_df.merge(gene_df[[eQTLSNP_col, f'TSTAT_{gene}']], on = eQTLSNP_col, how = 'left')\n",
    "\n",
    "        pindex.append(gene_df.shape[0])\n",
    "        \n",
    "        Zscore_eQTLs_list.append(Zscore_eQTLs_df.set_index(eQTLSNP_col))\n",
    "    # Zscore_eQTLs =  np.stack(Zscore_eQTLs)\n",
    "    assert all_equal == all((x.index == Zscore_eQTLs_list[0].index ).all() for x in Zscore_eQTLs_list)  # check if all index are the same\n",
    "    Zscore_eQTLs_df = pd.concat(Zscore_eQTLs_list, axis = 1)\n",
    "\n",
    "    # print(len(snplist))\n",
    "    # step4 generate Zscore_GWAS     \n",
    "    gwas['TSTAT'] = gwas[gwasBETA_col]/gwas[gwasSE_col]\n",
    "    if any(gwas['TSTAT'].isna()):\n",
    "        print(gwas[gwas['TSTAT'].isna()])\n",
    "        raise ValueError(f\"GWAS has NA value in your GWAS data at TSTAT col, plz check\")\n",
    "    snplist_df = pd.DataFrame({gwasSNP_col: snplist})\n",
    "    Zscore_GWAS_df = snplist_df.merge(gwas[[gwasSNP_col, 'TSTAT']], on = gwasSNP_col, how = 'left').set_index(gwasSNP_col)\n",
    "    # print(Zscore_GWAS_df.shape[0])\n",
    "    # step5 match LD with snplist\n",
    "    LD_eQTL = LD_martix.loc[snplist, snplist] # todo: support for different LD of eQTL and GWAS\n",
    "    LD_GWAS = LD_martix.loc[snplist, snplist]\n",
    "\n",
    "\n",
    "    #step6 geneExpression corr  \n",
    "    geneExpression_corr_df = geneExpression_df.corr()\n",
    "    # check all \n",
    "    print(Zscore_GWAS_df.shape[0], sum(pindex))\n",
    "    assert sum(pindex) == Zscore_GWAS_df.shape[0], print(f\"eQTLs snps are not equal to GWAS snps, eQTLs snps are {sum(pindex)} and GWAS snps are {Zscore_GWAS_df.shape[0]}\")\n",
    "    assert [col.split(\"_\")[1] for col in Zscore_eQTLs_df.columns] == genes_in_analysis\n",
    "    assert Zscore_GWAS_df.index.tolist() == Zscore_eQTLs_df.index.tolist()\n",
    "    assert Zscore_GWAS_df.index.tolist() == snplist\n",
    "    assert Zscore_eQTLs_df.index.tolist() == LD_eQTL.index.tolist()\n",
    "    assert Zscore_GWAS_df.index.tolist() == LD_GWAS.index.tolist()\n",
    "    assert geneExpression_corr_df.index.tolist() == genes_in_analysis\n",
    "    print(f\"eQTLs matrix shape is :{Zscore_eQTLs_df.shape}\")\n",
    "    return genes_in_analysis, snplist, pindex, Zscore_eQTLs_df, Zscore_GWAS_df, geneExpression_corr_df, LD_eQTL, LD_GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "used_genes=[\"AGMAT\", \"DNAJC16\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene, snplist, pindex, Zscore_eQTLs_df, Zscore_GWAS_df, geneExpression_corr_df, LD_eQTL, LD_GWAS = preprocess(gwasPath, eQTLsLocation, plink_vcor2_LDPath, geneExpressionPath, used_genes=used_genes, gff = Ensemble_gff3, force_same=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Zscore_eQTLs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Zscore_eQTLs_df.fillna(0, inplace=True)\n",
    "Zscore_GWAS_df.fillna(0, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to h5 \n",
    "import h5py\n",
    "with h5py.File(\"GIFT.h5\", \"w\") as f:\n",
    "    f.create_group(\"data\")\n",
    "    f['data'].create_dataset('gene', data=gene)\n",
    "    f['data'].create_dataset('snplist', data=snplist)\n",
    "    f['data'].create_dataset('pindex', data=pindex)\n",
    "    f['data'].create_dataset('Zscore_eQTLs_df', data=Zscore_eQTLs_df)\n",
    "    f['data'].create_dataset('Zscore_GWAS_df', data=Zscore_GWAS_df)\n",
    "    f['data'].create_dataset('geneExpression_corr_df', data=geneExpression_corr_df)\n",
    "    f['data'].create_dataset('LD_eQTL', data=LD_eQTL)\n",
    "    f['data'].create_dataset('LD_GWAS', data=LD_GWAS)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # to h5 \n",
    "# import h5py\n",
    "# with h5py.File(\"GIFT.h5\", \"w\") as f:\n",
    "#     f.create_group(\"data\")\n",
    "#     f['data'].create_dataset('gene', data=gene)\n",
    "#     f['data'].create_dataset('snplist', data=snplist)\n",
    "#     f['data'].create_dataset('pindex', data=pindex)\n",
    "#     f['data'].create_dataset('Zscore_eQTLs', data=Zscore_eQTLs_df.values)\n",
    "#     f['data'].create_dataset('Zscore_GWAS', data=Zscore_GWAS_df.values)\n",
    "#     f['data'].create_dataset('geneExpression_corr', data=geneExpression_corr_df.values)\n",
    "#     f['data'].create_dataset('LD_eQTL', data=LD_eQTL.values)\n",
    "#     f['data'].create_dataset('LD_GWAS', data=LD_GWAS.values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
