{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_file = \"/home/xutingfeng/GIFT/data/GWAS/T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz\"\n",
    "save_dir = \"/home/xutingfeng/GIFT/data/analysis/DNAJC16/coloc\"\n",
    "from pathlib import Path\n",
    "\n",
    "Path(save_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "locus_region: 1:14583355-16583355\n",
      "Will load region from file.\n"
     ]
    }
   ],
   "source": [
    "from finemap_tools.reader.gwas import load_GWASFormated_file\n",
    "from finemap_tools.utils import add_ID\n",
    "from finemap_tools.snpfilter import filter_pipline\n",
    "\n",
    "\n",
    "topLoci = (1, 15583355, 15583355)  # (chr, start, end)\n",
    "locus_range = 1000  # 100kb\n",
    "locus_range = locus_range * 1000\n",
    "locus_range_tuple = (topLoci[0], topLoci[1] - locus_range, topLoci[2] + locus_range)\n",
    "\n",
    "locus_region = f\"{locus_range_tuple[0]}:{locus_range_tuple[1]}-{locus_range_tuple[2]}\"\n",
    "print(f\"locus_region: {locus_region}\")\n",
    "\n",
    "\n",
    "sumstats = load_GWASFormated_file(gwas_file, region=locus_region)  # load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>pval</th>\n",
       "      <th>af</th>\n",
       "      <th>beta</th>\n",
       "      <th>sebeta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.826078</td>\n",
       "      <td>0.158770</td>\n",
       "      <td>0.005861</td>\n",
       "      <td>0.026674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14583907</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.131199</td>\n",
       "      <td>0.000677</td>\n",
       "      <td>-0.556895</td>\n",
       "      <td>0.368953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14583992</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.292274</td>\n",
       "      <td>0.000678</td>\n",
       "      <td>0.388817</td>\n",
       "      <td>0.369195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14584191</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.125667</td>\n",
       "      <td>0.001244</td>\n",
       "      <td>-0.423208</td>\n",
       "      <td>0.276351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14584347</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.161867</td>\n",
       "      <td>0.005443</td>\n",
       "      <td>0.183348</td>\n",
       "      <td>0.131073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16201</th>\n",
       "      <td>1</td>\n",
       "      <td>16582952</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.158365</td>\n",
       "      <td>0.060092</td>\n",
       "      <td>0.074419</td>\n",
       "      <td>0.052757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16202</th>\n",
       "      <td>1</td>\n",
       "      <td>16583062</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.302697</td>\n",
       "      <td>0.000915</td>\n",
       "      <td>0.331578</td>\n",
       "      <td>0.321712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16203</th>\n",
       "      <td>1</td>\n",
       "      <td>16583105</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.986135</td>\n",
       "      <td>0.011341</td>\n",
       "      <td>-0.001656</td>\n",
       "      <td>0.095266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16204</th>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.591985</td>\n",
       "      <td>0.002474</td>\n",
       "      <td>-0.106121</td>\n",
       "      <td>0.198001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16205</th>\n",
       "      <td>1</td>\n",
       "      <td>16583277</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.484518</td>\n",
       "      <td>0.003413</td>\n",
       "      <td>0.117562</td>\n",
       "      <td>0.168173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16206 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom       pos ref alt      pval        af      beta    sebeta\n",
       "0          1  14583587   G   A  0.826078  0.158770  0.005861  0.026674\n",
       "1          1  14583907   T   C  0.131199  0.000677 -0.556895  0.368953\n",
       "2          1  14583992   G   A  0.292274  0.000678  0.388817  0.369195\n",
       "3          1  14584191   G   A  0.125667  0.001244 -0.423208  0.276351\n",
       "4          1  14584347   C   T  0.161867  0.005443  0.183348  0.131073\n",
       "...      ...       ...  ..  ..       ...       ...       ...       ...\n",
       "16201      1  16582952   G   T  0.158365  0.060092  0.074419  0.052757\n",
       "16202      1  16583062   G   A  0.302697  0.000915  0.331578  0.321712\n",
       "16203      1  16583105   A   C  0.986135  0.011341 -0.001656  0.095266\n",
       "16204      1  16583157   G   A  0.591985  0.002474 -0.106121  0.198001\n",
       "16205      1  16583277   T   C  0.484518  0.003413  0.117562  0.168173\n",
       "\n",
       "[16206 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "default_coloc_map = {\n",
    "    \"snp\": \"SNP\",\n",
    "    \"chr\": \"chrom\",\n",
    "    \"position\": \"pos\",\n",
    "    \"beta\": \"beta\",\n",
    "    \"varbeta\": None,\n",
    "}\n",
    "\n",
    "\n",
    "def sumstats2coloc(sumstats, map_dict=None):\n",
    "    sumstats = sumstats.copy()\n",
    "    polyfun_cols = list(default_coloc_map.keys())\n",
    "    used_polyfun_map_dict = default_coloc_map.copy()\n",
    "    if map_dict:\n",
    "        used_polyfun_map_dict.update(map_dict)\n",
    "\n",
    "    for col in polyfun_cols:\n",
    "        current_check_col = used_polyfun_map_dict.get(col, None)\n",
    "\n",
    "        if (current_check_col not in sumstats.columns) | (current_check_col is None):\n",
    "            if col == \"varbeta\":\n",
    "                # assert beta and se in map_dict\n",
    "                assert (\n",
    "                    \"sebeta\" in used_polyfun_map_dict.keys()\n",
    "                ), \"sebeta not in map_dict\"\n",
    "                sumstats[\"varbeta\"] = sumstats[used_polyfun_map_dict[\"sebeta\"]] ** 2\n",
    "\n",
    "                used_polyfun_map_dict[\"varbeta\"] = \"varbeta\"\n",
    "            else:\n",
    "                raise ValueError(f\"{current_check_col} not in sumstats\")\n",
    "\n",
    "    rename_dict = {v: k for k, v in used_polyfun_map_dict.items() if v is not None}\n",
    "    sumstats.rename(columns=rename_dict, inplace=True)\n",
    "    return sumstats[\n",
    "        list(rename_dict.values())\n",
    "        + [col for col in sumstats.columns if col not in rename_dict.values()]\n",
    "    ]\n",
    "\n",
    "\n",
    "necessary_check_cols = [\"beta\", \"varbeta\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NA nums in sumstats: beta       0\n",
      "varbeta    0\n",
      "dtype: int64 at cols: ['beta', 'varbeta']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta</th>\n",
       "      <th>varbeta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>pval</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>0.005861</td>\n",
       "      <td>0.000711</td>\n",
       "      <td>0.026674</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.826078</td>\n",
       "      <td>0.158770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:14583907:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>14583907</td>\n",
       "      <td>-0.556895</td>\n",
       "      <td>0.136126</td>\n",
       "      <td>0.368953</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.131199</td>\n",
       "      <td>0.000677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:14583992:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583992</td>\n",
       "      <td>0.388817</td>\n",
       "      <td>0.136305</td>\n",
       "      <td>0.369195</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.292274</td>\n",
       "      <td>0.000678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:14584191:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14584191</td>\n",
       "      <td>-0.423208</td>\n",
       "      <td>0.076370</td>\n",
       "      <td>0.276351</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.125667</td>\n",
       "      <td>0.001244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:14584347:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>14584347</td>\n",
       "      <td>0.183348</td>\n",
       "      <td>0.017180</td>\n",
       "      <td>0.131073</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.161867</td>\n",
       "      <td>0.005443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16201</th>\n",
       "      <td>1:16582952:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>16582952</td>\n",
       "      <td>0.074419</td>\n",
       "      <td>0.002783</td>\n",
       "      <td>0.052757</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.158365</td>\n",
       "      <td>0.060092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16202</th>\n",
       "      <td>1:16583062:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583062</td>\n",
       "      <td>0.331578</td>\n",
       "      <td>0.103499</td>\n",
       "      <td>0.321712</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.302697</td>\n",
       "      <td>0.000915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16203</th>\n",
       "      <td>1:16583105:A:C</td>\n",
       "      <td>1</td>\n",
       "      <td>16583105</td>\n",
       "      <td>-0.001656</td>\n",
       "      <td>0.009076</td>\n",
       "      <td>0.095266</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.986135</td>\n",
       "      <td>0.011341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16204</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>-0.106121</td>\n",
       "      <td>0.039204</td>\n",
       "      <td>0.198001</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.591985</td>\n",
       "      <td>0.002474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16205</th>\n",
       "      <td>1:16583277:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>16583277</td>\n",
       "      <td>0.117562</td>\n",
       "      <td>0.028282</td>\n",
       "      <td>0.168173</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.484518</td>\n",
       "      <td>0.003413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16206 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp  chr  position      beta   varbeta    sebeta ref alt  \\\n",
       "0      1:14583587:G:A    1  14583587  0.005861  0.000711  0.026674   G   A   \n",
       "1      1:14583907:T:C    1  14583907 -0.556895  0.136126  0.368953   T   C   \n",
       "2      1:14583992:G:A    1  14583992  0.388817  0.136305  0.369195   G   A   \n",
       "3      1:14584191:G:A    1  14584191 -0.423208  0.076370  0.276351   G   A   \n",
       "4      1:14584347:C:T    1  14584347  0.183348  0.017180  0.131073   C   T   \n",
       "...               ...  ...       ...       ...       ...       ...  ..  ..   \n",
       "16201  1:16582952:G:T    1  16582952  0.074419  0.002783  0.052757   G   T   \n",
       "16202  1:16583062:G:A    1  16583062  0.331578  0.103499  0.321712   G   A   \n",
       "16203  1:16583105:A:C    1  16583105 -0.001656  0.009076  0.095266   A   C   \n",
       "16204  1:16583157:G:A    1  16583157 -0.106121  0.039204  0.198001   G   A   \n",
       "16205  1:16583277:T:C    1  16583277  0.117562  0.028282  0.168173   T   C   \n",
       "\n",
       "           pval        af  \n",
       "0      0.826078  0.158770  \n",
       "1      0.131199  0.000677  \n",
       "2      0.292274  0.000678  \n",
       "3      0.125667  0.001244  \n",
       "4      0.161867  0.005443  \n",
       "...         ...       ...  \n",
       "16201  0.158365  0.060092  \n",
       "16202  0.302697  0.000915  \n",
       "16203  0.986135  0.011341  \n",
       "16204  0.591985  0.002474  \n",
       "16205  0.484518  0.003413  \n",
       "\n",
       "[16206 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats[\"SNP\"] = add_ID(\n",
    "    sumstats,\n",
    "    col_list=[\"chrom\", \"pos\", \"ref\", \"alt\"],\n",
    ")\n",
    "sumstats_coloc = sumstats2coloc(sumstats, map_dict={\"sebeta\": \"sebeta\"})\n",
    "is_na_nums_sumstats = sumstats_coloc[necessary_check_cols].isna().sum()\n",
    "print(f\"NA nums in sumstats: {is_na_nums_sumstats} at cols: {necessary_check_cols}\")\n",
    "sumstats_coloc = sumstats_coloc.dropna(subset=necessary_check_cols)\n",
    "if (dup_nums := sumstats_coloc.duplicated(subset=[\"snp\"]).sum()) > 0:\n",
    "    print(f\"duplicated SNPs in sumstats {dup_nums}\")\n",
    "    sumstats_coloc = sumstats_coloc.drop_duplicates(subset=[\"SNP\"])\n",
    "\n",
    "sumstats_coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Will load region from file.\n",
      "eQTL: 10677 rows with NA in ['beta', 'varbeta']\n",
      "duplicated SNPs in eQTL 310\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta</th>\n",
       "      <th>varbeta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>variant</th>\n",
       "      <th>r2</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>maf</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>an</th>\n",
       "      <th>ac</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.523009</td>\n",
       "      <td>0.038039</td>\n",
       "      <td>0.195036</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010016</td>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>0.109959</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.494978</td>\n",
       "      <td>0.034707</td>\n",
       "      <td>0.186299</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010676</td>\n",
       "      <td>ENSG00000162438</td>\n",
       "      <td>ENSG00000162438</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000162438</td>\n",
       "      <td>0.193525</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.231571</td>\n",
       "      <td>0.014802</td>\n",
       "      <td>0.121662</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.062993</td>\n",
       "      <td>ENSG00000142634</td>\n",
       "      <td>ENSG00000142634</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000142634</td>\n",
       "      <td>18.635300</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.275230</td>\n",
       "      <td>0.022320</td>\n",
       "      <td>0.149399</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.071621</td>\n",
       "      <td>ENSG00000175147</td>\n",
       "      <td>ENSG00000175147</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000175147</td>\n",
       "      <td>3.083490</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.141635</td>\n",
       "      <td>0.013845</td>\n",
       "      <td>0.117663</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.234596</td>\n",
       "      <td>ENSG00000171729</td>\n",
       "      <td>ENSG00000171729</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000171729</td>\n",
       "      <td>30.311600</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386642</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>-0.123758</td>\n",
       "      <td>0.745492</td>\n",
       "      <td>0.863419</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.886626</td>\n",
       "      <td>ENSG00000224459</td>\n",
       "      <td>ENSG00000224459</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000224459</td>\n",
       "      <td>0.121452</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386643</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.095035</td>\n",
       "      <td>0.488745</td>\n",
       "      <td>0.699103</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.892439</td>\n",
       "      <td>ENSG00000233954</td>\n",
       "      <td>ENSG00000233954</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000233954</td>\n",
       "      <td>4.323420</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386644</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.040898</td>\n",
       "      <td>0.191090</td>\n",
       "      <td>0.437138</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.925850</td>\n",
       "      <td>ENSG00000162458</td>\n",
       "      <td>ENSG00000162458</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000162458</td>\n",
       "      <td>17.314400</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386645</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.028620</td>\n",
       "      <td>0.267724</td>\n",
       "      <td>0.517421</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.956119</td>\n",
       "      <td>ENSG00000183888</td>\n",
       "      <td>ENSG00000183888</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000183888</td>\n",
       "      <td>0.725811</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386646</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.032089</td>\n",
       "      <td>0.476866</td>\n",
       "      <td>0.690555</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.963130</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>12.527600</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>375660 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   snp  chr  position      beta   varbeta    sebeta  \\\n",
       "0       1:14583587:G:A    1  14583587 -0.523009  0.038039  0.195036   \n",
       "1       1:14583587:G:A    1  14583587 -0.494978  0.034707  0.186299   \n",
       "2       1:14583587:G:A    1  14583587 -0.231571  0.014802  0.121662   \n",
       "3       1:14583587:G:A    1  14583587 -0.275230  0.022320  0.149399   \n",
       "4       1:14583587:G:A    1  14583587 -0.141635  0.013845  0.117663   \n",
       "...                ...  ...       ...       ...       ...       ...   \n",
       "386642  1:16583157:G:A    1  16583157 -0.123758  0.745492  0.863419   \n",
       "386643  1:16583157:G:A    1  16583157  0.095035  0.488745  0.699103   \n",
       "386644  1:16583157:G:A    1  16583157  0.040898  0.191090  0.437138   \n",
       "386645  1:16583157:G:A    1  16583157  0.028620  0.267724  0.517421   \n",
       "386646  1:16583157:G:A    1  16583157  0.032089  0.476866  0.690555   \n",
       "\n",
       "                  variant  r2    pvalue molecular_trait_object_id  \\\n",
       "0       chr1_14583587_G_A NaN  0.010016           ENSG00000215704   \n",
       "1       chr1_14583587_G_A NaN  0.010676           ENSG00000162438   \n",
       "2       chr1_14583587_G_A NaN  0.062993           ENSG00000142634   \n",
       "3       chr1_14583587_G_A NaN  0.071621           ENSG00000175147   \n",
       "4       chr1_14583587_G_A NaN  0.234596           ENSG00000171729   \n",
       "...                   ...  ..       ...                       ...   \n",
       "386642  chr1_16583157_G_A NaN  0.886626           ENSG00000224459   \n",
       "386643  chr1_16583157_G_A NaN  0.892439           ENSG00000233954   \n",
       "386644  chr1_16583157_G_A NaN  0.925850           ENSG00000162458   \n",
       "386645  chr1_16583157_G_A NaN  0.956119           ENSG00000183888   \n",
       "386646  chr1_16583157_G_A NaN  0.963130           ENSG00000186301   \n",
       "\n",
       "       molecular_trait_id       maf          gene_id  median_tpm     an  ac  \\\n",
       "0         ENSG00000215704  0.191781  ENSG00000215704    0.109959  146.0  28   \n",
       "1         ENSG00000162438  0.191781  ENSG00000162438    0.193525  146.0  28   \n",
       "2         ENSG00000142634  0.191781  ENSG00000142634   18.635300  146.0  28   \n",
       "3         ENSG00000175147  0.191781  ENSG00000175147    3.083490  146.0  28   \n",
       "4         ENSG00000171729  0.191781  ENSG00000171729   30.311600  146.0  28   \n",
       "...                   ...       ...              ...         ...    ...  ..   \n",
       "386642    ENSG00000224459  0.013699  ENSG00000224459    0.121452  146.0   2   \n",
       "386643    ENSG00000233954  0.013699  ENSG00000233954    4.323420  146.0   2   \n",
       "386644    ENSG00000162458  0.013699  ENSG00000162458   17.314400  146.0   2   \n",
       "386645    ENSG00000183888  0.013699  ENSG00000183888    0.725811  146.0   2   \n",
       "386646    ENSG00000186301  0.013699  ENSG00000186301   12.527600  146.0   2   \n",
       "\n",
       "       ref alt type         rsid  \n",
       "0        G   A  SNP   rs34982023  \n",
       "1        G   A  SNP   rs34982023  \n",
       "2        G   A  SNP   rs34982023  \n",
       "3        G   A  SNP   rs34982023  \n",
       "4        G   A  SNP   rs34982023  \n",
       "...     ..  ..  ...          ...  \n",
       "386642   G   A  SNP  rs558467297  \n",
       "386643   G   A  SNP  rs558467297  \n",
       "386644   G   A  SNP  rs558467297  \n",
       "386645   G   A  SNP  rs558467297  \n",
       "386646   G   A  SNP  rs558467297  \n",
       "\n",
       "[375660 rows x 20 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from finemap_tools.reader.eQTL import GTEx_tabix_reader\n",
    "\n",
    "\n",
    "eQTL_path = \"/home/xutingfeng/GIFT/GTEx_V8/ge/Kidney_Cortex.tsv.gz\"\n",
    "\n",
    "eQTL_df = GTEx_tabix_reader(eQTL_path, region=locus_region)\n",
    "\n",
    "\n",
    "eQTL_df[\"snp\"] = add_ID(eQTL_df, [\"chromosome\", \"position\", \"ref\", \"alt\"])\n",
    "\n",
    "GTEx_coloc_map_dict = {\n",
    "    \"snp\": \"snp\",\n",
    "    \"chr\": \"chromosome\",\n",
    "    \"position\": \"position\",\n",
    "    \"beta\": \"beta\",\n",
    "    \"sebeta\": \"se\",\n",
    "}\n",
    "eQTL_df = sumstats2coloc(eQTL_df, map_dict=GTEx_coloc_map_dict)\n",
    "is_na_num_eqtl = eQTL_df[necessary_check_cols].isna().any(axis=1).sum()\n",
    "print(f\"eQTL: {is_na_num_eqtl} rows with NA in {necessary_check_cols}\")\n",
    "eQTL_df = eQTL_df.dropna(subset=necessary_check_cols)\n",
    "\n",
    "if (\n",
    "    dup_nums := eQTL_df.duplicated(\n",
    "        subset=[\n",
    "            \"snp\",\n",
    "            \"gene_id\",\n",
    "        ]\n",
    "    ).sum()\n",
    ") > 0:\n",
    "    print(f\"duplicated SNPs in eQTL {dup_nums}\")\n",
    "    eQTL_df = eQTL_df.drop_duplicates(\n",
    "        subset=[\n",
    "            \"snp\",\n",
    "            \"gene_id\",\n",
    "        ]\n",
    "    )\n",
    "\n",
    "eQTL_df\n",
    "# eQTL_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intersection: 8132\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta</th>\n",
       "      <th>varbeta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>pval</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>0.005861</td>\n",
       "      <td>0.000711</td>\n",
       "      <td>0.026674</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.826078</td>\n",
       "      <td>0.158770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:14584424:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14584424</td>\n",
       "      <td>-0.027372</td>\n",
       "      <td>0.000395</td>\n",
       "      <td>0.019872</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.168390</td>\n",
       "      <td>0.398367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:14584509:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>14584509</td>\n",
       "      <td>-0.224028</td>\n",
       "      <td>0.046741</td>\n",
       "      <td>0.216197</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.300099</td>\n",
       "      <td>0.002376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1:14584737:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>14584737</td>\n",
       "      <td>-0.005055</td>\n",
       "      <td>0.000680</td>\n",
       "      <td>0.026071</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.846272</td>\n",
       "      <td>0.833258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1:14585405:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>14585405</td>\n",
       "      <td>0.004201</td>\n",
       "      <td>0.000699</td>\n",
       "      <td>0.026432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.873712</td>\n",
       "      <td>0.163672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16171</th>\n",
       "      <td>1:16577626:T:G</td>\n",
       "      <td>1</td>\n",
       "      <td>16577626</td>\n",
       "      <td>0.048317</td>\n",
       "      <td>0.001663</td>\n",
       "      <td>0.040774</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.236014</td>\n",
       "      <td>0.659827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16181</th>\n",
       "      <td>1:16579584:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16579584</td>\n",
       "      <td>-0.174521</td>\n",
       "      <td>0.035891</td>\n",
       "      <td>0.189450</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.356948</td>\n",
       "      <td>0.002712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16185</th>\n",
       "      <td>1:16580288:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>16580288</td>\n",
       "      <td>-0.159916</td>\n",
       "      <td>0.037436</td>\n",
       "      <td>0.193484</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.408515</td>\n",
       "      <td>0.002593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16201</th>\n",
       "      <td>1:16582952:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>16582952</td>\n",
       "      <td>0.074419</td>\n",
       "      <td>0.002783</td>\n",
       "      <td>0.052757</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.158365</td>\n",
       "      <td>0.060092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16204</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>-0.106121</td>\n",
       "      <td>0.039204</td>\n",
       "      <td>0.198001</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.591985</td>\n",
       "      <td>0.002474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8132 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  snp  chr  position      beta   varbeta    sebeta ref alt  \\\n",
       "0      1:14583587:G:A    1  14583587  0.005861  0.000711  0.026674   G   A   \n",
       "6      1:14584424:G:A    1  14584424 -0.027372  0.000395  0.019872   G   A   \n",
       "7      1:14584509:T:C    1  14584509 -0.224028  0.046741  0.216197   T   C   \n",
       "10     1:14584737:T:C    1  14584737 -0.005055  0.000680  0.026071   T   C   \n",
       "12     1:14585405:C:T    1  14585405  0.004201  0.000699  0.026432   C   T   \n",
       "...               ...  ...       ...       ...       ...       ...  ..  ..   \n",
       "16171  1:16577626:T:G    1  16577626  0.048317  0.001663  0.040774   T   G   \n",
       "16181  1:16579584:T:A    1  16579584 -0.174521  0.035891  0.189450   T   A   \n",
       "16185  1:16580288:T:C    1  16580288 -0.159916  0.037436  0.193484   T   C   \n",
       "16201  1:16582952:G:T    1  16582952  0.074419  0.002783  0.052757   G   T   \n",
       "16204  1:16583157:G:A    1  16583157 -0.106121  0.039204  0.198001   G   A   \n",
       "\n",
       "           pval        af  \n",
       "0      0.826078  0.158770  \n",
       "6      0.168390  0.398367  \n",
       "7      0.300099  0.002376  \n",
       "10     0.846272  0.833258  \n",
       "12     0.873712  0.163672  \n",
       "...         ...       ...  \n",
       "16171  0.236014  0.659827  \n",
       "16181  0.356948  0.002712  \n",
       "16185  0.408515  0.002593  \n",
       "16201  0.158365  0.060092  \n",
       "16204  0.591985  0.002474  \n",
       "\n",
       "[8132 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersection = set(sumstats_coloc[\"snp\"]).intersection(set(eQTL_df[\"snp\"]))\n",
    "print(f\"intersection: {len(intersection)}\")\n",
    "eQTL_df_gene = eQTL_df[eQTL_df[\"snp\"].isin(intersection)]\n",
    "sumstats_coloc = sumstats_coloc[sumstats_coloc[\"snp\"].isin(intersection)]\n",
    "sumstats_coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta</th>\n",
       "      <th>varbeta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>variant</th>\n",
       "      <th>r2</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>maf</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>an</th>\n",
       "      <th>ac</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>rsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.523009</td>\n",
       "      <td>0.038039</td>\n",
       "      <td>0.195036</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010016</td>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000215704</td>\n",
       "      <td>0.109959</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.494978</td>\n",
       "      <td>0.034707</td>\n",
       "      <td>0.186299</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010676</td>\n",
       "      <td>ENSG00000162438</td>\n",
       "      <td>ENSG00000162438</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000162438</td>\n",
       "      <td>0.193525</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.231571</td>\n",
       "      <td>0.014802</td>\n",
       "      <td>0.121662</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.062993</td>\n",
       "      <td>ENSG00000142634</td>\n",
       "      <td>ENSG00000142634</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000142634</td>\n",
       "      <td>18.635300</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.275230</td>\n",
       "      <td>0.022320</td>\n",
       "      <td>0.149399</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.071621</td>\n",
       "      <td>ENSG00000175147</td>\n",
       "      <td>ENSG00000175147</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000175147</td>\n",
       "      <td>3.083490</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:14583587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>14583587</td>\n",
       "      <td>-0.141635</td>\n",
       "      <td>0.013845</td>\n",
       "      <td>0.117663</td>\n",
       "      <td>chr1_14583587_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.234596</td>\n",
       "      <td>ENSG00000171729</td>\n",
       "      <td>ENSG00000171729</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>ENSG00000171729</td>\n",
       "      <td>30.311600</td>\n",
       "      <td>146.0</td>\n",
       "      <td>28</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs34982023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386642</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>-0.123758</td>\n",
       "      <td>0.745492</td>\n",
       "      <td>0.863419</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.886626</td>\n",
       "      <td>ENSG00000224459</td>\n",
       "      <td>ENSG00000224459</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000224459</td>\n",
       "      <td>0.121452</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386643</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.095035</td>\n",
       "      <td>0.488745</td>\n",
       "      <td>0.699103</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.892439</td>\n",
       "      <td>ENSG00000233954</td>\n",
       "      <td>ENSG00000233954</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000233954</td>\n",
       "      <td>4.323420</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386644</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.040898</td>\n",
       "      <td>0.191090</td>\n",
       "      <td>0.437138</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.925850</td>\n",
       "      <td>ENSG00000162458</td>\n",
       "      <td>ENSG00000162458</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000162458</td>\n",
       "      <td>17.314400</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386645</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.028620</td>\n",
       "      <td>0.267724</td>\n",
       "      <td>0.517421</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.956119</td>\n",
       "      <td>ENSG00000183888</td>\n",
       "      <td>ENSG00000183888</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000183888</td>\n",
       "      <td>0.725811</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386646</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>0.032089</td>\n",
       "      <td>0.476866</td>\n",
       "      <td>0.690555</td>\n",
       "      <td>chr1_16583157_G_A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.963130</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>12.527600</td>\n",
       "      <td>146.0</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs558467297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>375660 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   snp  chr  position      beta   varbeta    sebeta  \\\n",
       "0       1:14583587:G:A    1  14583587 -0.523009  0.038039  0.195036   \n",
       "1       1:14583587:G:A    1  14583587 -0.494978  0.034707  0.186299   \n",
       "2       1:14583587:G:A    1  14583587 -0.231571  0.014802  0.121662   \n",
       "3       1:14583587:G:A    1  14583587 -0.275230  0.022320  0.149399   \n",
       "4       1:14583587:G:A    1  14583587 -0.141635  0.013845  0.117663   \n",
       "...                ...  ...       ...       ...       ...       ...   \n",
       "386642  1:16583157:G:A    1  16583157 -0.123758  0.745492  0.863419   \n",
       "386643  1:16583157:G:A    1  16583157  0.095035  0.488745  0.699103   \n",
       "386644  1:16583157:G:A    1  16583157  0.040898  0.191090  0.437138   \n",
       "386645  1:16583157:G:A    1  16583157  0.028620  0.267724  0.517421   \n",
       "386646  1:16583157:G:A    1  16583157  0.032089  0.476866  0.690555   \n",
       "\n",
       "                  variant  r2    pvalue molecular_trait_object_id  \\\n",
       "0       chr1_14583587_G_A NaN  0.010016           ENSG00000215704   \n",
       "1       chr1_14583587_G_A NaN  0.010676           ENSG00000162438   \n",
       "2       chr1_14583587_G_A NaN  0.062993           ENSG00000142634   \n",
       "3       chr1_14583587_G_A NaN  0.071621           ENSG00000175147   \n",
       "4       chr1_14583587_G_A NaN  0.234596           ENSG00000171729   \n",
       "...                   ...  ..       ...                       ...   \n",
       "386642  chr1_16583157_G_A NaN  0.886626           ENSG00000224459   \n",
       "386643  chr1_16583157_G_A NaN  0.892439           ENSG00000233954   \n",
       "386644  chr1_16583157_G_A NaN  0.925850           ENSG00000162458   \n",
       "386645  chr1_16583157_G_A NaN  0.956119           ENSG00000183888   \n",
       "386646  chr1_16583157_G_A NaN  0.963130           ENSG00000186301   \n",
       "\n",
       "       molecular_trait_id       maf          gene_id  median_tpm     an  ac  \\\n",
       "0         ENSG00000215704  0.191781  ENSG00000215704    0.109959  146.0  28   \n",
       "1         ENSG00000162438  0.191781  ENSG00000162438    0.193525  146.0  28   \n",
       "2         ENSG00000142634  0.191781  ENSG00000142634   18.635300  146.0  28   \n",
       "3         ENSG00000175147  0.191781  ENSG00000175147    3.083490  146.0  28   \n",
       "4         ENSG00000171729  0.191781  ENSG00000171729   30.311600  146.0  28   \n",
       "...                   ...       ...              ...         ...    ...  ..   \n",
       "386642    ENSG00000224459  0.013699  ENSG00000224459    0.121452  146.0   2   \n",
       "386643    ENSG00000233954  0.013699  ENSG00000233954    4.323420  146.0   2   \n",
       "386644    ENSG00000162458  0.013699  ENSG00000162458   17.314400  146.0   2   \n",
       "386645    ENSG00000183888  0.013699  ENSG00000183888    0.725811  146.0   2   \n",
       "386646    ENSG00000186301  0.013699  ENSG00000186301   12.527600  146.0   2   \n",
       "\n",
       "       ref alt type         rsid  \n",
       "0        G   A  SNP   rs34982023  \n",
       "1        G   A  SNP   rs34982023  \n",
       "2        G   A  SNP   rs34982023  \n",
       "3        G   A  SNP   rs34982023  \n",
       "4        G   A  SNP   rs34982023  \n",
       "...     ..  ..  ...          ...  \n",
       "386642   G   A  SNP  rs558467297  \n",
       "386643   G   A  SNP  rs558467297  \n",
       "386644   G   A  SNP  rs558467297  \n",
       "386645   G   A  SNP  rs558467297  \n",
       "386646   G   A  SNP  rs558467297  \n",
       "\n",
       "[375660 rows x 20 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eQTL_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sumstats_coloc.reset_index().to_feather(f\"{save_dir}/sumstats_coloc.feather\")\n",
    "eQTL_df.reset_index().to_feather(f\"{save_dir}/eQTL_coloc.feather\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "æµ‹è¯•rpy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2 import robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "pandas2ri.activate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: This is coloc version 5.2.3\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: â€˜dplyrâ€™\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:statsâ€™:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:baseâ€™:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: Registered S3 methods overwritten by 'treeio':\n",
      "  method              from    \n",
      "  MRCA.phylo          tidytree\n",
      "  MRCA.treedata       tidytree\n",
      "  Nnode.treedata      tidytree\n",
      "  Ntip.treedata       tidytree\n",
      "  ancestor.phylo      tidytree\n",
      "  ancestor.treedata   tidytree\n",
      "  child.phylo         tidytree\n",
      "  child.treedata      tidytree\n",
      "  full_join.phylo     tidytree\n",
      "  full_join.treedata  tidytree\n",
      "  groupClade.phylo    tidytree\n",
      "  groupClade.treedata tidytree\n",
      "  groupOTU.phylo      tidytree\n",
      "  groupOTU.treedata   tidytree\n",
      "  is.rooted.treedata  tidytree\n",
      "  nodeid.phylo        tidytree\n",
      "  nodeid.treedata     tidytree\n",
      "  nodelab.phylo       tidytree\n",
      "  nodelab.treedata    tidytree\n",
      "  offspring.phylo     tidytree\n",
      "  offspring.treedata  tidytree\n",
      "  parent.phylo        tidytree\n",
      "  parent.treedata     tidytree\n",
      "  root.treedata       tidytree\n",
      "  rootnode.phylo      tidytree\n",
      "  sibling.phylo       tidytree\n",
      "\n",
      "R[write to console]: clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: â€˜clusterProfilerâ€™\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from â€˜package:statsâ€™:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: AnnotationDbi\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: â€˜BiocGenericsâ€™\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:dplyrâ€™:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:statsâ€™:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:baseâ€™:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: â€˜S4Vectorsâ€™\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from â€˜package:clusterProfilerâ€™:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:dplyrâ€™:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:baseâ€™:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: â€˜IRangesâ€™\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from â€˜package:clusterProfilerâ€™:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from â€˜package:dplyrâ€™:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: â€˜AnnotationDbiâ€™\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from â€˜package:clusterProfilerâ€™:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from â€˜package:dplyrâ€™:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/miniforge3/envs/finemap/lib/python3.8/site-packages/IPython/core/formatters.py:344\u001b[0m, in \u001b[0;36mBaseFormatter.__call__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    342\u001b[0m     method \u001b[38;5;241m=\u001b[39m get_real_method(obj, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprint_method)\n\u001b[1;32m    343\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m method \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 344\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    345\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    346\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniforge3/envs/finemap/lib/python3.8/site-packages/rpy2/robjects/vectors.py:720\u001b[0m, in \u001b[0;36mListVector._repr_html_\u001b[0;34m(self, max_items)\u001b[0m\n\u001b[1;32m    717\u001b[0m rnames \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnames\n\u001b[1;32m    718\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m max_items:\n\u001b[1;32m    719\u001b[0m     names\u001b[38;5;241m.\u001b[39mextend(\n\u001b[0;32m--> 720\u001b[0m         rnames\n\u001b[1;32m    721\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m rnames \u001b[38;5;241m!=\u001b[39m rinterface\u001b[38;5;241m.\u001b[39mNULL\n\u001b[1;32m    722\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m[no name]\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m*\u001b[39m \u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    723\u001b[0m     )\n\u001b[1;32m    724\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    725\u001b[0m     half_items \u001b[38;5;241m=\u001b[39m max_items \u001b[38;5;241m/\u001b[39m\u001b[38;5;241m/\u001b[39m \u001b[38;5;241m2\u001b[39m\n",
      "\u001b[0;31mValueError\u001b[0m: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<rpy2.robjects.vectors.ListVector object at 0x7f77b5b967c0> [RTYPES.VECSXP]\n",
       "R classes: ('list',)\n",
       "[SexpClosure, BoolSexpVector]\n",
       "  value: <class 'rpy2.rinterface.SexpClosure'>\n",
       "  <rpy2.rinterface.SexpClosure object at 0x7f77b5af17c0> [RTYPES.CLOSXP]\n",
       "  visible: <class 'rpy2.rinterface.BoolSexpVector'>\n",
       "  <rpy2.rinterface.BoolSexpVector object at 0x7f77b5afba40> [RTYPES.LGLSXP]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r = robjects.r\n",
    "r.source(\"2.2_coloc_analysis.R\")  # load script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ENSG00000215704', 'ENSG00000162438', 'ENSG00000142634',\n",
       "       'ENSG00000175147', 'ENSG00000171729', 'ENSG00000278480',\n",
       "       'ENSG00000116138', 'ENSG00000116731', 'ENSG00000189337',\n",
       "       'ENSG00000132906', 'ENSG00000142621', 'ENSG00000233485',\n",
       "       'ENSG00000272510', 'ENSG00000142615', 'ENSG00000116771',\n",
       "       'ENSG00000235084', 'ENSG00000237301', 'ENSG00000197312',\n",
       "       'ENSG00000116786', 'ENSG00000162461', 'ENSG00000237938',\n",
       "       'ENSG00000162460', 'ENSG00000224459', 'ENSG00000162458',\n",
       "       'ENSG00000224321', 'ENSG00000233954', 'ENSG00000178715',\n",
       "       'ENSG00000179743', 'ENSG00000065526', 'ENSG00000116809',\n",
       "       'ENSG00000183888', 'ENSG00000186510', 'ENSG00000173641',\n",
       "       'ENSG00000232456', 'ENSG00000184908', 'ENSG00000185519',\n",
       "       'ENSG00000227959', 'ENSG00000142627', 'ENSG00000224621',\n",
       "       'ENSG00000142632', 'ENSG00000237276', 'ENSG00000132881',\n",
       "       'ENSG00000233929', 'ENSG00000037637', 'ENSG00000055070',\n",
       "       'ENSG00000157191', 'ENSG00000226029', 'ENSG00000080947',\n",
       "       'ENSG00000261135', 'ENSG00000233421', 'ENSG00000219481',\n",
       "       'ENSG00000186301', 'ENSG00000215908', 'ENSG00000236698',\n",
       "       'ENSG00000280114', 'ENSG00000268869', 'ENSG00000186715',\n",
       "       'ENSG00000228549', 'ENSG00000235241', 'ENSG00000238142',\n",
       "       'ENSG00000272426', 'ENSG00000058453', 'ENSG00000117122',\n",
       "       'ENSG00000159363', 'ENSG00000117118', 'ENSG00000117115',\n",
       "       'ENSG00000204362', 'ENSG00000142623', 'ENSG00000159339',\n",
       "       'ENSG00000179051', 'ENSG00000074964'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eQTL_df[\"gene_id\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = eQTL_df[eQTL_df[\"gene_id\"].isin([\"ENSG00000142623\", \"ENSG00000159339\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Gene ID: ENSG00000142623\"\n",
      "[1] \"Number of SNPs in common: 1135\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.7850    0.1020    0.0907    0.0118    0.0102 \n",
      "[1] \"PP abf for shared variant: 1.02%\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Joining with `by = join_by(snp)`\n",
      "[1] \"Gene: PADI1 PP.H4.abf 0.010165684705165\"\n",
      "[1] \"Gene ID: ENSG00000159339\"\n",
      "[1] \"Number of SNPs in common: 790\"\n",
      "PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf \n",
      "   0.8410    0.0782    0.0668    0.0062    0.0078 \n",
      "[1] \"PP abf for shared variant: 0.78%\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 'select()' returned 1:1 mapping between keys and columns\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Joining with `by = join_by(snp)`\n",
      "[1] \"Gene: PADI4 PP.H4.abf 0.00779744751547604\"\n"
     ]
    }
   ],
   "source": [
    "res = r.runColocAnalysis(\n",
    "    data_gwas=sumstats_coloc,\n",
    "    data_eQTLs=test_df,\n",
    "    output_folder=\"./test\",\n",
    "    n_gwas=4869,\n",
    "    n_eQTL=73,\n",
    "    gwas_sdY=1,\n",
    "    gwas_type=\"quant\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <span>R/rpy2 DataFrame (1135 x 11)</span>\n",
       "        <table>\n",
       "          <thead>\n",
       "            <tr>\n",
       "              \n",
       "              <th>snp</th>\n",
       "              \n",
       "              <th>beta_eQTL</th>\n",
       "              \n",
       "              <th>varbeta_eQTL</th>\n",
       "              \n",
       "              <th>...</th>\n",
       "              \n",
       "              <th>varbeta_gwas</th>\n",
       "              \n",
       "              <th>pvalue_gwas</th>\n",
       "              \n",
       "              <th>PADI1_PP.H4</th>\n",
       "              \n",
       "            </tr>\n",
       "          </thead>\n",
       "          <tbody>\n",
       "          \n",
       "          <tr>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "            <td>\n",
       "              ...\n",
       "            </td>\n",
       "            \n",
       "          </tr>\n",
       "          \n",
       "          </tbody>\n",
       "        </table>\n",
       "    "
      ],
      "text/plain": [
       "<rpy2.robjects.vectors.DataFrame object at 0x7f77bbc30840> [RTYPES.VECSXP]\n",
       "R classes: ('data.frame',)\n",
       "[StrSexpVe..., FloatSexp..., FloatSexp..., FloatSexp..., ..., FloatSexp..., FloatSexp..., FloatSexp..., FloatSexp...]\n",
       "  snp: <class 'numpy.ndarray'>\n",
       "  array(['1:16206003:G:A', '1:16206308:G:A', '1:16206353:A:G', ...,\n",
       "       '1:16580288:T:C', '1:16582952:G:T', '1:16583157:G:A'], dtype='<U35')\n",
       "  beta_eQTL: <class 'numpy.ndarray'>\n",
       "  array([ 0.308864 ,  0.0434383,  0.046588 , ..., -0.163994 , -0.283698 ,\n",
       "       -0.163994 ])\n",
       "  varbeta_eQTL: <class 'numpy.ndarray'>\n",
       "  array([0.03680834, 0.03003497, 0.02994526, ..., 0.67549002, 0.05003185,\n",
       "       0.67549002])\n",
       "  maf: <class 'numpy.ndarray'>\n",
       "  array([0.130137 , 0.212329 , 0.215278 , ..., 0.0136986, 0.116438 ,\n",
       "       0.0136986])\n",
       "...\n",
       "  pvalue_eQTL: <class 'numpy.ndarray'>\n",
       "  array([ 0.00952698, -0.0635466 , -0.0652266 , ..., -0.159916  ,\n",
       "        0.0744191 , -0.106121  ])\n",
       "  rsid: <class 'numpy.ndarray'>\n",
       "  array([0.00092641, 0.00048196, 0.00048206, ..., 0.03743606, 0.00278334,\n",
       "       0.0392044 ])\n",
       "  beta_gwas: <class 'numpy.ndarray'>\n",
       "  array([0.7542755 , 0.00379665, 0.00297016, ..., 0.40851499, 0.15836529,\n",
       "       0.59198541])\n",
       "  varbeta_gwas: <class 'numpy.ndarray'>\n",
       "  array([0.0002029 , 0.00537149, 0.00669518, ..., 0.00068502, 0.00065974,\n",
       "       0.00064043])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pandas2ri.ri2py_vector(res[1][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2.robjects.conversion import localconverter\n",
    "import rpy2.robjects as ro\n",
    "\n",
    "with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "    res_0 = res[0]\n",
    "    res_1 = []\n",
    "    for k in res[1].keys():\n",
    "        pp_h4 = res[1][k].columns[-1]\n",
    "        symbol_name = pp_h4.split(\"_\")[0]\n",
    "        current_df = res[1][k][\n",
    "            [\"snp\", \"beta_eQTL\", \"varbeta_eQTL\", \"pvalue_eQTL\", pp_h4]\n",
    "        ].rename(\n",
    "            columns={\n",
    "                \"beta_eQTL\": f\"beta_eQTL_{symbol_name}\",\n",
    "                \"varbeta_eQTL\": f\"varbeta_eQTL_{symbol_name}\",\n",
    "                \"pvalue_eQTL\": f\"pvalue_eQTL_{symbol_name}\",\n",
    "            }\n",
    "        )\n",
    "        res_1.append(current_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>beta_eQTL_PADI1</th>\n",
       "      <th>varbeta_eQTL_PADI1</th>\n",
       "      <th>pvalue_eQTL_PADI1</th>\n",
       "      <th>PADI1_PP.H4</th>\n",
       "      <th>beta_eQTL_PADI4</th>\n",
       "      <th>varbeta_eQTL_PADI4</th>\n",
       "      <th>pvalue_eQTL_PADI4</th>\n",
       "      <th>PADI4_PP.H4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:16308428:C:T</td>\n",
       "      <td>-0.140821</td>\n",
       "      <td>0.034713</td>\n",
       "      <td>0.453445</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>-0.317598</td>\n",
       "      <td>0.040203</td>\n",
       "      <td>0.119765</td>\n",
       "      <td>0.000203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:16308692:A:G</td>\n",
       "      <td>-1.309590</td>\n",
       "      <td>0.657946</td>\n",
       "      <td>0.112972</td>\n",
       "      <td>0.000832</td>\n",
       "      <td>1.071350</td>\n",
       "      <td>0.811516</td>\n",
       "      <td>0.240183</td>\n",
       "      <td>0.001142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:16309214:C:T</td>\n",
       "      <td>0.004877</td>\n",
       "      <td>0.454699</td>\n",
       "      <td>0.994260</td>\n",
       "      <td>0.000712</td>\n",
       "      <td>-0.191289</td>\n",
       "      <td>0.546856</td>\n",
       "      <td>0.796992</td>\n",
       "      <td>0.000996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:16309548:A:T</td>\n",
       "      <td>0.118970</td>\n",
       "      <td>0.029544</td>\n",
       "      <td>0.492176</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.094605</td>\n",
       "      <td>0.035750</td>\n",
       "      <td>0.619112</td>\n",
       "      <td>0.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:16309919:A:G</td>\n",
       "      <td>0.106196</td>\n",
       "      <td>0.026472</td>\n",
       "      <td>0.517063</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.095910</td>\n",
       "      <td>0.031973</td>\n",
       "      <td>0.594171</td>\n",
       "      <td>0.000124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>1:16577626:T:G</td>\n",
       "      <td>-0.260081</td>\n",
       "      <td>0.023916</td>\n",
       "      <td>0.099109</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>0.248346</td>\n",
       "      <td>0.029215</td>\n",
       "      <td>0.152743</td>\n",
       "      <td>0.000647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>1:16579584:T:A</td>\n",
       "      <td>-0.163994</td>\n",
       "      <td>0.675490</td>\n",
       "      <td>0.842687</td>\n",
       "      <td>0.000704</td>\n",
       "      <td>-0.637943</td>\n",
       "      <td>0.805722</td>\n",
       "      <td>0.480706</td>\n",
       "      <td>0.000989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>1:16580288:T:C</td>\n",
       "      <td>-0.163994</td>\n",
       "      <td>0.675490</td>\n",
       "      <td>0.842687</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>-0.637943</td>\n",
       "      <td>0.805722</td>\n",
       "      <td>0.480706</td>\n",
       "      <td>0.000962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>1:16582952:G:T</td>\n",
       "      <td>-0.283698</td>\n",
       "      <td>0.050032</td>\n",
       "      <td>0.210796</td>\n",
       "      <td>0.000660</td>\n",
       "      <td>-0.172939</td>\n",
       "      <td>0.061652</td>\n",
       "      <td>0.489477</td>\n",
       "      <td>0.000793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>-0.163994</td>\n",
       "      <td>0.675490</td>\n",
       "      <td>0.842687</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>-0.637943</td>\n",
       "      <td>0.805722</td>\n",
       "      <td>0.480706</td>\n",
       "      <td>0.000900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>790 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                snp  beta_eQTL_PADI1  varbeta_eQTL_PADI1  pvalue_eQTL_PADI1  \\\n",
       "0    1:16308428:C:T        -0.140821            0.034713           0.453445   \n",
       "1    1:16308692:A:G        -1.309590            0.657946           0.112972   \n",
       "2    1:16309214:C:T         0.004877            0.454699           0.994260   \n",
       "3    1:16309548:A:T         0.118970            0.029544           0.492176   \n",
       "4    1:16309919:A:G         0.106196            0.026472           0.517063   \n",
       "..              ...              ...                 ...                ...   \n",
       "785  1:16577626:T:G        -0.260081            0.023916           0.099109   \n",
       "786  1:16579584:T:A        -0.163994            0.675490           0.842687   \n",
       "787  1:16580288:T:C        -0.163994            0.675490           0.842687   \n",
       "788  1:16582952:G:T        -0.283698            0.050032           0.210796   \n",
       "789  1:16583157:G:A        -0.163994            0.675490           0.842687   \n",
       "\n",
       "     PADI1_PP.H4  beta_eQTL_PADI4  varbeta_eQTL_PADI4  pvalue_eQTL_PADI4  \\\n",
       "0       0.000103        -0.317598            0.040203           0.119765   \n",
       "1       0.000832         1.071350            0.811516           0.240183   \n",
       "2       0.000712        -0.191289            0.546856           0.796992   \n",
       "3       0.000094         0.094605            0.035750           0.619112   \n",
       "4       0.000091         0.095910            0.031973           0.594171   \n",
       "..           ...              ...                 ...                ...   \n",
       "785     0.000543         0.248346            0.029215           0.152743   \n",
       "786     0.000704        -0.637943            0.805722           0.480706   \n",
       "787     0.000685        -0.637943            0.805722           0.480706   \n",
       "788     0.000660        -0.172939            0.061652           0.489477   \n",
       "789     0.000640        -0.637943            0.805722           0.480706   \n",
       "\n",
       "     PADI4_PP.H4  \n",
       "0       0.000203  \n",
       "1       0.001142  \n",
       "2       0.000996  \n",
       "3       0.000126  \n",
       "4       0.000124  \n",
       "..           ...  \n",
       "785     0.000647  \n",
       "786     0.000989  \n",
       "787     0.000962  \n",
       "788     0.000793  \n",
       "789     0.000900  \n",
       "\n",
       "[790 rows x 9 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from functools import reduce\n",
    "\n",
    "\n",
    "tes = reduce(lambda x, y: x.merge(y, on=[\"snp\"]), res_1)\n",
    "tes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp</th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta_GWAS</th>\n",
       "      <th>varbeta_GWAS</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>pval</th>\n",
       "      <th>af</th>\n",
       "      <th>beta_eQTL_PADI1</th>\n",
       "      <th>varbeta_eQTL_PADI1</th>\n",
       "      <th>pvalue_eQTL_PADI1</th>\n",
       "      <th>PADI1_PP.H4</th>\n",
       "      <th>beta_eQTL_PADI4</th>\n",
       "      <th>varbeta_eQTL_PADI4</th>\n",
       "      <th>pvalue_eQTL_PADI4</th>\n",
       "      <th>PADI4_PP.H4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:16308428:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>16308428</td>\n",
       "      <td>-0.004002</td>\n",
       "      <td>0.000503</td>\n",
       "      <td>0.022424</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.858357</td>\n",
       "      <td>0.241968</td>\n",
       "      <td>-0.140821</td>\n",
       "      <td>0.034713</td>\n",
       "      <td>0.453445</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>-0.317598</td>\n",
       "      <td>0.040203</td>\n",
       "      <td>0.119765</td>\n",
       "      <td>0.000203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:16308692:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>16308692</td>\n",
       "      <td>0.138096</td>\n",
       "      <td>0.010390</td>\n",
       "      <td>0.101932</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.175486</td>\n",
       "      <td>0.009110</td>\n",
       "      <td>-1.309590</td>\n",
       "      <td>0.657946</td>\n",
       "      <td>0.112972</td>\n",
       "      <td>0.000832</td>\n",
       "      <td>1.071350</td>\n",
       "      <td>0.811516</td>\n",
       "      <td>0.240183</td>\n",
       "      <td>0.001142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:16309214:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>16309214</td>\n",
       "      <td>0.186176</td>\n",
       "      <td>0.127144</td>\n",
       "      <td>0.356572</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.601582</td>\n",
       "      <td>0.000790</td>\n",
       "      <td>0.004877</td>\n",
       "      <td>0.454699</td>\n",
       "      <td>0.994260</td>\n",
       "      <td>0.000712</td>\n",
       "      <td>-0.191289</td>\n",
       "      <td>0.546856</td>\n",
       "      <td>0.796992</td>\n",
       "      <td>0.000996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:16309548:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>16309548</td>\n",
       "      <td>0.007775</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>0.019973</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.697084</td>\n",
       "      <td>0.368201</td>\n",
       "      <td>0.118970</td>\n",
       "      <td>0.029544</td>\n",
       "      <td>0.492176</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.094605</td>\n",
       "      <td>0.035750</td>\n",
       "      <td>0.619112</td>\n",
       "      <td>0.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:16309919:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>16309919</td>\n",
       "      <td>0.007188</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.020010</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.719427</td>\n",
       "      <td>0.367149</td>\n",
       "      <td>0.106196</td>\n",
       "      <td>0.026472</td>\n",
       "      <td>0.517063</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.095910</td>\n",
       "      <td>0.031973</td>\n",
       "      <td>0.594171</td>\n",
       "      <td>0.000124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>1:16577626:T:G</td>\n",
       "      <td>1</td>\n",
       "      <td>16577626</td>\n",
       "      <td>0.048317</td>\n",
       "      <td>0.001663</td>\n",
       "      <td>0.040774</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.236014</td>\n",
       "      <td>0.659827</td>\n",
       "      <td>-0.260081</td>\n",
       "      <td>0.023916</td>\n",
       "      <td>0.099109</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>0.248346</td>\n",
       "      <td>0.029215</td>\n",
       "      <td>0.152743</td>\n",
       "      <td>0.000647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>786</th>\n",
       "      <td>1:16579584:T:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16579584</td>\n",
       "      <td>-0.174521</td>\n",
       "      <td>0.035891</td>\n",
       "      <td>0.189450</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.356948</td>\n",
       "      <td>0.002712</td>\n",
       "      <td>-0.163994</td>\n",
       "      <td>0.675490</td>\n",
       "      <td>0.842687</td>\n",
       "      <td>0.000704</td>\n",
       "      <td>-0.637943</td>\n",
       "      <td>0.805722</td>\n",
       "      <td>0.480706</td>\n",
       "      <td>0.000989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787</th>\n",
       "      <td>1:16580288:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>16580288</td>\n",
       "      <td>-0.159916</td>\n",
       "      <td>0.037436</td>\n",
       "      <td>0.193484</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.408515</td>\n",
       "      <td>0.002593</td>\n",
       "      <td>-0.163994</td>\n",
       "      <td>0.675490</td>\n",
       "      <td>0.842687</td>\n",
       "      <td>0.000685</td>\n",
       "      <td>-0.637943</td>\n",
       "      <td>0.805722</td>\n",
       "      <td>0.480706</td>\n",
       "      <td>0.000962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>1:16582952:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>16582952</td>\n",
       "      <td>0.074419</td>\n",
       "      <td>0.002783</td>\n",
       "      <td>0.052757</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.158365</td>\n",
       "      <td>0.060092</td>\n",
       "      <td>-0.283698</td>\n",
       "      <td>0.050032</td>\n",
       "      <td>0.210796</td>\n",
       "      <td>0.000660</td>\n",
       "      <td>-0.172939</td>\n",
       "      <td>0.061652</td>\n",
       "      <td>0.489477</td>\n",
       "      <td>0.000793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>1:16583157:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>16583157</td>\n",
       "      <td>-0.106121</td>\n",
       "      <td>0.039204</td>\n",
       "      <td>0.198001</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.591985</td>\n",
       "      <td>0.002474</td>\n",
       "      <td>-0.163994</td>\n",
       "      <td>0.675490</td>\n",
       "      <td>0.842687</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>-0.637943</td>\n",
       "      <td>0.805722</td>\n",
       "      <td>0.480706</td>\n",
       "      <td>0.000900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>790 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                snp  chr  position  beta_GWAS  varbeta_GWAS    sebeta ref alt  \\\n",
       "0    1:16308428:C:T    1  16308428  -0.004002      0.000503  0.022424   C   T   \n",
       "1    1:16308692:A:G    1  16308692   0.138096      0.010390  0.101932   A   G   \n",
       "2    1:16309214:C:T    1  16309214   0.186176      0.127144  0.356572   C   T   \n",
       "3    1:16309548:A:T    1  16309548   0.007775      0.000399  0.019973   A   T   \n",
       "4    1:16309919:A:G    1  16309919   0.007188      0.000400  0.020010   A   G   \n",
       "..              ...  ...       ...        ...           ...       ...  ..  ..   \n",
       "785  1:16577626:T:G    1  16577626   0.048317      0.001663  0.040774   T   G   \n",
       "786  1:16579584:T:A    1  16579584  -0.174521      0.035891  0.189450   T   A   \n",
       "787  1:16580288:T:C    1  16580288  -0.159916      0.037436  0.193484   T   C   \n",
       "788  1:16582952:G:T    1  16582952   0.074419      0.002783  0.052757   G   T   \n",
       "789  1:16583157:G:A    1  16583157  -0.106121      0.039204  0.198001   G   A   \n",
       "\n",
       "         pval        af  beta_eQTL_PADI1  varbeta_eQTL_PADI1  \\\n",
       "0    0.858357  0.241968        -0.140821            0.034713   \n",
       "1    0.175486  0.009110        -1.309590            0.657946   \n",
       "2    0.601582  0.000790         0.004877            0.454699   \n",
       "3    0.697084  0.368201         0.118970            0.029544   \n",
       "4    0.719427  0.367149         0.106196            0.026472   \n",
       "..        ...       ...              ...                 ...   \n",
       "785  0.236014  0.659827        -0.260081            0.023916   \n",
       "786  0.356948  0.002712        -0.163994            0.675490   \n",
       "787  0.408515  0.002593        -0.163994            0.675490   \n",
       "788  0.158365  0.060092        -0.283698            0.050032   \n",
       "789  0.591985  0.002474        -0.163994            0.675490   \n",
       "\n",
       "     pvalue_eQTL_PADI1  PADI1_PP.H4  beta_eQTL_PADI4  varbeta_eQTL_PADI4  \\\n",
       "0             0.453445     0.000103        -0.317598            0.040203   \n",
       "1             0.112972     0.000832         1.071350            0.811516   \n",
       "2             0.994260     0.000712        -0.191289            0.546856   \n",
       "3             0.492176     0.000094         0.094605            0.035750   \n",
       "4             0.517063     0.000091         0.095910            0.031973   \n",
       "..                 ...          ...              ...                 ...   \n",
       "785           0.099109     0.000543         0.248346            0.029215   \n",
       "786           0.842687     0.000704        -0.637943            0.805722   \n",
       "787           0.842687     0.000685        -0.637943            0.805722   \n",
       "788           0.210796     0.000660        -0.172939            0.061652   \n",
       "789           0.842687     0.000640        -0.637943            0.805722   \n",
       "\n",
       "     pvalue_eQTL_PADI4  PADI4_PP.H4  \n",
       "0             0.119765     0.000203  \n",
       "1             0.240183     0.001142  \n",
       "2             0.796992     0.000996  \n",
       "3             0.619112     0.000126  \n",
       "4             0.594171     0.000124  \n",
       "..                 ...          ...  \n",
       "785           0.152743     0.000647  \n",
       "786           0.480706     0.000989  \n",
       "787           0.480706     0.000962  \n",
       "788           0.489477     0.000793  \n",
       "789           0.480706     0.000900  \n",
       "\n",
       "[790 rows x 18 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats_coloc.rename(columns={\"beta\": \"beta_GWAS\", \"varbeta\": \"varbeta_GWAS\"}).merge(\n",
    "    tes, how=\"inner\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
