{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from finemap_tools.finemap.coloc import coloc_Wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "locus_region: 1:14083355-17083355\n"
     ]
    }
   ],
   "source": [
    "topLoci = (1, 15583355, 15583355)  # (chr, start, end)\n",
    "locus_range = 1500  # 100kb\n",
    "locus_range = locus_range * 1000\n",
    "locus_range_tuple = (topLoci[0], topLoci[1] - locus_range, topLoci[2] + locus_range)\n",
    "\n",
    "locus_region = f\"{locus_range_tuple[0]}:{locus_range_tuple[1]}-{locus_range_tuple[2]}\"\n",
    "print(f\"locus_region: {locus_region}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_file = \"/home/xutingfeng/GIFT/data/GWAS/T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz\"\n",
    "save_dir = \"/home/xutingfeng/GIFT/data/analysis/DNAJC16/coloc\"\n",
    "from pathlib import Path\n",
    "\n",
    "Path(save_dir).mkdir(parents=True, exist_ok=True)\n",
    "eQTL_path = \"/home/xutingfeng/GIFT/GTEx_V8/ge/Kidney_Cortex.tsv.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from finemap_tools.others.polyfun.polyfun_utils import (\n",
    "    configure_logger,\n",
    ")\n",
    "\n",
    "configure_logger(\"./log.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO]  loading with normal tabix reader to loading sumstats file with /home/xutingfeng/GIFT/data/GWAS/T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz......\n",
      "Will load region from file.\n",
      "[INFO]  the header of loaded file is Index(['chrom', 'pos', 'ref', 'alt', 'pval', 'af', 'beta', 'sebeta'], dtype='object'), if it is not the correct header, please check the file or with correct format specified.\n",
      "[INFO]  filtering SNP by finemap_tools with 23225 SNP at begining........\n",
      "Function find_amibiguous_alleles elapsed time: 0.1960 seconds\n",
      "[INFO]  drop 2966 ambiguous alleles\n",
      "Function find_Biallelic_snp elapsed time: 0.0115 seconds\n",
      "[INFO]  drop 36 biallelic snps\n",
      "[INFO]  after filtering, left 20223 SNP\n",
      "[INFO]  na numbers in sumstats T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz: 0\n",
      "[INFO]  loading GTExV8 format sumstats file with /home/xutingfeng/GIFT/GTEx_V8/ge/Kidney_Cortex.tsv.gz......\n",
      "Will load region from file.\n",
      "[INFO]  the header of loaded file is Index(['variant', 'r2', 'pvalue', 'molecular_trait_object_id',\n",
      "       'molecular_trait_id', 'maf', 'gene_id', 'median_tpm', 'beta', 'se',\n",
      "       'an', 'ac', 'chromosome', 'position', 'ref', 'alt', 'type', 'rsid'],\n",
      "      dtype='object'), if it is not the correct header, please check the file or with correct format specified.\n",
      "[INFO]  filtering SNP by finemap_tools with 460789 SNP at begining........\n",
      "Function find_amibiguous_alleles elapsed time: 3.0842 seconds\n",
      "[INFO]  drop 59502 ambiguous alleles\n",
      "Function find_Biallelic_snp elapsed time: 0.0076 seconds\n",
      "[INFO]  drop 2 biallelic snps\n",
      "[INFO]  after filtering, left 401227 SNP\n",
      "[INFO]  na numbers in sumstats Kidney_Cortex.tsv.gz: 1\n"
     ]
    }
   ],
   "source": [
    "coloc = coloc_Wrapper(\n",
    "    sumstats_file1=gwas_file,\n",
    "    sumstats_file1_format=\"format\",\n",
    "    sumstats_file2=eQTL_path,\n",
    "    sumstats_file2_format=\"GTExV8\",\n",
    "    n1=4786,\n",
    "    n2=73,\n",
    "    type1=\"quant\",\n",
    "    sdY1=1,\n",
    "    chr_num=locus_range_tuple[0],\n",
    "    start=locus_range_tuple[1],\n",
    "    end=locus_range_tuple[2],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta</th>\n",
       "      <th>varbeta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref</th>\n",
       "      <th>variant</th>\n",
       "      <th>r2</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>maf</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>an</th>\n",
       "      <th>ac</th>\n",
       "      <th>type</th>\n",
       "      <th>rsid</th>\n",
       "      <th>snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14083559</td>\n",
       "      <td>0.101611</td>\n",
       "      <td>0.037131</td>\n",
       "      <td>0.192693</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14083559_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.600399</td>\n",
       "      <td>ENSG00000116731</td>\n",
       "      <td>ENSG00000116731</td>\n",
       "      <td>0.034247</td>\n",
       "      <td>ENSG00000116731</td>\n",
       "      <td>5.466610</td>\n",
       "      <td>146.0</td>\n",
       "      <td>5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs116430980</td>\n",
       "      <td>1:14083559:A:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14083559</td>\n",
       "      <td>0.148856</td>\n",
       "      <td>0.117537</td>\n",
       "      <td>0.342836</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14083559_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.666095</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>0.034247</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>8.443570</td>\n",
       "      <td>146.0</td>\n",
       "      <td>5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs116430980</td>\n",
       "      <td>1:14083559:A:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14083559</td>\n",
       "      <td>0.047898</td>\n",
       "      <td>0.121123</td>\n",
       "      <td>0.348027</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14083559_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.891112</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>0.034247</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>2.191650</td>\n",
       "      <td>146.0</td>\n",
       "      <td>5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs116430980</td>\n",
       "      <td>1:14083559:A:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14083614</td>\n",
       "      <td>0.169648</td>\n",
       "      <td>0.030147</td>\n",
       "      <td>0.173629</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14083614_T_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.333433</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>8.443570</td>\n",
       "      <td>146.0</td>\n",
       "      <td>20</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs35028645</td>\n",
       "      <td>1:14083614:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14083614</td>\n",
       "      <td>0.095978</td>\n",
       "      <td>0.031381</td>\n",
       "      <td>0.177148</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14083614_T_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.590463</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>2.191650</td>\n",
       "      <td>146.0</td>\n",
       "      <td>20</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs35028645</td>\n",
       "      <td>1:14083614:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460749</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>0.076296</td>\n",
       "      <td>0.061443</td>\n",
       "      <td>0.247877</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.759567</td>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>2.563180</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460750</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.141632</td>\n",
       "      <td>0.349522</td>\n",
       "      <td>0.591204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.811687</td>\n",
       "      <td>ENSG00000186715</td>\n",
       "      <td>ENSG00000186715</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000186715</td>\n",
       "      <td>13.809000</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460751</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.039539</td>\n",
       "      <td>0.142180</td>\n",
       "      <td>0.377067</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.916925</td>\n",
       "      <td>ENSG00000179051</td>\n",
       "      <td>ENSG00000179051</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000179051</td>\n",
       "      <td>14.486300</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460752</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.040268</td>\n",
       "      <td>0.282896</td>\n",
       "      <td>0.531880</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.939965</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>12.527600</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460753</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.012926</td>\n",
       "      <td>0.423629</td>\n",
       "      <td>0.650868</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.984237</td>\n",
       "      <td>ENSG00000204362</td>\n",
       "      <td>ENSG00000204362</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000204362</td>\n",
       "      <td>0.036613</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>391256 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr  position      beta   varbeta    sebeta alt ref  \\\n",
       "0         1  14083559  0.101611  0.037131  0.192693   C   A   \n",
       "1         1  14083559  0.148856  0.117537  0.342836   C   A   \n",
       "2         1  14083559  0.047898  0.121123  0.348027   C   A   \n",
       "3         1  14083614  0.169648  0.030147  0.173629   C   T   \n",
       "4         1  14083614  0.095978  0.031381  0.177148   C   T   \n",
       "...     ...       ...       ...       ...       ...  ..  ..   \n",
       "460749    1  17083276  0.076296  0.061443  0.247877   T   C   \n",
       "460750    1  17083276 -0.141632  0.349522  0.591204   T   C   \n",
       "460751    1  17083276 -0.039539  0.142180  0.377067   T   C   \n",
       "460752    1  17083276 -0.040268  0.282896  0.531880   T   C   \n",
       "460753    1  17083276 -0.012926  0.423629  0.650868   T   C   \n",
       "\n",
       "                  variant  r2    pvalue molecular_trait_object_id  \\\n",
       "0       chr1_14083559_A_C NaN  0.600399           ENSG00000116731   \n",
       "1       chr1_14083559_A_C NaN  0.666095           ENSG00000189337   \n",
       "2       chr1_14083559_A_C NaN  0.891112           ENSG00000162493   \n",
       "3       chr1_14083614_T_C NaN  0.333433           ENSG00000189337   \n",
       "4       chr1_14083614_T_C NaN  0.590463           ENSG00000162493   \n",
       "...                   ...  ..       ...                       ...   \n",
       "460749  chr1_17083276_C_T NaN  0.759567           ENSG00000080947   \n",
       "460750  chr1_17083276_C_T NaN  0.811687           ENSG00000186715   \n",
       "460751  chr1_17083276_C_T NaN  0.916925           ENSG00000179051   \n",
       "460752  chr1_17083276_C_T NaN  0.939965           ENSG00000186301   \n",
       "460753  chr1_17083276_C_T NaN  0.984237           ENSG00000204362   \n",
       "\n",
       "       molecular_trait_id       maf          gene_id  median_tpm     an  ac  \\\n",
       "0         ENSG00000116731  0.034247  ENSG00000116731    5.466610  146.0   5   \n",
       "1         ENSG00000189337  0.034247  ENSG00000189337    8.443570  146.0   5   \n",
       "2         ENSG00000162493  0.034247  ENSG00000162493    2.191650  146.0   5   \n",
       "3         ENSG00000189337  0.136986  ENSG00000189337    8.443570  146.0  20   \n",
       "4         ENSG00000162493  0.136986  ENSG00000162493    2.191650  146.0  20   \n",
       "...                   ...       ...              ...         ...    ...  ..   \n",
       "460749    ENSG00000080947  0.020548  ENSG00000080947    2.563180  146.0   3   \n",
       "460750    ENSG00000186715  0.020548  ENSG00000186715   13.809000  146.0   3   \n",
       "460751    ENSG00000179051  0.020548  ENSG00000179051   14.486300  146.0   3   \n",
       "460752    ENSG00000186301  0.020548  ENSG00000186301   12.527600  146.0   3   \n",
       "460753    ENSG00000204362  0.020548  ENSG00000204362    0.036613  146.0   3   \n",
       "\n",
       "       type         rsid             snp  \n",
       "0       SNP  rs116430980  1:14083559:A:C  \n",
       "1       SNP  rs116430980  1:14083559:A:C  \n",
       "2       SNP  rs116430980  1:14083559:A:C  \n",
       "3       SNP   rs35028645  1:14083614:T:C  \n",
       "4       SNP   rs35028645  1:14083614:T:C  \n",
       "...     ...          ...             ...  \n",
       "460749  SNP    rs2746525  1:17083276:C:T  \n",
       "460750  SNP    rs2746525  1:17083276:C:T  \n",
       "460751  SNP    rs2746525  1:17083276:C:T  \n",
       "460752  SNP    rs2746525  1:17083276:C:T  \n",
       "460753  SNP    rs2746525  1:17083276:C:T  \n",
       "\n",
       "[391256 rows x 20 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coloc.sumstats_file2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(73,)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coloc.sumstats_file2[\"gene_id\"].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_col = \"gene_id\"\n",
    "# drop duplicated SNPs\n",
    "sumstats1_duplicated_num = coloc.sumstats_file1.duplicated(subset=[\"snp\"]).sum()\n",
    "sumstats2_duplicated_num = coloc.sumstats_file2.duplicated(\n",
    "    subset=[\"snp\", gene_col]\n",
    ").sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO]  before sync , sumstats1 have 20223 SNPs, sumstats2 have 391070 SNPs\n",
      "[INFO]  after sync, left 9818 SNPs\n"
     ]
    }
   ],
   "source": [
    "# if sumstats1_duplicated_num > 0:\n",
    "#     logging.warning(\n",
    "#         f\"sumstats1 have {sumstats1_duplicated_num} duplicated SNPs and will be dropped\"\n",
    "#     )\n",
    "# if sumstats2_duplicated_num > 0:\n",
    "#     logging.warning(\n",
    "#         f\"sumstats2 have {sumstats2_duplicated_num} duplicated SNPs and will be dropped\"\n",
    "#     )\n",
    "\n",
    "coloc.sumstats_file1 = coloc.sumstats_file1.drop_duplicates(subset=[\"snp\"])\n",
    "coloc.sumstats_file2 = coloc.sumstats_file2.drop_duplicates(subset=[\"snp\", gene_col])\n",
    "\n",
    "assert (\n",
    "    coloc.sumstats_file1.shape[0] > 0 and coloc.sumstats_file2.shape[0] > 0\n",
    "), \"no valid SNPs to run coloc after drop duplicated SNPs\"\n",
    "\n",
    "# sync sumstats\n",
    "coloc.sync_sumstats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>position</th>\n",
       "      <th>beta</th>\n",
       "      <th>varbeta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref</th>\n",
       "      <th>variant</th>\n",
       "      <th>r2</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>molecular_trait_object_id</th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>maf</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>median_tpm</th>\n",
       "      <th>an</th>\n",
       "      <th>ac</th>\n",
       "      <th>type</th>\n",
       "      <th>rsid</th>\n",
       "      <th>snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14083559</td>\n",
       "      <td>0.101611</td>\n",
       "      <td>0.037131</td>\n",
       "      <td>0.192693</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14083559_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.600399</td>\n",
       "      <td>ENSG00000116731</td>\n",
       "      <td>ENSG00000116731</td>\n",
       "      <td>0.034247</td>\n",
       "      <td>ENSG00000116731</td>\n",
       "      <td>5.466610</td>\n",
       "      <td>146.0</td>\n",
       "      <td>5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs116430980</td>\n",
       "      <td>1:14083559:A:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14083559</td>\n",
       "      <td>0.148856</td>\n",
       "      <td>0.117537</td>\n",
       "      <td>0.342836</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14083559_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.666095</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>0.034247</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>8.443570</td>\n",
       "      <td>146.0</td>\n",
       "      <td>5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs116430980</td>\n",
       "      <td>1:14083559:A:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14083559</td>\n",
       "      <td>0.047898</td>\n",
       "      <td>0.121123</td>\n",
       "      <td>0.348027</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1_14083559_A_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.891112</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>0.034247</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>2.191650</td>\n",
       "      <td>146.0</td>\n",
       "      <td>5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs116430980</td>\n",
       "      <td>1:14083559:A:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>14083614</td>\n",
       "      <td>0.169648</td>\n",
       "      <td>0.030147</td>\n",
       "      <td>0.173629</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14083614_T_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.333433</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>ENSG00000189337</td>\n",
       "      <td>8.443570</td>\n",
       "      <td>146.0</td>\n",
       "      <td>20</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs35028645</td>\n",
       "      <td>1:14083614:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>14083614</td>\n",
       "      <td>0.095978</td>\n",
       "      <td>0.031381</td>\n",
       "      <td>0.177148</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1_14083614_T_C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.590463</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>0.136986</td>\n",
       "      <td>ENSG00000162493</td>\n",
       "      <td>2.191650</td>\n",
       "      <td>146.0</td>\n",
       "      <td>20</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs35028645</td>\n",
       "      <td>1:14083614:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460749</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>0.076296</td>\n",
       "      <td>0.061443</td>\n",
       "      <td>0.247877</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.759567</td>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000080947</td>\n",
       "      <td>2.563180</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460750</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.141632</td>\n",
       "      <td>0.349522</td>\n",
       "      <td>0.591204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.811687</td>\n",
       "      <td>ENSG00000186715</td>\n",
       "      <td>ENSG00000186715</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000186715</td>\n",
       "      <td>13.809000</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460751</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.039539</td>\n",
       "      <td>0.142180</td>\n",
       "      <td>0.377067</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.916925</td>\n",
       "      <td>ENSG00000179051</td>\n",
       "      <td>ENSG00000179051</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000179051</td>\n",
       "      <td>14.486300</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460752</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.040268</td>\n",
       "      <td>0.282896</td>\n",
       "      <td>0.531880</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.939965</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000186301</td>\n",
       "      <td>12.527600</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460753</th>\n",
       "      <td>1</td>\n",
       "      <td>17083276</td>\n",
       "      <td>-0.012926</td>\n",
       "      <td>0.423629</td>\n",
       "      <td>0.650868</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1_17083276_C_T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.984237</td>\n",
       "      <td>ENSG00000204362</td>\n",
       "      <td>ENSG00000204362</td>\n",
       "      <td>0.020548</td>\n",
       "      <td>ENSG00000204362</td>\n",
       "      <td>0.036613</td>\n",
       "      <td>146.0</td>\n",
       "      <td>3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>rs2746525</td>\n",
       "      <td>1:17083276:C:T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>363919 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr  position      beta   varbeta    sebeta alt ref  \\\n",
       "0         1  14083559  0.101611  0.037131  0.192693   C   A   \n",
       "1         1  14083559  0.148856  0.117537  0.342836   C   A   \n",
       "2         1  14083559  0.047898  0.121123  0.348027   C   A   \n",
       "3         1  14083614  0.169648  0.030147  0.173629   C   T   \n",
       "4         1  14083614  0.095978  0.031381  0.177148   C   T   \n",
       "...     ...       ...       ...       ...       ...  ..  ..   \n",
       "460749    1  17083276  0.076296  0.061443  0.247877   T   C   \n",
       "460750    1  17083276 -0.141632  0.349522  0.591204   T   C   \n",
       "460751    1  17083276 -0.039539  0.142180  0.377067   T   C   \n",
       "460752    1  17083276 -0.040268  0.282896  0.531880   T   C   \n",
       "460753    1  17083276 -0.012926  0.423629  0.650868   T   C   \n",
       "\n",
       "                  variant  r2    pvalue molecular_trait_object_id  \\\n",
       "0       chr1_14083559_A_C NaN  0.600399           ENSG00000116731   \n",
       "1       chr1_14083559_A_C NaN  0.666095           ENSG00000189337   \n",
       "2       chr1_14083559_A_C NaN  0.891112           ENSG00000162493   \n",
       "3       chr1_14083614_T_C NaN  0.333433           ENSG00000189337   \n",
       "4       chr1_14083614_T_C NaN  0.590463           ENSG00000162493   \n",
       "...                   ...  ..       ...                       ...   \n",
       "460749  chr1_17083276_C_T NaN  0.759567           ENSG00000080947   \n",
       "460750  chr1_17083276_C_T NaN  0.811687           ENSG00000186715   \n",
       "460751  chr1_17083276_C_T NaN  0.916925           ENSG00000179051   \n",
       "460752  chr1_17083276_C_T NaN  0.939965           ENSG00000186301   \n",
       "460753  chr1_17083276_C_T NaN  0.984237           ENSG00000204362   \n",
       "\n",
       "       molecular_trait_id       maf          gene_id  median_tpm     an  ac  \\\n",
       "0         ENSG00000116731  0.034247  ENSG00000116731    5.466610  146.0   5   \n",
       "1         ENSG00000189337  0.034247  ENSG00000189337    8.443570  146.0   5   \n",
       "2         ENSG00000162493  0.034247  ENSG00000162493    2.191650  146.0   5   \n",
       "3         ENSG00000189337  0.136986  ENSG00000189337    8.443570  146.0  20   \n",
       "4         ENSG00000162493  0.136986  ENSG00000162493    2.191650  146.0  20   \n",
       "...                   ...       ...              ...         ...    ...  ..   \n",
       "460749    ENSG00000080947  0.020548  ENSG00000080947    2.563180  146.0   3   \n",
       "460750    ENSG00000186715  0.020548  ENSG00000186715   13.809000  146.0   3   \n",
       "460751    ENSG00000179051  0.020548  ENSG00000179051   14.486300  146.0   3   \n",
       "460752    ENSG00000186301  0.020548  ENSG00000186301   12.527600  146.0   3   \n",
       "460753    ENSG00000204362  0.020548  ENSG00000204362    0.036613  146.0   3   \n",
       "\n",
       "       type         rsid             snp  \n",
       "0       SNP  rs116430980  1:14083559:A:C  \n",
       "1       SNP  rs116430980  1:14083559:A:C  \n",
       "2       SNP  rs116430980  1:14083559:A:C  \n",
       "3       SNP   rs35028645  1:14083614:T:C  \n",
       "4       SNP   rs35028645  1:14083614:T:C  \n",
       "...     ...          ...             ...  \n",
       "460749  SNP    rs2746525  1:17083276:C:T  \n",
       "460750  SNP    rs2746525  1:17083276:C:T  \n",
       "460751  SNP    rs2746525  1:17083276:C:T  \n",
       "460752  SNP    rs2746525  1:17083276:C:T  \n",
       "460753  SNP    rs2746525  1:17083276:C:T  \n",
       "\n",
       "[363919 rows x 20 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coloc.sumstats_file2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_duplicated"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "finemap",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
