{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from finemap_tools.finemap.coloc import coloc_Wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "locus_region: 1:15383355-15783355\n"
     ]
    }
   ],
   "source": [
    "topLoci = (1, 15583355, 15583355)  # (chr, start, end)\n",
    "locus_range = 200  # 100kb\n",
    "locus_range = locus_range * 1000\n",
    "locus_range_tuple = (topLoci[0], topLoci[1] - locus_range, topLoci[2] + locus_range)\n",
    "\n",
    "locus_region = f\"{locus_range_tuple[0]}:{locus_range_tuple[1]}-{locus_range_tuple[2]}\"\n",
    "print(f\"locus_region: {locus_region}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_file = \"/home/xutingfeng/GIFT/data/GWAS/T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz\"\n",
    "gwas2_file = (\n",
    "    \"/home/xutingfeng/GIFT/data/GWAS/CKD_overall_ALL_JW_20180223_nstud30.dbgap.tsv.gz\"\n",
    ")\n",
    "save_dir = \"/home/xutingfeng/GIFT/data/analysis/DNAJC16/coloc\"\n",
    "from pathlib import Path\n",
    "\n",
    "Path(save_dir).mkdir(parents=True, exist_ok=True)\n",
    "# eQTL_path = \"/home/xutingfeng/GIFT/GTEx_V8/ge/Kidney_Cortex.tsv.gz\"\n",
    "eQTL_path = \"/home/xutingfeng/GIFT/GTEx_V8/ge/Adipose_Subcutaneous.tsv.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from finemap_tools.others.polyfun.polyfun_utils import (\n",
    "    configure_logger,\n",
    ")\n",
    "\n",
    "configure_logger(\"./log.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # test two sample coloc\n",
    "# coloc = coloc_Wrapper(\n",
    "#     sumstats_file=gwas_file,\n",
    "#     sumstats_file_format=\"pheweb\",\n",
    "#     sumstats_file2=gwas2_file,\n",
    "#     sumstats_file2_format=\"pheweb\",\n",
    "#     n1=4786,\n",
    "#     n2=73,\n",
    "#     type1=\"quant\",\n",
    "#     type2=\"cc\",\n",
    "#     sdY1=1,\n",
    "#     sdY2=None,\n",
    "#     chr_num=locus_range_tuple[0],\n",
    "#     start=locus_range_tuple[1],\n",
    "#     end=locus_range_tuple[2],\n",
    "# )\n",
    "# s1, s2, ld = coloc.finemap(\n",
    "#     num_causal_snps=1,\n",
    "#     out_dir=\"./test_two_gwas\",\n",
    "#     ld_file=\"/home/xutingfeng/GIFT/data/analysis/DNAJC16/polyfun/ld/DNAJC16_GRCh38.1.14403350.15833356.npz\",\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coloc = coloc_Wrapper(\n",
    "#     sumstats_file=gwas_file,\n",
    "#     sumstats_file_format=\"pheweb\",\n",
    "#     sumstats_file2=eQTL_path,\n",
    "#     sumstats_file2_format=\"GTExV8\",\n",
    "#     n1=4786,\n",
    "#     n2=73,\n",
    "#     type1=\"quant\",\n",
    "#     type2=\"quant\",\n",
    "#     sdY1=1,\n",
    "#     sdY2=None,\n",
    "#     chr_num=locus_range_tuple[0],\n",
    "#     start=locus_range_tuple[1],\n",
    "#     end=locus_range_tuple[2],\n",
    "# )\n",
    "# s1 = coloc.finemap(\n",
    "#     num_causal_snps=1,\n",
    "#     out_dir=\"./test\",\n",
    "\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO]  Loading sumstats file...\n",
      "[INFO]  loading with normal tabix reader to loading sumstats file with /home/xutingfeng/GIFT/data/GWAS/T1Mapping_Cortex_20240129.csv_firstorder_Median_all_2023_GRCh38_unionKidneys.tsv.gz......\n",
      "Will load region from file.\n",
      "[INFO]  the header of loaded file is Index(['chrom', 'pos', 'ref', 'alt', 'pval', 'af', 'beta', 'sebeta'], dtype='object'), if it is not the correct header, please check the file or with correct format specified.\n",
      "[INFO]  Loaded sumstats for 3167 SNPs in 0.11 seconds\n",
      "[INFO]  filtering SNP by finemap_tools with 3167 SNP at begining........\n",
      "Function find_amibiguous_alleles elapsed time: 0.0251 seconds\n",
      "[INFO]  drop 380 ambiguous alleles\n",
      "Function find_Biallelic_snp elapsed time: 0.0022 seconds\n",
      "[INFO]  drop 2 biallelic snps\n",
      "[INFO]  after filtering, left 2785 SNP\n",
      "[INFO]  loading GTExV8 format sumstats file with /home/xutingfeng/GIFT/GTEx_V8/ge/Adipose_Subcutaneous.tsv.gz......\n",
      "Will load region from file.\n",
      "[INFO]  the header of loaded file is Index(['variant', 'r2', 'pvalue', 'molecular_trait_object_id',\n",
      "       'molecular_trait_id', 'maf', 'gene_id', 'median_tpm', 'beta', 'se',\n",
      "       'an', 'ac', 'chromosome', 'position', 'ref', 'alt', 'type', 'rsid'],\n",
      "      dtype='object'), if it is not the correct header, please check the file or with correct format specified.\n",
      "[INFO]  filtering SNP by finemap_tools with 62822 SNP at begining........\n",
      "Function find_amibiguous_alleles elapsed time: 0.4110 seconds\n",
      "[INFO]  drop 7285 ambiguous alleles\n",
      "Function find_Biallelic_snp elapsed time: 0.0019 seconds\n",
      "[INFO]  drop 1 biallelic snps\n",
      "[INFO]  after filtering, left 55491 SNP\n",
      "[INFO]  na numbers in sumstats Adipose_Subcutaneous.tsv.gz: 0\n",
      "[INFO]  run coloc with 2 causal SNPs, this will run coloc.susie\n",
      "[INFO]  Loading LD file /home/xutingfeng/GIFT/data/analysis/DNAJC16/polyfun/ld/DNAJC16_GRCh38.1.14403350.15833356\n",
      "[INFO]  Done in 5.79 seconds\n",
      "[WARNING]  there are 10291 nan in R matrix in the ld_arr\n",
      "[INFO]  remove 0 SNPs with NaNs in LD, this should be think carefully.\n",
      "[WARNING]  15 variants with sumstats were not found in the LD file and will be omitted (please note that this may lead to false positives if the omitted SNPs are causal!)\n",
      "[WARNING]  4 of the missing variants were indels. Check that the allele order (A1/A2) matches between the sumstats and the LD file. Also see the flag --allow-swapped-indel-alleles\n",
      "[INFO]  before sync , sumstats1 have 2785 SNPs, qtl have 55491 SNPs\n",
      "[INFO]  after sync,sumstats1 left 1314 SNPs, qtl 54597 SNPs\n",
      "[INFO]  df_sumstats2 finally inlcude 54597 rows\n",
      "[INFO]  cffi mode is CFFI_MODE.ANY\n",
      "[DEBUG]  Looking for R home with: R RHOME\n",
      "[INFO]  R home found: /home/xutingfeng/miniforge3/envs/finemap/lib/R\n",
      "[DEBUG]  Looking for LD_LIBRARY_PATH with: /home/xutingfeng/miniforge3/envs/finemap/lib/R/bin/Rscript -e cat(Sys.getenv(\"LD_LIBRARY_PATH\"))\n",
      "[INFO]  R library path: \n",
      "[INFO]  LD_LIBRARY_PATH: \n",
      "[DEBUG]  cffi mode is InterfaceType.API\n",
      "[INFO]  Default options to initialize R: rpy2, --quiet, --no-save\n",
      "[INFO]  R is already initialized. No need to initialize.\n",
      "[WARNING]  R[write to console]: This is coloc version 5.2.3\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "\n",
      "[WARNING]  R[write to console]: Registered S3 methods overwritten by 'treeio':\n",
      "  method              from    \n",
      "  MRCA.phylo          tidytree\n",
      "  MRCA.treedata       tidytree\n",
      "  Nnode.treedata      tidytree\n",
      "  Ntip.treedata       tidytree\n",
      "  ancestor.phylo      tidytree\n",
      "  ancestor.treedata   tidytree\n",
      "  child.phylo         tidytree\n",
      "  child.treedata      tidytree\n",
      "  full_join.phylo     tidytree\n",
      "  full_join.treedata  tidytree\n",
      "  groupClade.phylo    tidytree\n",
      "  groupClade.treedata tidytree\n",
      "  groupOTU.phylo      tidytree\n",
      "  groupOTU.treedata   tidytree\n",
      "  is.rooted.treedata  tidytree\n",
      "  nodeid.phylo        tidytree\n",
      "  nodeid.treedata     tidytree\n",
      "  nodelab.phylo       tidytree\n",
      "  nodelab.treedata    tidytree\n",
      "  offspring.phylo     tidytree\n",
      "  offspring.treedata  tidytree\n",
      "  parent.phylo        tidytree\n",
      "  parent.treedata     tidytree\n",
      "  root.treedata       tidytree\n",
      "  rootnode.phylo      tidytree\n",
      "  sibling.phylo       tidytree\n",
      "\n",
      "[WARNING]  R[write to console]: clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: Loading required package: AnnotationDbi\n",
      "\n",
      "[WARNING]  R[write to console]: Loading required package: stats4\n",
      "\n",
      "[WARNING]  R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: Loading required package: Biobase\n",
      "\n",
      "[WARNING]  R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: Loading required package: IRanges\n",
      "\n",
      "[WARNING]  R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "[WARNING]  R[write to console]: \n",
      "\n",
      "[INFO]  run coloc with ENSG00000037637 as xQTL\n",
      "[1] \"gene_id column found in sumstats2, using gene_id to split the data\"\n",
      "[1] \"Number of SNPs in common:1311\"\n",
      "[1] 1311 1311\n",
      "[1] 1311   24\n",
      "[WARNING]  R[write to console]: running max iterations: 100\n",
      "\n",
      "[WARNING]  R[write to console]: \tconverged: TRUE\n",
      "\n",
      "[WARNING]  R[write to console]: running max iterations: 100\n",
      "\n",
      "[WARNING]  R[write to console]: \tconverged: TRUE\n",
      "\n",
      "NULL\n",
      "[WARNING]  R[write to console]: Error in (function (sumstats1, sumstats2, output_folder, n1, n2, sdY1,  : \n",
      "  object 're' not found\n",
      "\n",
      "[WARNING]  R[write to console]: In addition: \n",
      "[WARNING]  R[write to console]: Warning messages:\n",
      "\n",
      "[WARNING]  R[write to console]: 1: \n",
      "[WARNING]  R[write to console]: In check_dataset(dataset1) :\n",
      "[WARNING]  R[write to console]:  minimum p value is: 0.0030845\n",
      "If this is what you expected, this is not a problem.\n",
      "If this is not as small as you expected, please check you supplied var(beta) and not sd(beta) for the varbeta argument. If that's not the explanation, please check the 02_data vignette.\n",
      "\n",
      "[WARNING]  R[write to console]: 2: \n",
      "[WARNING]  R[write to console]: In check_dataset(d, suffix, req = c(\"beta\", \"varbeta\", \"LD\", \"snp\",  :\n",
      "[WARNING]  R[write to console]: \n",
      " \n",
      "[WARNING]  R[write to console]:  minimum p value is: 0.0030845\n",
      "If this is what you expected, this is not a problem.\n",
      "If this is not as small as you expected, please check you supplied var(beta) and not sd(beta) for the varbeta argument. If that's not the explanation, please check the 02_data vignette.\n",
      "\n",
      "[WARNING]  R[write to console]: 3: ggrepel: 1 unlabeled data points (too many overlaps). Consider increasing max.overlaps \n",
      "\n"
     ]
    },
    {
     "ename": "RRuntimeError",
     "evalue": "Error in (function (sumstats1, sumstats2, output_folder, n1, n2, sdY1,  : \n  object 're' not found\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 16\u001b[0m\n\u001b[1;32m      1\u001b[0m coloc \u001b[38;5;241m=\u001b[39m coloc_Wrapper(\n\u001b[1;32m      2\u001b[0m     sumstats_file\u001b[38;5;241m=\u001b[39mgwas_file,\n\u001b[1;32m      3\u001b[0m     sumstats_file_format\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpheweb\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     14\u001b[0m     end\u001b[38;5;241m=\u001b[39mlocus_range_tuple[\u001b[38;5;241m2\u001b[39m],\n\u001b[1;32m     15\u001b[0m )\n\u001b[0;32m---> 16\u001b[0m s1 \u001b[38;5;241m=\u001b[39m \u001b[43mcoloc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfinemap\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_causal_snps\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[43m    \u001b[49m\u001b[43mout_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m./test\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m    \u001b[49m\u001b[43mld_file\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/home/xutingfeng/GIFT/data/analysis/DNAJC16/polyfun/ld/DNAJC16_GRCh38.1.14403350.15833356.npz\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m    \u001b[49m\u001b[43mallow_missing\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     21\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/github_code/mine/finemap_tools/finemap_tools/finemap/coloc.py:506\u001b[0m, in \u001b[0;36mfinemap\u001b[0;34m(self, num_causal_snps, out_dir, gene_col, ld_file, verbose, allow_missing, save_all)\u001b[0m\n\u001b[1;32m    504\u001b[0m     )\n\u001b[1;32m    505\u001b[0m logging.info(f\"\"\n\u001b[0;32m--> 506\u001b[0m current_res = r.coloc_analysis(\n\u001b[1;32m    507\u001b[0m     sumstats1=current_df1,\n\u001b[1;32m    508\u001b[0m     sumstats2=current_df2,\n",
      "File \u001b[0;32m~/miniforge3/envs/finemap/lib/python3.8/site-packages/rpy2/robjects/functions.py:208\u001b[0m, in \u001b[0;36mSignatureTranslatedFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    206\u001b[0m         v \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mpop(k)\n\u001b[1;32m    207\u001b[0m         kwargs[r_k] \u001b[38;5;241m=\u001b[39m v\n\u001b[0;32m--> 208\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m (\u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mSignatureTranslatedFunction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    209\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m~/miniforge3/envs/finemap/lib/python3.8/site-packages/rpy2/robjects/functions.py:131\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    129\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    130\u001b[0m         new_kwargs[k] \u001b[38;5;241m=\u001b[39m cv\u001b[38;5;241m.\u001b[39mpy2rpy(v)\n\u001b[0;32m--> 131\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mFunction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    132\u001b[0m res \u001b[38;5;241m=\u001b[39m cv\u001b[38;5;241m.\u001b[39mrpy2py(res)\n\u001b[1;32m    133\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m res\n",
      "File \u001b[0;32m~/miniforge3/envs/finemap/lib/python3.8/site-packages/rpy2/rinterface_lib/conversion.py:45\u001b[0m, in \u001b[0;36m_cdata_res_to_rinterface.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 45\u001b[0m     cdata \u001b[38;5;241m=\u001b[39m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m     \u001b[38;5;66;03m# TODO: test cdata is of the expected CType\u001b[39;00m\n\u001b[1;32m     47\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _cdata_to_rinterface(cdata)\n",
      "File \u001b[0;32m~/miniforge3/envs/finemap/lib/python3.8/site-packages/rpy2/rinterface.py:817\u001b[0m, in \u001b[0;36mSexpClosure.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    810\u001b[0m     res \u001b[38;5;241m=\u001b[39m rmemory\u001b[38;5;241m.\u001b[39mprotect(\n\u001b[1;32m    811\u001b[0m         openrlib\u001b[38;5;241m.\u001b[39mrlib\u001b[38;5;241m.\u001b[39mR_tryEval(\n\u001b[1;32m    812\u001b[0m             call_r,\n\u001b[1;32m    813\u001b[0m             call_context\u001b[38;5;241m.\u001b[39m__sexp__\u001b[38;5;241m.\u001b[39m_cdata,\n\u001b[1;32m    814\u001b[0m             error_occured)\n\u001b[1;32m    815\u001b[0m     )\n\u001b[1;32m    816\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m error_occured[\u001b[38;5;241m0\u001b[39m]:\n\u001b[0;32m--> 817\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m embedded\u001b[38;5;241m.\u001b[39mRRuntimeError(_rinterface\u001b[38;5;241m.\u001b[39m_geterrmessage())\n\u001b[1;32m    818\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m res\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in (function (sumstats1, sumstats2, output_folder, n1, n2, sdY1,  : \n  object 're' not found\n"
     ]
    }
   ],
   "source": [
    "coloc = coloc_Wrapper(\n",
    "    sumstats_file=gwas_file,\n",
    "    sumstats_file_format=\"pheweb\",\n",
    "    sumstats_file2=eQTL_path,\n",
    "    sumstats_file2_format=\"GTExV8\",\n",
    "    n1=4786,\n",
    "    n2=73,\n",
    "    type1=\"quant\",\n",
    "    type2=\"quant\",\n",
    "    sdY1=1,\n",
    "    sdY2=None,\n",
    "    chr_num=locus_range_tuple[0],\n",
    "    start=locus_range_tuple[1],\n",
    "    end=locus_range_tuple[2],\n",
    ")\n",
    "s1 = coloc.finemap(\n",
    "    num_causal_snps=2,\n",
    "    out_dir=\"./test\",\n",
    "    ld_file=\"/home/xutingfeng/GIFT/data/analysis/DNAJC16/polyfun/ld/DNAJC16_GRCh38.1.14403350.15833356.npz\",\n",
    "    allow_missing=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  0\n",
       "0  0  0\n",
       "1  1  1\n",
       "2  2  2\n",
       "3  3  3\n",
       "4  4  4\n",
       "5  5  5\n",
       "6  6  6\n",
       "7  7  7\n",
       "8  8  8\n",
       "9  9  9"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pd.DataFrame(range(10)).reset_index(drop=False, names=[\"a\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ld.loc[common, common]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ld.reset_index(drop=False).to_feather(f\"{save_dir}/ld.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rpy2 import robjects\n",
    "\n",
    "robjects.r(\"NULL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ld.index.name = \"\"\n",
    "ld.columns.name = \"\"\n",
    "ld"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc.sumstats_file2[coloc.sumstats_file2[\"gene_id\"] == \"ENSG00000132906\"].sort_values(\n",
    "    \"pvalue\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc.sumstats_file2[\"gene_id\"].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_col = \"gene_id\"\n",
    "# drop duplicated SNPs\n",
    "sumstats1_duplicated_num = coloc.sumstats_file1.duplicated(subset=[\"snp\"]).sum()\n",
    "sumstats2_duplicated_num = coloc.sumstats_file2.duplicated(\n",
    "    subset=[\"snp\", gene_col]\n",
    ").sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if sumstats1_duplicated_num > 0:\n",
    "#     logging.warning(\n",
    "#         f\"sumstats1 have {sumstats1_duplicated_num} duplicated SNPs and will be dropped\"\n",
    "#     )\n",
    "# if sumstats2_duplicated_num > 0:\n",
    "#     logging.warning(\n",
    "#         f\"sumstats2 have {sumstats2_duplicated_num} duplicated SNPs and will be dropped\"\n",
    "#     )\n",
    "\n",
    "coloc.sumstats_file1 = coloc.sumstats_file1.drop_duplicates(subset=[\"snp\"])\n",
    "coloc.sumstats_file2 = coloc.sumstats_file2.drop_duplicates(subset=[\"snp\", gene_col])\n",
    "\n",
    "assert (\n",
    "    coloc.sumstats_file1.shape[0] > 0 and coloc.sumstats_file2.shape[0] > 0\n",
    "), \"no valid SNPs to run coloc after drop duplicated SNPs\"\n",
    "\n",
    "# sync sumstats\n",
    "coloc.sync_sumstats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc.sumstats_file2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_duplicated"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "finemap",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
