{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f461b4a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/xutingfeng/miniforge3/envs/postGWAS/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n"
     ]
    }
   ],
   "source": [
    "import gwaslab as gl \n",
    "import genal\n",
    "from config import *\n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fac75e1",
   "metadata": {},
   "source": [
    "\n",
    "## Load Data\n",
    "\n",
    "1. The `gl.Sumstats` module supports loading data from various sources such as directories, Pandas DataFrames, and more.\n",
    "2. You can build a pipeline to convert and save raw summary statistics into supported formats. Several commonly used formats are already supported — see the [formats directory](https://github.com/Cloufield/formatbook/tree/main/formats) for details (e.g., GWAS Catalog format).\n",
    "\n",
    "> **Note:** When working with large datasets, using CSV or gzip-compressed files can be time-consuming. It is highly recommended to preprocess your data and save it in the `feather` format for significantly faster loading and downstream analysis.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e17fc7ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "      <th>rs_id</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.003855</td>\n",
       "      <td>0.002427</td>\n",
       "      <td>0.1637</td>\n",
       "      <td>0.112194</td>\n",
       "      <td>rs3094315</td>\n",
       "      <td>663139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>885689</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.012647</td>\n",
       "      <td>0.004002</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>0.001575</td>\n",
       "      <td>rs4970452</td>\n",
       "      <td>656956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>885699</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.012867</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>0.001216</td>\n",
       "      <td>rs4970376</td>\n",
       "      <td>656954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>886006</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.012928</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>0.001164</td>\n",
       "      <td>rs4970375</td>\n",
       "      <td>656955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>887801</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.012811</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>0.001278</td>\n",
       "      <td>rs3828047</td>\n",
       "      <td>656954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753790</th>\n",
       "      <td>22</td>\n",
       "      <td>51172460</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.004041</td>\n",
       "      <td>0.005831</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>0.488332</td>\n",
       "      <td>rs5770824</td>\n",
       "      <td>715356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753791</th>\n",
       "      <td>22</td>\n",
       "      <td>51175626</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001762</td>\n",
       "      <td>0.003547</td>\n",
       "      <td>0.0613</td>\n",
       "      <td>0.619277</td>\n",
       "      <td>rs3810648</td>\n",
       "      <td>739079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753792</th>\n",
       "      <td>22</td>\n",
       "      <td>51177257</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.000743</td>\n",
       "      <td>0.004725</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>0.875139</td>\n",
       "      <td>rs73174437</td>\n",
       "      <td>713374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753793</th>\n",
       "      <td>22</td>\n",
       "      <td>51178090</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.001494</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>0.0549</td>\n",
       "      <td>0.689183</td>\n",
       "      <td>rs2285395</td>\n",
       "      <td>739077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753794</th>\n",
       "      <td>22</td>\n",
       "      <td>51182399</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.003232</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.854675</td>\n",
       "      <td>rs56238942</td>\n",
       "      <td>720624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5753795 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome  base_pair_location effect_allele other_allele      beta  \\\n",
       "0                 1              752566             G            A -0.003855   \n",
       "1                 1              885689             G            A -0.012647   \n",
       "2                 1              885699             A            G -0.012867   \n",
       "3                 1              886006             T            C -0.012928   \n",
       "4                 1              887801             A            G -0.012811   \n",
       "...             ...                 ...           ...          ...       ...   \n",
       "5753790          22            51172460             C            T -0.004041   \n",
       "5753791          22            51175626             G            A  0.001762   \n",
       "5753792          22            51177257             T            C -0.000743   \n",
       "5753793          22            51178090             A            G  0.001494   \n",
       "5753794          22            51182399             G            A  0.000592   \n",
       "\n",
       "         standard_error  effect_allele_frequency   p_value       rs_id       n  \n",
       "0              0.002427                   0.1637  0.112194   rs3094315  663139  \n",
       "1              0.004002                   0.0537  0.001575   rs4970452  656956  \n",
       "2              0.003977                   0.0544  0.001216   rs4970376  656954  \n",
       "3              0.003981                   0.0543  0.001164   rs4970375  656955  \n",
       "4              0.003977                   0.0544  0.001278   rs3828047  656954  \n",
       "...                 ...                      ...       ...         ...     ...  \n",
       "5753790        0.005831                   0.0225  0.488332   rs5770824  715356  \n",
       "5753791        0.003547                   0.0613  0.619277   rs3810648  739079  \n",
       "5753792        0.004725                   0.0348  0.875139  rs73174437  713374  \n",
       "5753793        0.003735                   0.0549  0.689183   rs2285395  739077  \n",
       "5753794        0.003232                   0.0770  0.854675  rs56238942  720624  \n",
       "\n",
       "[5753795 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "filename = \"GCST90449056.tsv.gz\"\n",
    "data = pd.read_csv(data_dir / filename, sep = r\"\\s+\", compression=\"gzip\")\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b6bc10b",
   "metadata": {},
   "source": [
    "This summary statistics file was downloaded directly from the GWAS Catalog, it's already in a supported format. That means you can load it without needing additional conversion:\n",
    "\n",
    ">Note: This file was downloaded from the GWAS Catalog, so we can load it directly using `gl.Sumstats()`.\n",
    "\n",
    "As rsid may not always be present in the summary statistics, so we can check or "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6698cdb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/06/10 01:27:22 GWASLab v3.6.3 https://cloufield.github.io/gwaslab/\n",
      "2025/06/10 01:27:22 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/06/10 01:27:22 Python version: 3.9.23 | packaged by conda-forge | (main, Jun  4 2025, 17:57:12) \n",
      "[GCC 13.3.0]\n",
      "2025/06/10 01:27:22 Start to load format from formatbook....\n",
      "2025/06/10 01:27:22  -gwascatalog format meta info:\n",
      "2025/06/10 01:27:22   - format_name  : gwascatalog\n",
      "2025/06/10 01:27:22   - format_source  : https://www.ebi.ac.uk/gwas/docs/methods/summary-statistics\n",
      "2025/06/10 01:27:22   - format_version  : 20220726\n",
      "2025/06/10 01:27:22  -gwascatalog to gwaslab format dictionary:\n",
      "2025/06/10 01:27:22   - gwascatalog keys: variant_id,chromosome,base_pair_location,other_allele,effect_allele,beta,effect_allele_frequency,standard_error,p-value,odds_ratio,ci_lower,ci_upper\n",
      "2025/06/10 01:27:22   - gwaslab values: SNPID,CHR,POS,NEA,EA,BETA,EAF,SE,P,OR,OR_95L,OR_95U\n",
      "2025/06/10 01:27:22 Start to initialize gl.Sumstats from pandas DataFrame ...\n",
      "2025/06/10 01:27:24  -Reading columns          : other_allele,beta,chromosome,rs_id,effect_allele_frequency,standard_error,base_pair_location,n,effect_allele\n",
      "2025/06/10 01:27:24  -Renaming columns to      : NEA,BETA,CHR,rsID,EAF,SE,POS,N,EA\n",
      "2025/06/10 01:27:24  -Current Dataframe shape : 5753795  x  9\n",
      "2025/06/10 01:27:24  -Initiating a status column: STATUS ...\n",
      "2025/06/10 01:27:24  -Genomic coordinates are based on GRCh38/hg38...\n",
      "2025/06/10 01:27:25 Start to reorder the columns...v3.6.3\n",
      "2025/06/10 01:27:25  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 367.16 MB\n",
      "2025/06/10 01:27:25  -Reordering columns to    : rsID,CHR,POS,EA,NEA,EAF,BETA,SE,N,STATUS\n",
      "2025/06/10 01:27:25 Finished reordering the columns.\n",
      "2025/06/10 01:27:25  -Column  : rsID   CHR   POS   EA       NEA      EAF     BETA    SE      N     STATUS  \n",
      "2025/06/10 01:27:25  -DType   : object Int64 int64 category category float64 float64 float64 int64 category\n",
      "2025/06/10 01:27:25  -Verified: T      T     T     T        T        T       T       T       T     T       \n",
      "2025/06/10 01:27:25  -Current Dataframe memory usage: 367.16 MB\n",
      "2025/06/10 01:27:25 Finished loading data successfully!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1637</td>\n",
       "      <td>-0.003855</td>\n",
       "      <td>0.002427</td>\n",
       "      <td>663139</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs4970452</td>\n",
       "      <td>1</td>\n",
       "      <td>885689</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>-0.012647</td>\n",
       "      <td>0.004002</td>\n",
       "      <td>656956</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4970376</td>\n",
       "      <td>1</td>\n",
       "      <td>885699</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012867</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4970375</td>\n",
       "      <td>1</td>\n",
       "      <td>886006</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>-0.012928</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>656955</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3828047</td>\n",
       "      <td>1</td>\n",
       "      <td>887801</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012811</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753790</th>\n",
       "      <td>rs5770824</td>\n",
       "      <td>22</td>\n",
       "      <td>51172460</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>-0.004041</td>\n",
       "      <td>0.005831</td>\n",
       "      <td>715356</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753791</th>\n",
       "      <td>rs3810648</td>\n",
       "      <td>22</td>\n",
       "      <td>51175626</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0613</td>\n",
       "      <td>0.001762</td>\n",
       "      <td>0.003547</td>\n",
       "      <td>739079</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753792</th>\n",
       "      <td>rs73174437</td>\n",
       "      <td>22</td>\n",
       "      <td>51177257</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>-0.000743</td>\n",
       "      <td>0.004725</td>\n",
       "      <td>713374</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753793</th>\n",
       "      <td>rs2285395</td>\n",
       "      <td>22</td>\n",
       "      <td>51178090</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0549</td>\n",
       "      <td>0.001494</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>739077</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753794</th>\n",
       "      <td>rs56238942</td>\n",
       "      <td>22</td>\n",
       "      <td>51182399</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.003232</td>\n",
       "      <td>720624</td>\n",
       "      <td>3899999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5753795 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               rsID  CHR       POS EA NEA     EAF      BETA        SE       N  \\\n",
       "0         rs3094315    1    752566  G   A  0.1637 -0.003855  0.002427  663139   \n",
       "1         rs4970452    1    885689  G   A  0.0537 -0.012647  0.004002  656956   \n",
       "2         rs4970376    1    885699  A   G  0.0544 -0.012867  0.003977  656954   \n",
       "3         rs4970375    1    886006  T   C  0.0543 -0.012928  0.003981  656955   \n",
       "4         rs3828047    1    887801  A   G  0.0544 -0.012811  0.003977  656954   \n",
       "...             ...  ...       ... ..  ..     ...       ...       ...     ...   \n",
       "5753790   rs5770824   22  51172460  C   T  0.0225 -0.004041  0.005831  715356   \n",
       "5753791   rs3810648   22  51175626  G   A  0.0613  0.001762  0.003547  739079   \n",
       "5753792  rs73174437   22  51177257  T   C  0.0348 -0.000743  0.004725  713374   \n",
       "5753793   rs2285395   22  51178090  A   G  0.0549  0.001494  0.003735  739077   \n",
       "5753794  rs56238942   22  51182399  G   A  0.0770  0.000592  0.003232  720624   \n",
       "\n",
       "          STATUS  \n",
       "0        3899999  \n",
       "1        3899999  \n",
       "2        3899999  \n",
       "3        3899999  \n",
       "4        3899999  \n",
       "...          ...  \n",
       "5753790  3899999  \n",
       "5753791  3899999  \n",
       "5753792  3899999  \n",
       "5753793  3899999  \n",
       "5753794  3899999  \n",
       "\n",
       "[5753795 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define fmt_dict to map columns\n",
    "fmt_dict = dict(\n",
    "    fmt = \"gwascatalog\",\n",
    "    build = \"38\"  # check genome build or versionConvert in the beginning\n",
    ")\n",
    "if \"rs_id\" in data.columns:\n",
    "    fmt_dict[\"rsid\"] = \"rs_id\"\n",
    "if \"n\" in data.columns:\n",
    "    fmt_dict[\"n\"] = \"n\"\n",
    "\n",
    "if \"snpid\" in data.columns:\n",
    "    fmt_dict[\"snpid\"] = \"variant_id\"\n",
    "\n",
    "sumstats = gl.Sumstats(\n",
    "    sumstats=data,\n",
    "    **fmt_dict\n",
    ")\n",
    "sumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "434dd1ee",
   "metadata": {},
   "source": [
    "standarization of data to remove dup and fix some errors, so the data can be suitable for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "faa38971",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/06/10 01:28:44 Start to check SNPID/rsID...v3.6.3\n",
      "2025/06/10 01:28:44  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 356.19 MB\n",
      "2025/06/10 01:28:44  -Checking rsID data type...\n",
      "2025/06/10 01:28:44  -Checking if rsID is rsxxxxxx...\n",
      "2025/06/10 01:28:51  -Checking if CHR:POS:NEA:EA is mixed in rsID column ...\n",
      "2025/06/10 01:28:53  -Number of CHR:POS:NEA:EA mixed in rsID column : 0\n",
      "2025/06/10 01:28:54  -Number of Unrecognized rsID : 0\n",
      "2025/06/10 01:28:54  -A look at the unrecognized rsID : set() ...\n",
      "2025/06/10 01:28:54 Finished checking SNPID/rsID.\n",
      "2025/06/10 01:28:54 Start to fix chromosome notation (CHR)...v3.6.3\n",
      "2025/06/10 01:28:54  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 356.19 MB\n",
      "2025/06/10 01:28:54  -Checking CHR data type...\n",
      "2025/06/10 01:28:55  -Variants with standardized chromosome notation: 5753795\n",
      "2025/06/10 01:28:56  -All CHR are already fixed...\n",
      "2025/06/10 01:29:01 Finished fixing chromosome notation (CHR).\n",
      "2025/06/10 01:29:01 Start to fix basepair positions (POS)...v3.6.3\n",
      "2025/06/10 01:29:01  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 356.19 MB\n",
      "2025/06/10 01:29:01  -Converting to Int64 data type ...\n",
      "2025/06/10 01:29:07  -Position bound:(0 , 250,000,000)\n",
      "2025/06/10 01:29:07  -Removed outliers: 0\n",
      "2025/06/10 01:29:08  -Removed 0 variants with bad positions.\n",
      "2025/06/10 01:29:08 Finished fixing basepair positions (POS).\n",
      "2025/06/10 01:29:08 Start to fix alleles (EA and NEA)...v3.6.3\n",
      "2025/06/10 01:29:08  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:08  -Converted all bases to string datatype and UPPERCASE.\n",
      "2025/06/10 01:29:09  -Variants with bad EA  : 0\n",
      "2025/06/10 01:29:10  -Variants with bad NEA : 0\n",
      "2025/06/10 01:29:10  -Variants with NA for EA or NEA: 0\n",
      "2025/06/10 01:29:10  -Variants with same EA and NEA: 0\n",
      "2025/06/10 01:29:11  -Removed 0 variants with NA alleles or alleles that contain bases other than A/C/T/G.\n",
      "2025/06/10 01:29:12  -Removed 0 variants with same allele for EA and NEA.\n",
      "2025/06/10 01:29:17 Finished fixing alleles (EA and NEA).\n",
      "2025/06/10 01:29:18 Start to perform sanity check for statistics...v3.6.3\n",
      "2025/06/10 01:29:18  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:18  -Comparison tolerance for floats: 1e-07\n",
      "2025/06/10 01:29:18  -Checking if 0 <= N <= 2147483647 ...\n",
      "2025/06/10 01:29:19  -Removed 0 variants with bad/na N.\n",
      "2025/06/10 01:29:19  -Checking if -1e-07 < EAF < 1.0000001 ...\n",
      "2025/06/10 01:29:19  -Removed 0 variants with bad/na EAF.\n",
      "2025/06/10 01:29:19  -Checking if -100.0000001 < BETA < 100.0000001 ...\n",
      "2025/06/10 01:29:20  -Removed 0 variants with bad/na BETA.\n",
      "2025/06/10 01:29:20  -Checking if -1e-07 < SE < inf ...\n",
      "2025/06/10 01:29:21  -Removed 0 variants with bad/na SE.\n",
      "2025/06/10 01:29:21  -Checking STATUS and converting STATUS to categories....\n",
      "2025/06/10 01:29:22  -Removed 0 variants with bad statistics in total.\n",
      "2025/06/10 01:29:22  -Data types for each column:\n",
      "2025/06/10 01:29:22  -Column  : rsID   CHR   POS   EA       NEA      EAF     BETA    SE      N     STATUS  \n",
      "2025/06/10 01:29:22  -DType   : string Int64 Int64 category category float32 float64 float64 Int64 category\n",
      "2025/06/10 01:29:22  -Verified: T      T     T     T        T        T       T       T       T     T       \n",
      "2025/06/10 01:29:22 Finished sanity check for statistics.\n",
      "2025/06/10 01:29:22 Start to check data consistency across columns...v3.6.3\n",
      "2025/06/10 01:29:22  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:22  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2025/06/10 01:29:22  -No availalbe columns for data consistency checking...Skipping...\n",
      "2025/06/10 01:29:22 Finished checking data consistency across columns.\n",
      "2025/06/10 01:29:22 Start to normalize indels...v3.6.3\n",
      "2025/06/10 01:29:22  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:24  -No available variants to normalize..\n",
      "2025/06/10 01:29:24 Finished normalizing variants successfully!\n",
      "2025/06/10 01:29:24 Start to remove duplicated/multiallelic variants...v3.6.3\n",
      "2025/06/10 01:29:24  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:24  -Removing mode:dm\n",
      "2025/06/10 01:29:24 ColumnP was not detected... skipping... \n",
      "2025/06/10 01:29:24 Start to remove duplicated variants based on rsID...\n",
      "2025/06/10 01:29:24  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:28  -Removed  0  based on rsID...\n",
      "2025/06/10 01:29:28 Start to remove duplicated variants based on CHR,POS,EA and NEA...\n",
      "2025/06/10 01:29:28  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:28  -Which variant to keep:  first\n",
      "2025/06/10 01:29:28 Start to remove multiallelic variants based on chr:pos...\n",
      "2025/06/10 01:29:28  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 400.09 MB\n",
      "2025/06/10 01:29:28  -Which variant to keep:  first\n",
      "2025/06/10 01:29:31  -Removed  0  multiallelic variants...\n",
      "2025/06/10 01:29:31  -Removed  0  variants in total.\n",
      "2025/06/10 01:29:31  -Sort the coordinates based on CHR and POS...\n",
      "2025/06/10 01:29:35 Finished removing duplicated/multiallelic variants.\n",
      "2025/06/10 01:29:35 Start to sort the genome coordinates...v3.6.3\n",
      "2025/06/10 01:29:35  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 356.19 MB\n",
      "2025/06/10 01:29:38 Finished sorting coordinates.\n",
      "2025/06/10 01:29:38 Start to reorder the columns...v3.6.3\n",
      "2025/06/10 01:29:38  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 356.19 MB\n",
      "2025/06/10 01:29:38  -Reordering columns to    : rsID,CHR,POS,EA,NEA,EAF,BETA,SE,N,STATUS\n",
      "2025/06/10 01:29:38 Finished reordering the columns.\n"
     ]
    }
   ],
   "source": [
    "n_cores = 6 # or by number of cores\n",
    "sumstats.basic_check(remove=True,remove_dup=True, n_cores=n_cores)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3d2399a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1637</td>\n",
       "      <td>-0.003855</td>\n",
       "      <td>0.002427</td>\n",
       "      <td>663139</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs4970452</td>\n",
       "      <td>1</td>\n",
       "      <td>885689</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>-0.012647</td>\n",
       "      <td>0.004002</td>\n",
       "      <td>656956</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4970376</td>\n",
       "      <td>1</td>\n",
       "      <td>885699</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012867</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4970375</td>\n",
       "      <td>1</td>\n",
       "      <td>886006</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>-0.012928</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>656955</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3828047</td>\n",
       "      <td>1</td>\n",
       "      <td>887801</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012811</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753790</th>\n",
       "      <td>rs5770824</td>\n",
       "      <td>22</td>\n",
       "      <td>51172460</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>-0.004041</td>\n",
       "      <td>0.005831</td>\n",
       "      <td>715356</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753791</th>\n",
       "      <td>rs3810648</td>\n",
       "      <td>22</td>\n",
       "      <td>51175626</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0613</td>\n",
       "      <td>0.001762</td>\n",
       "      <td>0.003547</td>\n",
       "      <td>739079</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753792</th>\n",
       "      <td>rs73174437</td>\n",
       "      <td>22</td>\n",
       "      <td>51177257</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>-0.000743</td>\n",
       "      <td>0.004725</td>\n",
       "      <td>713374</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753793</th>\n",
       "      <td>rs2285395</td>\n",
       "      <td>22</td>\n",
       "      <td>51178090</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0549</td>\n",
       "      <td>0.001494</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>739077</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753794</th>\n",
       "      <td>rs56238942</td>\n",
       "      <td>22</td>\n",
       "      <td>51182399</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.003232</td>\n",
       "      <td>720624</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5753795 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               rsID  CHR       POS EA NEA     EAF      BETA        SE       N  \\\n",
       "0         rs3094315    1    752566  G   A  0.1637 -0.003855  0.002427  663139   \n",
       "1         rs4970452    1    885689  G   A  0.0537 -0.012647  0.004002  656956   \n",
       "2         rs4970376    1    885699  A   G  0.0544 -0.012867  0.003977  656954   \n",
       "3         rs4970375    1    886006  T   C  0.0543 -0.012928  0.003981  656955   \n",
       "4         rs3828047    1    887801  A   G  0.0544 -0.012811  0.003977  656954   \n",
       "...             ...  ...       ... ..  ..     ...       ...       ...     ...   \n",
       "5753790   rs5770824   22  51172460  C   T  0.0225 -0.004041  0.005831  715356   \n",
       "5753791   rs3810648   22  51175626  G   A  0.0613  0.001762  0.003547  739079   \n",
       "5753792  rs73174437   22  51177257  T   C  0.0348 -0.000743  0.004725  713374   \n",
       "5753793   rs2285395   22  51178090  A   G  0.0549  0.001494  0.003735  739077   \n",
       "5753794  rs56238942   22  51182399  G   A  0.0770  0.000592  0.003232  720624   \n",
       "\n",
       "          STATUS  \n",
       "0        3850099  \n",
       "1        3850099  \n",
       "2        3850099  \n",
       "3        3850099  \n",
       "4        3850099  \n",
       "...          ...  \n",
       "5753790  3850099  \n",
       "5753791  3850099  \n",
       "5753792  3850099  \n",
       "5753793  3850099  \n",
       "5753794  3850099  \n",
       "\n",
       "[5753795 rows x 10 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0a9a12f",
   "metadata": {},
   "source": [
    "Now this file is ok with some basic check and unified column names. \n",
    "\n",
    "Save to feather format for later use in python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "45a6eeaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCST90449056'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename.split(\".\")[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7a265f07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/06/10 01:38:43 Start to convert the output sumstats in:  gwaslab  format\n",
      "2025/06/10 01:38:44  -Start outputting sumstats in gwaslab format...\n",
      "2025/06/10 01:38:44  -gwaslab format will be loaded...\n",
      "2025/06/10 01:38:44  -gwaslab format meta info:\n",
      "2025/06/10 01:38:44   - format_name  : gwaslab\n",
      "2025/06/10 01:38:44   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2025/06/10 01:38:44   - format_version  : 20231220_v4\n",
      "2025/06/10 01:38:44  -Output path: /home/xutingfeng/job/finemap_tools/data/parquet/GCST90449056.gwaslab.parquet\n",
      "2025/06/10 01:38:44  -Output columns: rsID,CHR,POS,EA,NEA,EAF,BETA,SE,N,STATUS\n",
      "2025/06/10 01:38:44  -Writing sumstats to: /home/xutingfeng/job/finemap_tools/data/parquet/GCST90449056.gwaslab.parquet...\n",
      "2025/06/10 01:38:48  -Saving log file to: /home/xutingfeng/job/finemap_tools/data/parquet/GCST90449056.gwaslab.log\n",
      "2025/06/10 01:38:48 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "parquet_path =  data_dir / \"parquet\"\n",
    "\n",
    "parquet_path.mkdir(exist_ok=True, parents=True)\n",
    "to_file = str(parquet_path/filename.split(\".\")[0])\n",
    "sumstats.to_format(\n",
    "        path=to_file,\n",
    "        fmt=\"gwaslab\", \n",
    "   tab_fmt = \"parquet\" \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2fa88ee",
   "metadata": {},
   "source": [
    "## Reformat to other format\n",
    "\n",
    "Sometimes we need to reformat the data to other format, such as ldsc or other format to load in other software.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49db8143",
   "metadata": {},
   "source": [
    "### LDSC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "46e53795",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1637</td>\n",
       "      <td>-0.003855</td>\n",
       "      <td>0.002427</td>\n",
       "      <td>663139</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs4970452</td>\n",
       "      <td>1</td>\n",
       "      <td>885689</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>-0.012647</td>\n",
       "      <td>0.004002</td>\n",
       "      <td>656956</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4970376</td>\n",
       "      <td>1</td>\n",
       "      <td>885699</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012867</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4970375</td>\n",
       "      <td>1</td>\n",
       "      <td>886006</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>-0.012928</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>656955</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3828047</td>\n",
       "      <td>1</td>\n",
       "      <td>887801</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012811</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753790</th>\n",
       "      <td>rs5770824</td>\n",
       "      <td>22</td>\n",
       "      <td>51172460</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>-0.004041</td>\n",
       "      <td>0.005831</td>\n",
       "      <td>715356</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753791</th>\n",
       "      <td>rs3810648</td>\n",
       "      <td>22</td>\n",
       "      <td>51175626</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0613</td>\n",
       "      <td>0.001762</td>\n",
       "      <td>0.003547</td>\n",
       "      <td>739079</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753792</th>\n",
       "      <td>rs73174437</td>\n",
       "      <td>22</td>\n",
       "      <td>51177257</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>-0.000743</td>\n",
       "      <td>0.004725</td>\n",
       "      <td>713374</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753793</th>\n",
       "      <td>rs2285395</td>\n",
       "      <td>22</td>\n",
       "      <td>51178090</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0549</td>\n",
       "      <td>0.001494</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>739077</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753794</th>\n",
       "      <td>rs56238942</td>\n",
       "      <td>22</td>\n",
       "      <td>51182399</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.003232</td>\n",
       "      <td>720624</td>\n",
       "      <td>3850099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5753795 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               rsID CHR       POS EA NEA     EAF      BETA        SE       N  \\\n",
       "0         rs3094315   1    752566  G   A  0.1637 -0.003855  0.002427  663139   \n",
       "1         rs4970452   1    885689  G   A  0.0537 -0.012647  0.004002  656956   \n",
       "2         rs4970376   1    885699  A   G  0.0544 -0.012867  0.003977  656954   \n",
       "3         rs4970375   1    886006  T   C  0.0543 -0.012928  0.003981  656955   \n",
       "4         rs3828047   1    887801  A   G  0.0544 -0.012811  0.003977  656954   \n",
       "...             ...  ..       ... ..  ..     ...       ...       ...     ...   \n",
       "5753790   rs5770824  22  51172460  C   T  0.0225 -0.004041  0.005831  715356   \n",
       "5753791   rs3810648  22  51175626  G   A  0.0613  0.001762  0.003547  739079   \n",
       "5753792  rs73174437  22  51177257  T   C  0.0348 -0.000743  0.004725  713374   \n",
       "5753793   rs2285395  22  51178090  A   G  0.0549  0.001494  0.003735  739077   \n",
       "5753794  rs56238942  22  51182399  G   A  0.0770  0.000592  0.003232  720624   \n",
       "\n",
       "          STATUS  \n",
       "0        3850099  \n",
       "1        3850099  \n",
       "2        3850099  \n",
       "3        3850099  \n",
       "4        3850099  \n",
       "...          ...  \n",
       "5753790  3850099  \n",
       "5753791  3850099  \n",
       "5753792  3850099  \n",
       "5753793  3850099  \n",
       "5753794  3850099  \n",
       "\n",
       "[5753795 rows x 10 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"GCST90449056.gwaslab.parquet\"\n",
    "to_load_file = parquet_path / filename\n",
    "\n",
    "data = pd.read_parquet(to_load_file)\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60b542f5",
   "metadata": {},
   "source": [
    "Now turn it to gl.Sumstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b324c1d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/06/10 01:42:11 GWASLab v3.6.3 https://cloufield.github.io/gwaslab/\n",
      "2025/06/10 01:42:11 (C) 2022-2025, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2025/06/10 01:42:11 Python version: 3.9.23 | packaged by conda-forge | (main, Jun  4 2025, 17:57:12) \n",
      "[GCC 13.3.0]\n",
      "2025/06/10 01:42:11 Start to load format from formatbook....\n",
      "2025/06/10 01:42:11  -gwaslab format meta info:\n",
      "2025/06/10 01:42:11   - format_name  : gwaslab\n",
      "2025/06/10 01:42:11   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2025/06/10 01:42:11   - format_version  : 20231220_v4\n",
      "2025/06/10 01:42:11 Start to initialize gl.Sumstats from pandas DataFrame ...\n",
      "2025/06/10 01:42:12  -Reading columns          : EAF,EA,CHR,POS,rsID,SE,NEA,STATUS,N,BETA\n",
      "2025/06/10 01:42:12  -Renaming columns to      : EAF,EA,CHR,POS,rsID,SE,NEA,STATUS,N,BETA\n",
      "2025/06/10 01:42:12  -Current Dataframe shape : 5753795  x  10\n",
      "2025/06/10 01:42:12  -Initiating a status column: STATUS ...\n",
      "2025/06/10 01:42:12  #WARNING! Version of genomic coordinates is unknown...\n",
      "2025/06/10 01:42:14 Start to reorder the columns...v3.6.3\n",
      "2025/06/10 01:42:14  -Current Dataframe shape : 5753795 x 10 ; Memory usage: 356.19 MB\n",
      "2025/06/10 01:42:14  -Reordering columns to    : rsID,CHR,POS,EA,NEA,EAF,BETA,SE,N,STATUS\n",
      "2025/06/10 01:42:14 Finished reordering the columns.\n",
      "2025/06/10 01:42:14  -Column  : rsID   CHR   POS   EA       NEA      EAF     BETA    SE      N     STATUS  \n",
      "2025/06/10 01:42:14  -DType   : string Int64 Int64 category category float32 float64 float64 Int64 category\n",
      "2025/06/10 01:42:14  -Verified: T      T     T     T        T        T       T       T       T     T       \n",
      "2025/06/10 01:42:14  -Current Dataframe memory usage: 356.19 MB\n",
      "2025/06/10 01:42:14 Finished loading data successfully!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>N</th>\n",
       "      <th>STATUS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>752566</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1637</td>\n",
       "      <td>-0.003855</td>\n",
       "      <td>0.002427</td>\n",
       "      <td>663139</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs4970452</td>\n",
       "      <td>1</td>\n",
       "      <td>885689</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0537</td>\n",
       "      <td>-0.012647</td>\n",
       "      <td>0.004002</td>\n",
       "      <td>656956</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4970376</td>\n",
       "      <td>1</td>\n",
       "      <td>885699</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012867</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4970375</td>\n",
       "      <td>1</td>\n",
       "      <td>886006</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0543</td>\n",
       "      <td>-0.012928</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>656955</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3828047</td>\n",
       "      <td>1</td>\n",
       "      <td>887801</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0544</td>\n",
       "      <td>-0.012811</td>\n",
       "      <td>0.003977</td>\n",
       "      <td>656954</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753790</th>\n",
       "      <td>rs5770824</td>\n",
       "      <td>22</td>\n",
       "      <td>51172460</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0225</td>\n",
       "      <td>-0.004041</td>\n",
       "      <td>0.005831</td>\n",
       "      <td>715356</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753791</th>\n",
       "      <td>rs3810648</td>\n",
       "      <td>22</td>\n",
       "      <td>51175626</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0613</td>\n",
       "      <td>0.001762</td>\n",
       "      <td>0.003547</td>\n",
       "      <td>739079</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753792</th>\n",
       "      <td>rs73174437</td>\n",
       "      <td>22</td>\n",
       "      <td>51177257</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0348</td>\n",
       "      <td>-0.000743</td>\n",
       "      <td>0.004725</td>\n",
       "      <td>713374</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753793</th>\n",
       "      <td>rs2285395</td>\n",
       "      <td>22</td>\n",
       "      <td>51178090</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0549</td>\n",
       "      <td>0.001494</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>739077</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5753794</th>\n",
       "      <td>rs56238942</td>\n",
       "      <td>22</td>\n",
       "      <td>51182399</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0770</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.003232</td>\n",
       "      <td>720624</td>\n",
       "      <td>9999999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5753795 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               rsID  CHR       POS EA NEA     EAF      BETA        SE       N  \\\n",
       "0         rs3094315    1    752566  G   A  0.1637 -0.003855  0.002427  663139   \n",
       "1         rs4970452    1    885689  G   A  0.0537 -0.012647  0.004002  656956   \n",
       "2         rs4970376    1    885699  A   G  0.0544 -0.012867  0.003977  656954   \n",
       "3         rs4970375    1    886006  T   C  0.0543 -0.012928  0.003981  656955   \n",
       "4         rs3828047    1    887801  A   G  0.0544 -0.012811  0.003977  656954   \n",
       "...             ...  ...       ... ..  ..     ...       ...       ...     ...   \n",
       "5753790   rs5770824   22  51172460  C   T  0.0225 -0.004041  0.005831  715356   \n",
       "5753791   rs3810648   22  51175626  G   A  0.0613  0.001762  0.003547  739079   \n",
       "5753792  rs73174437   22  51177257  T   C  0.0348 -0.000743  0.004725  713374   \n",
       "5753793   rs2285395   22  51178090  A   G  0.0549  0.001494  0.003735  739077   \n",
       "5753794  rs56238942   22  51182399  G   A  0.0770  0.000592  0.003232  720624   \n",
       "\n",
       "          STATUS  \n",
       "0        9999999  \n",
       "1        9999999  \n",
       "2        9999999  \n",
       "3        9999999  \n",
       "4        9999999  \n",
       "...          ...  \n",
       "5753790  9999999  \n",
       "5753791  9999999  \n",
       "5753792  9999999  \n",
       "5753793  9999999  \n",
       "5753794  9999999  \n",
       "\n",
       "[5753795 rows x 10 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    data, \n",
    "    fmt = \"gwaslab\"\n",
    ")\n",
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "134ffd95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025/06/10 01:44:06 Start to convert the output sumstats in:  ldsc  format\n",
      "2025/06/10 01:44:07  -Excluded 32666 variants in HLA region (chr6: 25000000-34000000 )...\n",
      "2025/06/10 01:44:07  -Formatting statistics ...\n",
      "2025/06/10 01:44:13  -Float statistics formats:\n",
      "2025/06/10 01:44:13   - Columns       : ['EAF', 'BETA', 'SE']\n",
      "2025/06/10 01:44:13   - Output formats: ['{:.4g}', '{:.4f}', '{:.4f}']\n",
      "2025/06/10 01:44:13  -Start outputting sumstats in ldsc format...\n",
      "2025/06/10 01:44:13  -ldsc format will be loaded...\n",
      "2025/06/10 01:44:13  -ldsc format meta info:\n",
      "2025/06/10 01:44:13   - format_name  : ldsc\n",
      "2025/06/10 01:44:13   - format_source  : https://github.com/bulik/ldsc/wiki/Summary-Statistics-File-Format\n",
      "2025/06/10 01:44:13   - format_source2  : https://github.com/bulik/ldsc/blob/master/munge_sumstats.py\n",
      "2025/06/10 01:44:13   - format_version  :  20150306\n",
      "2025/06/10 01:44:13  -gwaslab to ldsc format dictionary:\n",
      "2025/06/10 01:44:13   - gwaslab keys: rsID,NEA,EA,EAF,N,BETA,P,Z,INFO,OR,CHR,POS\n",
      "2025/06/10 01:44:13   - ldsc values: SNP,A2,A1,Frq,N,Beta,P,Z,INFO,OR,CHR,POS\n",
      "2025/06/10 01:44:13  -Output path: /home/xutingfeng/job/finemap_tools/data/ldsc/GCST90449056.noMHC.ldsc.tsv.gz\n",
      "2025/06/10 01:44:15  -Output columns: SNP,CHR,POS,A1,A2,Frq,Beta,N\n",
      "2025/06/10 01:44:15  -Writing sumstats to: /home/xutingfeng/job/finemap_tools/data/ldsc/GCST90449056.noMHC.ldsc.tsv.gz...\n",
      "2025/06/10 01:44:15   -Fast to csv mode...\n",
      "2025/06/10 01:44:21  -md5sum hashing for the file: /home/xutingfeng/job/finemap_tools/data/ldsc/GCST90449056.noMHC.ldsc.tsv.gz\n",
      "2025/06/10 01:44:21  -md5sum path: /home/xutingfeng/job/finemap_tools/data/ldsc/GCST90449056.noMHC.ldsc.tsv.gz.md5sum\n",
      "2025/06/10 01:44:21  -md5sum: 2bef770e0fb52e1b3188773807ab04f7\n",
      "2025/06/10 01:44:22  -Saving log file to: /home/xutingfeng/job/finemap_tools/data/ldsc/GCST90449056.noMHC.ldsc.log\n",
      "2025/06/10 01:44:22 Finished outputting successfully!\n"
     ]
    }
   ],
   "source": [
    "ldsc_dir = data_dir / \"ldsc\"\n",
    "ldsc_dir.mkdir(exist_ok=True, parents=True)\n",
    "to_file = str(ldsc_dir/filename.split(\".\")[0])\n",
    "mysumstats.to_format(to_file,fmt=\"ldsc\",hapmap3=False,exclude_hla=True,md5sum=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a8c0c868",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/xutingfeng/job/finemap_tools/data/ldsc/GCST90449056'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_file = str(ldsc_dir/filename.split(\".\")[0])\n",
    "to_file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b9d603",
   "metadata": {},
   "source": [
    "## other situation\n",
    "\n",
    "- if not GRCh38, use `versionConvert.py`\n",
    "- if no rsid , at least should have `variant_id` columns with `chr:pos:ref:alt` format [sorted of ref and alt] by `resetID2.py`.\n",
    "\n",
    "> why ref:alt should be sorted?  Reason: 1) ref and alt may are presented as A1 A2 or EA NEA etc, which we can not make sure which one are ref and alt or this is not important for variant name, as this is working as mapping between different variant name.\n",
    "\n",
    "For preprocessing stage, we recommend to be consistent format with columns or varaint name, but not preprocess with variants to keep or rm, which this should be done in downstream analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0dda814",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "postGWAS",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
